---
title: "README"
author:
output: github_document
date: "`r Sys.Date()`"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Quantifying Isoform-Level Diversity with lrRNA-Seq in WT Mouse Brain

## Authors

Emma Jones, TC Howton, Victoria Flanary, and Brittany Lasseigne. 

## Purpose 

The purpose of this project is to analyze Oxford Nanopore RNA sequencing from four wildtype mouse brain regions balanced for sex, to assay isoform usage differences across brain region and sex.

## Scripts 

"Scripts"
Include a "script tree" on how to reproduce your results with descriptions for all scripts

```{r echo = FALSE}
fs::dir_tree("src")
```

### Reproducibility

HPC user session info: I used the short partition with 1-2 nodes and 80GB per node, which is generous. None of my scripts take longer than 12 hours so I would stay on the short partition, unless you want to have one long session on another partition which is fine.

*Script 00: preprocessing/run_nanoseq.sh* - This script can be skipped if you are working from counts files. Does not require a docker. You need raw input fasta files, a genome file, and genome annotation file. I used GENCODE mouse release M31.

**Start here**

*Script 01: dtu_analyses/calculate_cpm.R* - This script is very short and depends on having transcript counts available either by downloading or running the nanoseq pipeline. Takes less than one minute.

*Script 02: preprocessing/get_genome_annotations.sh* - You need genome annotations available for gffread to run, but you also need genome annotations to run the nanoseq pipeline so frankly you should already have them somewhere and could just move them. You do need to get genome annotations if you didn't run the nanoseq pipeline and are just using counts, so I am keeping it as script 2. It took me 20 minutes to download.

*Script 03: preprocessing/create_isoform_fa.sh* - This script depends on script 02. It is a single command but needs to be run in a docker (either RStudio docker I made has gffread) or on a local machine with gffread.

*Script 04: dtu_analyses/pca_eda.Rmd* - This script is for exploratory data analysis and PCA. It depends on script 01. If data looks bad, do not proceed to script 05, but it is technically independent. This script takes 3 minutes to run.

*Script 05: dtu_analyses/dtu_region_region.Rmd* - This script depends on the outputs from script 01 which are read as an RDS file. This script also depends on script 03 - the gffread script. Run in github.1.2 docker so you have the most up-to-date version of the package. This script takes about 10 minutes to run.

*Script 06: dtu_analyses/dtu_region_others.Rmd* - This script is meant for DTU analysis comparing one brain region to all others. This script depends on scripts 01-03. Run this in the github.1.2 docker. Takes about 10 minutes to run.

*Script 07: dtu_analyses/dtu_region_sex.Rmd* - This script is meant for DTU analysis across sexes within brain regions. This script depends on scripts 01-03. Run this in the github.1.2 docker. Takes about 20 minutes to run (I think).

*Script 08: dtu_analyses/dtu_neuro_diseases.Rmd* - This script is for comparing brain-region-specific DTU genes to known Alzheimer's disease, psychiatric disorder, and CPAM case genes. It is dependent on scripts 01 - 06. It takes 4 minutes to run.

*Script 09: dtu_analyses/dtu_isoform_switching.Rmd* - This script is for adding open reading frames to the existing switchlist objects. It is also for plotting and saving switch plots, which show the significant isoform switching events. It is dependent on scripts 01-07. This script also takes 36 minutes to run.

*Script 10: de_analysis/DESeq2_region_region.Rmd* - This script is dependent on script 00/01 or having gene and transcript level count data with metadata in your /data/ directory. The purpose of this script is to run DESeq2 across brain regions at the gene and transcript level. It takes less than 5 minutes to run.

*Script 11: de_analysis/DESeq2_region_others.Rmd* - This script is dependent on script 00/01 or having gene and transcript level count data with metadata in your /data/ directory. The purpose of this script is to run DESeq2 for each brain region compared to an aggregate of other brain regions at the gene and transcript level. It takes about 10 minutes to run.

*Script 12: de_analysis/DESeq2_region_sex.Rmd* - This script is dependent on script 00/01 or having gene and transcript level count data with metadata in your /data/ directory. The purpose of this script is to compare expression at the gene and transcript level across sexes within brain regions. It takes less than 10 minutes to run.

*Script 13: de_analysis/incorporate_de_results.Rmd* - This script is fully dependent on scripts 01-09 and the de_analysis DESeq2 scripts 10-12. The purpose of this script is to wrap in all DESeq2 significance values into my isoformSwitchAnalyzeR objects and subsequently, plots. This enables us to plot them all with significance values for DGE, DTE, and DTU. This script takes 12 minutes to run.

*Script 14: dtu_analyses/protein_domain_info.Rmd* - This script depends on scripts 00-13, and also includes a shell script for running perl code. The purpose of thei script is to extract nucleotide and amino acid sequences and run pfam to annotate the protein domains. IT MUST RUN IN PFAM DOCKER!!!

*Script 15: de_analysis/compare_DTU_DGE.Rmd* - This script depends on scripts 00-14. The purpose of this script is to compare genes with differential gene expression, differential transcript usage, and differential transcript expression. It is not currently finished.

## Lasseigne Lab 

[What is Happening in the Lasseigne Lab?](https://www.lasseigne.org/)

<img src="https://www.lasseigne.org/img/main/lablogo.png" width="75" height="75">

## Funding 

NHGRI R00HG009678 (PI: Lasseigne)  
Pittman Scholar (PI: Lasseigne)  
UAB Lab Startup funds (PI: Lasseigne)  

## Acknowledgements 

We acknowledge the members of the Lasseigne Lab for their thoughtful feedback.

## License 

This repository is licensed under the MIT License, see LICENSE documentation within this repository for more details.
