---
title: "PCA Outlier Analysis"
author: "Emma Jones"
date: "2023-03-01"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Run PCA for Exploratory Data Analysis

Before performing DTU analysis, we want to ensure data is good quality. So, we will do some exploratory data analysis including principal component analysis to see what is influencing gene expression.

For time tracking, I will use proc.time to see how long this scripts takes.

```{r proc time}
ptm <- proc.time()
```

#### Load in packages

```{r load packages}
library(tidyverse)
library(styler)
library(lintr)
library(here)

set.seed(123)
```

#### Load in data

Next, I want to read in data and get cpm values. Since I'll be doing this for all 3 analyses, I made it an R script that can be sourced.

```{r read in data and get cpm values}
load(here("data", "cpm_out", "cpm_counts_metadata.RData"))
source("functions.R", local = knitr::knit_global())
```

### Run PCA

```{r run pca}
cpm_drop <- cpm[which(apply(cpm, 1, var) != 0), ] # drop rows with var = 0

pca_scaled <- prcomp(t(cpm_drop), scale = TRUE)

pca_metadata <- as_tibble(pca_scaled$x)
pca_metadata$sample_id <- rownames(pca_scaled$x)
pca_metadata <- left_join(pca_metadata, sample_collection_metadata)

var_explained <- pca_scaled$sdev^2 / sum(pca_scaled$sdev^2)
var_explained[1:5]
```

I expect the outputs from variance explained to be 0.06121201 0.03620078 0.03448017 0.03221352 0.03123769.

### Plot PCA

Now that the PCA has been run, we can plot. The plots can be colored by different aspects to see what is driving the principal components. My plot pca function is derived from the functions script.

```{r plot PCA, warning=FALSE}
if (!dir.exists(
  here(
    "results",
    "plots",
    "pca"
  )
)) {
  dir.create(
    here(
      "results",
      "plots",
      "pca"
    )
  )
}
pdf(
  here(
    "results",
    "plots",
    "pca",
    "pca_plots.pdf"
  )
)
# color by tissue and sex
plot_pca(
  metadata = pca_metadata,
  firstPC = PC1,
  secondPC = PC2,
  color = tissue,
  shape = sex
)

# color by sample
plot_pca(
  metadata = pca_metadata,
  firstPC = PC1,
  secondPC = PC2,
  color = sample_id
)

# color by batch
plot_pca(
  metadata = pca_metadata,
  firstPC = PC1,
  secondPC = PC2,
  color = batch
)

# color by rin score
plot_pca(
  metadata = pca_metadata,
  firstPC = PC1,
  secondPC = PC2,
  color = RIN
)

# color by median length
plot_pca(
  metadata = pca_metadata,
  firstPC = PC1,
  secondPC = PC2,
  color = med_len
)

# color by median quality
plot_pca(
  metadata = pca_metadata,
  firstPC = PC1,
  secondPC = PC2,
  color = med_qual
)

# color by total reads
plot_pca(
  metadata = pca_metadata,
  firstPC = PC1,
  secondPC = PC2,
  color = total_reads
)

# color by mouse id
plot_pca(
  metadata = pca_metadata,
  firstPC = PC1,
  secondPC = PC2,
  color = mouse_id
)
dev.off()
```
#### Clean up script

```{r tidy script}
style_file("04_pca_eda.Rmd")
lint("04_pca_eda.Rmd",
  linters = linters_with_defaults(
    object_length_linter = NULL,
    object_name_linter = NULL,
    object_usage_linter = NULL
  )
)
```

Finally, get final processing time.

```{r proc time finish}
fptm <- proc.time() - ptm
fptm[3] / 60
```

This script takes 3 minutes to run.

#### Software versions

```{r versions}
sessionInfo()
```

R version 4.3.0 beta (2023-04-12 r84240)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.2 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8    LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] IsoformSwitchAnalyzeR_1.99.17 pfamAnalyzeR_1.0.0            sva_3.48.0                   
 [4] genefilter_1.82.0             mgcv_1.8-42                   nlme_3.1-162                 
 [7] satuRn_1.8.0                  DEXSeq_1.46.0                 RColorBrewer_1.1-3           
[10] AnnotationDbi_1.62.0          DESeq2_1.40.0                 SummarizedExperiment_1.30.0  
[13] GenomicRanges_1.52.0          GenomeInfoDb_1.36.0           IRanges_2.34.0               
[16] S4Vectors_0.38.0              MatrixGenerics_1.12.0         matrixStats_0.63.0           
[19] Biobase_2.60.0                BiocGenerics_0.46.0           BiocParallel_1.34.0          
[22] limma_3.56.0                  here_1.0.1                    lintr_3.0.2                  
[25] styler_1.9.1                  lubridate_1.9.2               forcats_1.0.0                
[28] stringr_1.5.0                 dplyr_1.1.2                   purrr_1.0.1                  
[31] readr_2.1.4                   tidyr_1.3.0                   tibble_3.2.1                 
[34] ggplot2_3.4.2                 tidyverse_2.0.0              

loaded via a namespace (and not attached):
  [1] splines_4.3.0                 later_1.3.0                   BiocIO_1.10.0                
  [4] bitops_1.0-7                  filelock_1.0.2                R.oo_1.25.0                  
  [7] XML_3.99-0.14                 rex_1.2.1                     lifecycle_1.0.3              
 [10] edgeR_3.42.0                  rprojroot_2.0.3               vroom_1.6.3                  
 [13] ensembldb_2.24.0              processx_3.8.0                lattice_0.21-8               
 [16] magrittr_2.0.3                rmarkdown_2.21                yaml_2.3.7                   
 [19] remotes_2.4.2                 httpuv_1.6.9                  pbapply_1.7-0                
 [22] DBI_1.1.3                     zlibbioc_1.46.0               R.cache_0.16.0               
 [25] R.utils_2.12.2                AnnotationFilter_1.24.0       RCurl_1.98-1.12              
 [28] rappdirs_0.3.3                GenomeInfoDbData_1.2.10       annotate_1.78.0              
 [31] codetools_0.2-19              DelayedArray_0.25.0           xml2_1.3.4                   
 [34] tidyselect_1.2.0              futile.logger_1.4.3           locfdr_1.1-8                 
 [37] farver_2.1.1                  BiocFileCache_2.8.0           jsonlite_1.8.4               
 [40] GenomicAlignments_1.36.0      ellipsis_0.3.2                survival_3.5-5               
 [43] tools_4.3.0                   progress_1.2.2                Rcpp_1.0.10                  
 [46] glue_1.6.2                    gridExtra_2.3                 xfun_0.39                    
 [49] withr_2.5.0                   formatR_1.14                  BiocManager_1.30.20          
 [52] fastmap_1.1.1                 boot_1.3-28.1                 fansi_1.0.4                  
 [55] callr_3.7.3                   digest_0.6.31                 timechange_0.2.0             
 [58] R6_2.5.1                      mime_0.12                     colorspace_2.1-0             
 [61] biomaRt_2.56.0                RSQLite_2.3.1                 R.methodsS3_1.8.2            
 [64] utf8_1.2.3                    generics_0.1.3                tximeta_1.18.0               
 [67] rtracklayer_1.60.0            prettyunits_1.1.1             httr_1.4.5                   
 [70] pkgconfig_2.0.3               gtable_0.3.3                  blob_1.2.4                   
 [73] hwriter_1.3.2.1               XVector_0.40.0                htmltools_0.5.5              
 [76] geneplotter_1.78.0            ProtGenerics_1.32.0           scales_1.2.1                 
 [79] cyclocomp_1.1.0               png_0.1-8                     knitr_1.42                   
 [82] lambda.r_1.2.4                rstudioapi_0.14               tzdb_0.3.0                   
 [85] reshape2_1.4.4                rjson_0.2.21                  curl_5.0.0                   
 [88] cachem_1.0.8                  BiocVersion_3.17.1            parallel_4.3.0               
 [91] restfulr_0.0.15               desc_1.4.2                    pillar_1.9.0                 
 [94] grid_4.3.0                    vctrs_0.6.2                   promises_1.2.0.1             
 [97] dbplyr_2.3.2                  xtable_1.8-4                  tximport_1.28.0              
[100] evaluate_0.20                 GenomicFeatures_1.52.0        VennDiagram_1.7.3            
[103] cli_3.6.1                     locfit_1.5-9.7                compiler_4.3.0               
[106] futile.options_1.0.1          Rsamtools_2.16.0              rlang_1.1.1                  
[109] crayon_1.5.2                  labeling_0.4.2                ps_1.7.4                     
[112] plyr_1.8.8                    stringi_1.7.12                munsell_0.5.0                
[115] Biostrings_2.68.0             lazyeval_0.2.2                Matrix_1.5-4                 
[118] BSgenome_1.68.0               hms_1.1.3                     bit64_4.0.5                  
[121] KEGGREST_1.40.0               statmod_1.5.0                 shiny_1.7.4                  
[124] interactiveDisplayBase_1.38.0 AnnotationHub_3.8.0           memoise_2.0.1                
[127] bit_4.0.5                     xmlparsedata_1.0.5 
