---
title: 'DTU Analysis: Pairwise Brain Regions'
author: "Emma Jones"
date: "2023-02-23"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Differential Transcript Usage (DTU) Analysis in Mouse Brain lrRNA-Seq Data

For code review, I am tracking the amount of time that this analysis will take.

```{r proc time}
ptm <- proc.time()
```

First, we need to load in all packages.

#### Load in packages

```{r load packages and set seed}
suppressPackageStartupMessages({
  library(tidyverse)
  library(IsoformSwitchAnalyzeR)
  library(ComplexHeatmap)
  library(viridis)
  library(org.Mm.eg.db)
  library(gprofiler2)
  library(styler)
  library(lintr)
  library(here)
})

set.seed(123)
```

#### Load in data

Next, I want to read in data and get cpm values. Since I'll be doing this for all 3 analyses, I made it an R script that can be run beforehand and saves the data to an RData file. I will also source all of the functions I wrote.

```{r read in data and get cpm values}
load(here("data", "cpm_out", "cpm_counts_metadata.RData"))
source("functions.R", local = knitr::knit_global())
```

## Begin DTU analysis for pairwise brain region comparisons

The package requires a design matrix in specific format.

```{r make design matrix}
exp_design <- data.frame(
  sampleID = sample_collection_metadata$sample_id,
  condition = sample_collection_metadata$tissue
)
```

### Run satuRn

Now, I can make a switchlists and run satuRn. Because my data includes counts of less than 1, I still want to include them in my analysis. Therefore, I made the geneExpressionCutoff = NULL. This is a custom function from the functions script.

```{r make switchlist and run saturn}
region_region_switchlist_analyzed <- make_switchlist_saturn(
  isoformCountMatrix = merged_counts_iso,
  isoformRepExpression = cpm_iso,
  designMatrix = exp_design,
  isoformExonAnnoation = here(
    "data", "nextflow", "bambu",
    "extended_annotations.gtf"
  ),
  isoformNtFasta = here("data", "gffread", "isoform_sequences.fa"),
  reduceToSwitchingGenes = FALSE
)

# summarize number of isoforms, switches, and genes
extractSwitchSummary(region_region_switchlist_analyzed)
```
Expected results: 

cerebellum vs cortex	475	331	319	
cerebellum vs hippocampus	175	129	120	
cerebellum vs striatum	548	419	355	
cortex vs hippocampus	41	32	31	
cortex vs striatum	273	182	187	
hippocampus vs striatum	72	67	50	
Combined	1002	819	648	

Next, we want to add gene symbols to the created switchlist object.

```{r add gene symbols}
# get gene symbols and save object with gene symbols
region_region_switchlist_analyzed <- get_gene_symbols(
  switchlist_obj = region_region_switchlist_analyzed
)
```

Then, we want to save the analyzed switchlist for later use.

```{r save switchlist}
saveRDS(
  region_region_switchlist_analyzed,
  here(
    "data", "switchlist_objects", "raw",
    "region_region_switchlist_analyzed.Rds"
  )
)
```

Filter genes for significance.

```{r filter for significance}
region_region_sig_features <- dplyr::filter(
  region_region_switchlist_analyzed$isoformFeatures,
  abs(dIF) > 0.1 & isoform_switch_q_value < 0.05
)
```

Save significant transcripts into results directory.

```{r save sig transcripts}
write.csv(
  region_region_sig_features,
  here(
    "results",
    "dtu_transcripts",
    "region_region_sig_features.csv"
  ),
  row.names = FALSE
)
```

We can also do this at the gene level, to make more direct comparisons with DGE.

```{r collapse to gene level}
region_region_sig_genes <- unique(region_region_sig_features$gene_name)
```

And save this output as well.

```{r save sig genes}
write.csv(
  region_region_sig_genes,
  here(
    "results",
    "dtu_genes",
    "region_region_sig_genes.csv"
  ),
  row.names = FALSE
)
```

### Make UpSet Plots

With UpSet plots, we are interested in specifically how many genes with DTU have overlap. So, I must to get a list of genes for each comparison.

First we have to make data into lists and format it as a combination matrix.

```{r get list of genes for each comparison}
# make vector of comparisons
comparisons <- c(
  "cerebellum_cortex",
  "cerebellum_hippocampus",
  "cerebellum_striatum",
  "cortex_hippocampus",
  "cortex_striatum",
  "hippocampus_striatum"
)

# use function to get list of genes for each comparison
sig_overlap_list <- filter_genes(
  comparisons = comparisons,
  switchlist_obj = region_region_switchlist_analyzed,
  sig_isoform_features = region_region_sig_features
)
# make UpSet plot (ComplexHeatmap)
# create matrix
sig_overlap_list_mat <- ComplexHeatmap::list_to_matrix(sig_overlap_list)

# generate combination matrix
sig_overlap_list_comb_mat <- make_comb_mat(sig_overlap_list_mat)
```

Then, we can create and save the upset plot.

```{r create and save upset plot}
# make UpSet plot (ComplexHeatmap)
pdf(here("results", "plots", "upset_plots", "region_region_upset.pdf"),
  width = 12, height = 6
)
upset_plot <- UpSet(sig_overlap_list_comb_mat,
  comb_order = order(-comb_size(sig_overlap_list_comb_mat)),
  comb_col = viridis(6)[comb_degree(sig_overlap_list_comb_mat)],
  top_annotation = upset_top_annotation(sig_overlap_list_comb_mat,
    add_numbers = TRUE
  ),
  right_annotation = upset_right_annotation(sig_overlap_list_comb_mat,
    add_numbers = TRUE
  )
)
draw(upset_plot)
dev.off()
```

### Volcano Plots

Making volcano plots is a classic way to show the relationship between fold changes and p values. I will make some here with the differential isoform fraction (dIF), instead of the more common differential expression (log2FC) seen in DGE experiments.

```{r volcano plot}
# use function on all comparisons
cerebellum_cortex_volcano <- create_volcano_plot(
  switchlist = region_region_switchlist_analyzed,
  condition1 = "cerebellum",
  condition2 = "cortex"
)

cerebellum_hippocampus_volcano <- create_volcano_plot(
  switchlist = region_region_switchlist_analyzed,
  condition1 = "cerebellum",
  condition2 = "hippocampus"
)

cerebellum_striatum_volcano <- create_volcano_plot(
  switchlist = region_region_switchlist_analyzed,
  condition1 = "cerebellum",
  condition2 = "striatum"
)

cortex_hippocampus_volcano <- create_volcano_plot(
  switchlist = region_region_switchlist_analyzed,
  condition1 = "cortex",
  condition2 = "hippocampus"
)

cortex_striatum_volcano <- create_volcano_plot(
  switchlist = region_region_switchlist_analyzed,
  condition1 = "cortex",
  condition2 = "striatum"
)

hippocampus_striatum_volcano <- create_volcano_plot(
  switchlist = region_region_switchlist_analyzed,
  condition1 = "hippocampus",
  condition2 = "striatum"
)
```

Now, we can save the volcano plots.

```{r save volcano plots}
ggsave(
  here(
    "results",
    "plots",
    "satuRn_volcano",
    "cerebellum_cortex_volcano.png"
  ),
  plot = cerebellum_cortex_volcano
)
ggsave(
  here(
    "results",
    "plots",
    "satuRn_volcano",
    "cerebellum_hippocampus_volcano.png"
  ),
  plot = cerebellum_hippocampus_volcano
)
ggsave(
  here(
    "results",
    "plots",
    "satuRn_volcano",
    "cerebellum_striatum_volcano.png"
  ),
  plot = cerebellum_striatum_volcano
)
ggsave(
  here(
    "results",
    "plots",
    "satuRn_volcano",
    "cortex_hippocampus_volcano.png"
  ),
  plot = cortex_hippocampus_volcano
)
ggsave(
  here(
    "results",
    "plots",
    "satuRn_volcano",
    "cortex_striatum_volcano.png"
  ),
  plot = cortex_striatum_volcano
)
ggsave(
  here(
    "results",
    "plots",
    "satuRn_volcano",
    "hippocampus_striatum_volcano.png"
  ),
  plot = hippocampus_striatum_volcano
)
```

## GO Anlaysis of Gene Sets

The nest step of doing DTU is a GO analysis with my gene sets. I am going to use the package gprofiler2. I guess I want to do GO for every comparison list and then see if they have similar processes.

I need to get the correct number of background genes, so this is all genes that were tested for significance, whether significant or not.

```{r get correct list of background genes}
bg_genes <- unique(region_region_switchlist_analyzed$isoformFeatures$gene_id)
bg_genes <- str_extract(bg_genes, "ENSMUSG...........")
```

Now I can run gprofiler.

```{r gprofiler2}
run_plot_gprofiler(
  gene_list = sig_overlap_list$cerebellum_cortex,
  name = "cerebellum_cortex",
  save_path = here("results", "plots", "gprofiler2")
)

run_plot_gprofiler(
  gene_list = sig_overlap_list$cerebellum_hippocampus,
  name = "cerebellum_hippocampus",
  save_path = here("results", "plots", "gprofiler2")
)

run_plot_gprofiler(
  gene_list = sig_overlap_list$cerebellum_striatum,
  name = "cerebellum_striatum",
  save_path = here("results", "plots", "gprofiler2")
)

run_plot_gprofiler(
  gene_list = sig_overlap_list$cortex_hippocampus,
  name = "cortex_hippocampus",
  save_path = here("results", "plots", "gprofiler2")
)

run_plot_gprofiler(
  gene_list = sig_overlap_list$cortex_striatum,
  name = "cortex_striatum",
  save_path = here("results", "plots", "gprofiler2")
)

run_plot_gprofiler(
  gene_list = sig_overlap_list$hippocampus_striatum,
  name = "hippocampus_striatum",
  save_path = here("results", "plots", "gprofiler2")
)
```

This amount of GO is not really seeming beneficial to me, but maybe could be a supplemental figure.

#### Clean up script

```{r tidy script}
style_file("05_dtu_region_region.Rmd")

lint("05_dtu_region_region.Rmd",
  linters = linters_with_defaults(
    object_length_linter = NULL,
    object_name_linter = NULL,
    object_usage_linter = NULL
  )
)
```

Also, needed to get processing time.

```{r proc time finish}
fptm <- proc.time() - ptm
fptm[3] / 60
```

I was able to run everything in 8 minutes.

#### Software versions

My software versions will be commented below.

```{r versions}
sessionInfo()
```
R version 4.3.0 beta (2023-04-12 r84240)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.2 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8    LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] grid      stats4    stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] here_1.0.1                    lintr_3.0.2                   styler_1.9.1                 
 [4] gprofiler2_0.2.1              org.Mm.eg.db_3.17.0           viridis_0.6.2                
 [7] viridisLite_0.4.1             ComplexHeatmap_2.16.0         IsoformSwitchAnalyzeR_1.99.17
[10] pfamAnalyzeR_1.0.0            sva_3.48.0                    genefilter_1.82.0            
[13] mgcv_1.8-42                   nlme_3.1-162                  satuRn_1.8.0                 
[16] DEXSeq_1.46.0                 RColorBrewer_1.1-3            AnnotationDbi_1.62.0         
[19] DESeq2_1.40.0                 SummarizedExperiment_1.30.0   GenomicRanges_1.52.0         
[22] GenomeInfoDb_1.36.0           IRanges_2.34.0                S4Vectors_0.38.0             
[25] MatrixGenerics_1.12.0         matrixStats_0.63.0            Biobase_2.60.0               
[28] BiocGenerics_0.46.0           BiocParallel_1.34.0           limma_3.56.0                 
[31] lubridate_1.9.2               forcats_1.0.0                 stringr_1.5.0                
[34] dplyr_1.1.2                   purrr_1.0.1                   readr_2.1.4                  
[37] tidyr_1.3.0                   tibble_3.2.1                  ggplot2_3.4.2                
[40] tidyverse_2.0.0              

loaded via a namespace (and not attached):
  [1] splines_4.3.0                 later_1.3.0                   BiocIO_1.10.0                
  [4] bitops_1.0-7                  filelock_1.0.2                R.oo_1.25.0                  
  [7] rex_1.2.1                     XML_3.99-0.14                 lifecycle_1.0.3              
 [10] rprojroot_2.0.3               edgeR_3.42.0                  doParallel_1.0.17            
 [13] MASS_7.3-58.4                 processx_3.8.0                lattice_0.21-8               
 [16] ensembldb_2.24.0              magrittr_2.0.3                plotly_4.10.1                
 [19] rmarkdown_2.21                remotes_2.4.2                 yaml_2.3.7                   
 [22] httpuv_1.6.9                  pbapply_1.7-0                 DBI_1.1.3                    
 [25] zlibbioc_1.46.0               R.cache_0.16.0                R.utils_2.12.2               
 [28] AnnotationFilter_1.24.0       RCurl_1.98-1.12               rappdirs_0.3.3               
 [31] circlize_0.4.15               GenomeInfoDbData_1.2.10       annotate_1.78.0              
 [34] codetools_0.2-19              DelayedArray_0.25.0           xml2_1.3.4                   
 [37] tidyselect_1.2.0              shape_1.4.6                   futile.logger_1.4.3          
 [40] locfdr_1.1-8                  farver_2.1.1                  BiocFileCache_2.8.0          
 [43] GenomicAlignments_1.36.0      jsonlite_1.8.4                GetoptLong_1.0.5             
 [46] ellipsis_0.3.2                survival_3.5-5                iterators_1.0.14             
 [49] systemfonts_1.0.4             foreach_1.5.2                 tools_4.3.0                  
 [52] progress_1.2.2                ragg_1.2.5                    Rcpp_1.0.10                  
 [55] glue_1.6.2                    gridExtra_2.3                 xfun_0.39                    
 [58] withr_2.5.0                   formatR_1.14                  BiocManager_1.30.20          
 [61] fastmap_1.1.1                 boot_1.3-28.1                 fansi_1.0.4                  
 [64] callr_3.7.3                   digest_0.6.31                 timechange_0.2.0             
 [67] R6_2.5.1                      mime_0.12                     textshaping_0.3.6            
 [70] colorspace_2.1-0              biomaRt_2.56.0                RSQLite_2.3.1                
 [73] R.methodsS3_1.8.2             utf8_1.2.3                    generics_0.1.3               
 [76] data.table_1.14.8             tximeta_1.18.0                rtracklayer_1.60.0           
 [79] htmlwidgets_1.6.2             prettyunits_1.1.1             httr_1.4.5                   
 [82] pkgconfig_2.0.3               gtable_0.3.3                  blob_1.2.4                   
 [85] hwriter_1.3.2.1               XVector_0.40.0                htmltools_0.5.5              
 [88] geneplotter_1.78.0            ProtGenerics_1.32.0           clue_0.3-64                  
 [91] scales_1.2.1                  cyclocomp_1.1.0               png_0.1-8                    
 [94] knitr_1.42                    lambda.r_1.2.4                rstudioapi_0.14              
 [97] tzdb_0.3.0                    reshape2_1.4.4                rjson_0.2.21                 
[100] curl_5.0.0                    cachem_1.0.8                  GlobalOptions_0.1.2          
[103] BiocVersion_3.17.1            parallel_4.3.0                desc_1.4.2                   
[106] restfulr_0.0.15               pillar_1.9.0                  vctrs_0.6.2                  
[109] promises_1.2.0.1              dbplyr_2.3.2                  xtable_1.8-4                 
[112] cluster_2.1.4                 tximport_1.28.0               evaluate_0.20                
[115] VennDiagram_1.7.3             GenomicFeatures_1.52.0        cli_3.6.1                    
[118] locfit_1.5-9.7                compiler_4.3.0                futile.options_1.0.1         
[121] Rsamtools_2.16.0              rlang_1.1.1                   crayon_1.5.2                 
[124] labeling_0.4.2                ps_1.7.4                      plyr_1.8.8                   
[127] stringi_1.7.12                munsell_0.5.0                 Biostrings_2.68.0            
[130] lazyeval_0.2.2                Matrix_1.5-4                  BSgenome_1.68.0              
[133] hms_1.1.3                     bit64_4.0.5                   KEGGREST_1.40.0              
[136] statmod_1.5.0                 shiny_1.7.4                   interactiveDisplayBase_1.38.0
[139] AnnotationHub_3.8.0           memoise_2.0.1                 bit_4.0.5                    
[142] xmlparsedata_1.0.5
