---
title: "DTU Analysis: Brain Region vs All"
author: "Emma Jones"
date: "2023-02-24"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Differential Transcript Usage (DTU) Analysis by Region in Mouse Brain lrRNA-Seq Data

For code review, I am tracking the amount of time that this analysis will take.

```{r proc time}
ptm <- proc.time()
```


#### Load in packages

```{r load packages}
suppressPackageStartupMessages({
  library(tidyverse)
  library(IsoformSwitchAnalyzeR)
  library(ComplexHeatmap)
  library(viridis)
  library(org.Mm.eg.db)
  library(styler)
  library(lintr)
  library(here)
})

set.seed(123)
```

#### Load in data

Next, I want to read in data and get cpm values. Since I'll be doing this for all 3 analyses, I made it an R script that can be sourced.

```{r read in data and get cpm values}
load(here("data", "cpm_out", "cpm_counts_metadata.RData"))
source("functions.R", local = knitr::knit_global())
```

## DTU for one region versus all other regions

First, I need to update my design matrix because I need to make pairwise comparisons for each region versus all other regions. Then, I can make a new switchAnalyzeRlist object. This is required to run an isoform switch test with DEXSeq. I've made this into a function to make it faster.

Once switchAnalyzeRlists are made, you can run the satuRn isoform switch test. You also need to prefilter the data before running the test, but I added all that to the my function so all it needs is to run.

```{r make design matrix}
tissues <- c(
  "cerebellum",
  "cortex",
  "hippocampus",
  "striatum"
)
designs <- list()
for (tissue in tissues) {
  temp_design <- data.frame(
    sampleID = sample_collection_metadata$sample_id,
    condition = sample_collection_metadata$tissue == tissue
  )
  
  temp_design["condition"][temp_design["condition"] == TRUE] <- tissue
  temp_design["condition"][temp_design["condition"] == FALSE] <- "other"
  designs[[tissue]] <- temp_design
}
```

Make a switchlist, which is an object required for running isoformswitchanalyzer.

Now, I can run satuRn. Because my data includes counts of less than 1, I still want to include them in my analysis. Therefore, I made the geneExpressionCutoff = NULL.

```{r add fasta of transcripts to object}
region_all_switchlist_list <- list()
for(tissue in tissues) {
  switchlist_obj <- make_switchlist_saturn(
    isoformCountMatrix = merged_counts_iso,
    isoformRepExpression = cpm_iso,
    designMatrix = designs[[tissue]],
    isoformExonAnnoation = here(
    "data", "nextflow", "bambu",
    "extended_annotations.gtf"
    ),
    isoformNtFasta = here("data", "gffread", "isoform_sequences.fa"),
    reduceToSwitchingGenes = FALSE
  )
  region_all_switchlist_list[[tissue]] <- switchlist_obj
}

# sumarize number of isoforms, switches, and genes
extractSwitchSummary(region_all_switchlist_list$cerebellum)
extractSwitchSummary(region_all_switchlist_list$cortex)
extractSwitchSummary(region_all_switchlist_list$hippocampus)
extractSwitchSummary(region_all_switchlist_list$striatum)
```

```{r add gene symbols}
# get gene symbols and save object with gene symbols
for(tissue in tissues) {
  temp_switchlist <- get_gene_symbols(
    switchlist_obj = region_all_switchlist_list[[tissue]]
  )
  region_all_switchlist_list[[tissue]] <- temp_switchlist
}
```

```{r save switchlist}
saveRDS(
  region_all_switchlist_list, 
  here("data", "switchlist_objects", "raw", "region_all_switchlist_list.Rds")
)
```


```{r filter for significance}
region_all_sig_features_list <- list()
for(tissue in tissues) {
  temp_sig_features <- dplyr::filter(
    region_all_switchlist_list[[tissue]]$isoformFeatures,
    abs(dIF) > 0.1 & isoform_switch_q_value < 0.05
  )
  region_all_sig_features_list[[tissue]] <- temp_sig_features
}
```

```{r save sig features}
for(i in 1:length(region_all_sig_features_list)) {
  tissue_name <- names(region_all_sig_features_list)[i]
  write.csv(
    region_all_sig_features_list[[tissue_name]], 
    here(
      "results",
      "dtu_transcripts",
      paste0(
        tissue_name,
        "_other_sig_features.csv"
      )
    ), 
    row.names=FALSE
  )
}
```

```{r collapse to gene level}
region_all_sig_genes_list <- list()
for(i in 1:length(region_all_sig_features_list)) {
  tissue_name <- names(region_all_sig_features_list)[i]
  tissue_sig_genes <- unique(region_all_sig_features_list[[i]]$gene_name)
  region_all_sig_genes_list[[tissue_name]] <- tissue_sig_genes
}
```

```{r save sig genes}
for(i in 1:length(region_all_sig_genes_list)) {
  tissue_name <- names(region_all_sig_genes_list)[i]
  write.csv(
    region_all_sig_genes_list[[tissue_name]], 
    here(
      "results",
      "dtu_genes",
      paste0(
        tissue_name,
        "_other_sig_genes.csv"
      )
    ), 
    row.names=FALSE
  )
}
```

These switchlists are fine, but I'm going to apply a function for adding gene names to make downstream analysis better and more interpretable.

### Make plots for tissue:all DTU

Now, with the list created I can create a ComplexHeatmap UpSet plot.

#### Make UpSet plots

```{r complexheatmap tissue:all UpSet genes}
for(tissue in tissues) {
  region_all_sig_genes_list[[tissue]] <- na.omit(region_all_sig_genes_list[[tissue]])
}
# make UpSet plot (ComplexHeatmap)
# create matrix
all_gene_list_mat <- list_to_matrix(region_all_sig_genes_list)

# generate combination matrix
all_gene_list_comb_mat <- make_comb_mat(all_gene_list_mat)

# make UpSet plot (ComplexHeatmap)
UpSet(all_gene_list_comb_mat,
      pt_size = unit(3, "mm"),
      comb_col = viridis(4)[comb_degree(all_gene_list_comb_mat)]
)
pdf(here("results", "plots", "upset_plots", "region_others_gene_upset.pdf"))
upset_gene <- UpSet(all_gene_list_comb_mat,
                    comb_order = order(-comb_size(all_gene_list_comb_mat)),
                    comb_col = viridis(4)[comb_degree(all_gene_list_comb_mat)],
                    top_annotation = upset_top_annotation(
                      all_gene_list_comb_mat,
                      add_numbers = TRUE
                    ),
                    right_annotation = upset_right_annotation(
                      all_gene_list_comb_mat,
                      add_numbers = TRUE
                    )
)

draw(upset_gene, column_title = "Overlap of DTU genes by brain region")
dev.off()
```

Cerebellum has most unique genes overall, but what happens when we extend this to isoforms?

```{r extract sigificant tissue:all isoforms}
unique_sig_isoform_list <- list()
for(tissue in tissues) {
  unique_sig_isoform_list[[tissue]] <- unique(region_all_sig_features_list[[tissue]]$isoform_id)
}

# make UpSet plot (ComplexHeatmap)
# create matrix
all_isoform_list_mat <- list_to_matrix(unique_sig_isoform_list)

# generate combination matrix
all_isoform_list_comb_mat <- make_comb_mat(all_isoform_list_mat)

# make UpSet plot (ComplexHeatmap)
pdf(here("results", "plots", "upset_plots", "region_others_isoform_upset.pdf"))
upset_isoform <- UpSet(all_isoform_list_comb_mat,
                    comb_order = order(-comb_size(all_isoform_list_comb_mat)),
                    comb_col = viridis(4)[comb_degree(all_isoform_list_comb_mat)],
                    top_annotation = upset_top_annotation(
                      all_isoform_list_comb_mat,
                      add_numbers = TRUE
                    ),
                    right_annotation = upset_right_annotation(
                      all_isoform_list_comb_mat,
                      add_numbers = TRUE
                    )
)

draw(upset_isoform, column_title = "Overlap of DTU isoforms by brain region")
dev.off()
```

#### Make volcano plots

I also want to make volcano plots to visualize this.

```{r make volcano plots for region v others}
cerebellum_all_volcano <- create_volcano_plot(
  switchlist = region_all_switchlist_list$cerebellum,
  condition1 = "cerebellum",
  condition2 = "other"
)

cortex_all_volcano <- create_volcano_plot(
  switchlist = region_all_switchlist_list$cortex,
  condition1 = "cortex",
  condition2 = "other"
)

hippocampus_all_volcano <- create_volcano_plot(
  switchlist = region_all_switchlist_list$hippocampus,
  condition1 = "hippocampus",
  condition2 = "other"
)

striatum_all_volcano <- create_volcano_plot(
  switchlist = region_all_switchlist_list$striatum,
  condition1 = "striatum",
  condition2 = "other"
)
```

```{r save volcano plots}
ggsave(
  here(
    "results",
    "plots",
    "satuRn_volcano",
    "cerebellum_all_volcano.png"
  ),
  plot = cerebellum_all_volcano
)

ggsave(
  here(
    "results",
    "plots",
    "satuRn_volcano",
    "cortex_all_volcano.png"
  ),
  plot = cortex_all_volcano
)

ggsave(
  here(
    "results",
    "plots",
    "satuRn_volcano",
    "hippocampus_all_volcano.png"
  ),
  plot = hippocampus_all_volcano
)

ggsave(
  here(
    "results",
    "plots",
    "satuRn_volcano",
    "striatum_all_volcano.png"
  ),
  plot = striatum_all_volcano
)
```

#### Clean up script

```{r tidy script}
style_file("06_dtu_region_others.Rmd")
lint("06_dtu_region_others.Rmd")
```

Get processing time.

```{r proc time finish}
fptm <- proc.time() - ptm
fptm[3] / 60
```

Estimated processing time was 9 minutes. In a slow session, this went up to 41 minutes but I think that is an anomaly.

#### Software versions

```{r versions}
sessionInfo()
```
R version 4.3.0 beta (2023-04-12 r84240)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.2 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
 [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] grid      stats4    stats     graphics  grDevices utils     datasets  methods  
[9] base     

other attached packages:
 [1] here_1.0.1                    lintr_3.0.2                  
 [3] styler_1.9.1                  org.Mm.eg.db_3.17.0          
 [5] viridis_0.6.2                 viridisLite_0.4.1            
 [7] ComplexHeatmap_2.16.0         IsoformSwitchAnalyzeR_1.99.17
 [9] pfamAnalyzeR_1.0.0            sva_3.48.0                   
[11] genefilter_1.82.0             mgcv_1.8-42                  
[13] nlme_3.1-162                  satuRn_1.8.0                 
[15] DEXSeq_1.46.0                 RColorBrewer_1.1-3           
[17] AnnotationDbi_1.62.0          DESeq2_1.40.0                
[19] SummarizedExperiment_1.30.0   GenomicRanges_1.52.0         
[21] GenomeInfoDb_1.36.0           IRanges_2.34.0               
[23] S4Vectors_0.38.0              MatrixGenerics_1.12.0        
[25] matrixStats_0.63.0            Biobase_2.60.0               
[27] BiocGenerics_0.46.0           BiocParallel_1.34.0          
[29] limma_3.56.0                  lubridate_1.9.2              
[31] forcats_1.0.0                 stringr_1.5.0                
[33] dplyr_1.1.2                   purrr_1.0.1                  
[35] readr_2.1.4                   tidyr_1.3.0                  
[37] tibble_3.2.1                  ggplot2_3.4.2                
[39] tidyverse_2.0.0              

loaded via a namespace (and not attached):
  [1] splines_4.3.0                 later_1.3.0                  
  [3] BiocIO_1.10.0                 bitops_1.0-7                 
  [5] filelock_1.0.2                R.oo_1.25.0                  
  [7] rex_1.2.1                     XML_3.99-0.14                
  [9] lifecycle_1.0.3               rprojroot_2.0.3              
 [11] edgeR_3.42.0                  doParallel_1.0.17            
 [13] MASS_7.3-58.4                 processx_3.8.0               
 [15] lattice_0.21-8                ensembldb_2.24.0             
 [17] magrittr_2.0.3                rmarkdown_2.21               
 [19] remotes_2.4.2                 yaml_2.3.7                   
 [21] httpuv_1.6.9                  pbapply_1.7-0                
 [23] DBI_1.1.3                     zlibbioc_1.46.0              
 [25] R.cache_0.16.0                R.utils_2.12.2               
 [27] AnnotationFilter_1.24.0       RCurl_1.98-1.12              
 [29] rappdirs_0.3.3                circlize_0.4.15              
 [31] GenomeInfoDbData_1.2.10       annotate_1.78.0              
 [33] codetools_0.2-19              DelayedArray_0.25.0          
 [35] xml2_1.3.4                    tidyselect_1.2.0             
 [37] shape_1.4.6                   futile.logger_1.4.3          
 [39] locfdr_1.1-8                  farver_2.1.1                 
 [41] BiocFileCache_2.8.0           GenomicAlignments_1.36.0     
 [43] jsonlite_1.8.4                GetoptLong_1.0.5             
 [45] ellipsis_0.3.2                survival_3.5-5               
 [47] iterators_1.0.14              systemfonts_1.0.4            
 [49] foreach_1.5.2                 tools_4.3.0                  
 [51] progress_1.2.2                ragg_1.2.5                   
 [53] Rcpp_1.0.10                   glue_1.6.2                   
 [55] gridExtra_2.3                 xfun_0.39                    
 [57] withr_2.5.0                   formatR_1.14                 
 [59] BiocManager_1.30.20           fastmap_1.1.1                
 [61] boot_1.3-28.1                 fansi_1.0.4                  
 [63] callr_3.7.3                   digest_0.6.31                
 [65] timechange_0.2.0              R6_2.5.1                     
 [67] mime_0.12                     textshaping_0.3.6            
 [69] colorspace_2.1-0              biomaRt_2.56.0               
 [71] RSQLite_2.3.1                 R.methodsS3_1.8.2            
 [73] utf8_1.2.3                    generics_0.1.3               
 [75] tximeta_1.18.0                rtracklayer_1.60.0           
 [77] prettyunits_1.1.1             httr_1.4.5                   
 [79] pkgconfig_2.0.3               gtable_0.3.3                 
 [81] blob_1.2.4                    hwriter_1.3.2.1              
 [83] XVector_0.40.0                htmltools_0.5.5              
 [85] geneplotter_1.78.0            ProtGenerics_1.32.0          
 [87] clue_0.3-64                   scales_1.2.1                 
 [89] cyclocomp_1.1.0               png_0.1-8                    
 [91] knitr_1.42                    lambda.r_1.2.4               
 [93] rstudioapi_0.14               tzdb_0.3.0                   
 [95] reshape2_1.4.4                rjson_0.2.21                 
 [97] curl_5.0.0                    cachem_1.0.8                 
 [99] GlobalOptions_0.1.2           BiocVersion_3.17.1           
[101] parallel_4.3.0                desc_1.4.2                   
[103] restfulr_0.0.15               pillar_1.9.0                 
[105] vctrs_0.6.2                   promises_1.2.0.1             
[107] dbplyr_2.3.2                  xtable_1.8-4                 
[109] cluster_2.1.4                 tximport_1.28.0              
[111] evaluate_0.20                 VennDiagram_1.7.3            
[113] GenomicFeatures_1.52.0        cli_3.6.1                    
[115] locfit_1.5-9.7                compiler_4.3.0               
[117] futile.options_1.0.1          Rsamtools_2.16.0             
[119] rlang_1.1.1                   crayon_1.5.2                 
[121] labeling_0.4.2                ps_1.7.4                     
[123] plyr_1.8.8                    stringi_1.7.12               
[125] munsell_0.5.0                 Biostrings_2.68.0            
[127] lazyeval_0.2.2                Matrix_1.5-4                 
[129] BSgenome_1.68.0               hms_1.1.3                    
[131] bit64_4.0.5                   KEGGREST_1.40.0              
[133] statmod_1.5.0                 shiny_1.7.4                  
[135] interactiveDisplayBase_1.38.0 AnnotationHub_3.8.0          
[137] memoise_2.0.1                 bit_4.0.5                    
[139] xmlparsedata_1.0.5  
