---
title: 'DTU Analysis: Brain Region and Sex'
author: "Emma Jones"
date: "2023-02-27"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Differential Transcript Usage (DTU) Analysis across Sex in Mouse Brain lrRNA-Seq Data

For code review, I am tracking the amount of time that this analysis will take.

```{r proc time}
ptm <- proc.time()
```


#### Load in packages

```{r load packages}
suppressPackageStartupMessages({
  library(tidyverse)
  library(IsoformSwitchAnalyzeR)
  library(ComplexHeatmap)
  library(viridis)
  library(org.Mm.eg.db)
  library(gprofiler2)
  library(styler)
  library(lintr)
  library(here)
})

set.seed(123)
```

#### Load in data

Next, I want to read in data and get cpm values. Since I'll be doing this for all 3 analyses, I made it an R script and load in the output.

```{r read in data and get cpm values}
load(here("data", "cpm_out", "cpm_counts_metadata.RData"))
source("functions.R", local = knitr::knit_global())
```

### Get DTU by sex

Now that I have shown DTU differences across tissue, I am going to look at differences across sex. I'll need to create a new switchAnalyzeRlist object with metadata for sex. In this chunk we are making design matrices for all brain regions combined and separated.

```{r make design matrices}
# make sex design for combined regions
sex_design_combined <- data.frame(
  sampleID = sample_collection_metadata$sample_id,
  condition = sample_collection_metadata$sex
)


# set the design experiment for sex for each region
tissues <- c(
  "cerebellum",
  "cortex",
  "hippocampus",
  "striatum"
)
designs <- list()
# create designs list
for (tissue in tissues) {
  temp_design <- data.frame(
    sampleID = sample_collection_metadata$sample_id[
      sample_collection_metadata$tissue == tissue
    ],
    condition = sample_collection_metadata$sex[
      sample_collection_metadata$tissue == tissue
    ]
  )
  designs[[tissue]] <- temp_design
}
```

First, we will make a switchlist for all tissues combined, comparing grossly across sex. Because there are many replicates, we can use satuRn.

```{r make switchlist and run saturn for sex combined}
# make switchlist
sex_switchlist <- make_switchlist_saturn(
  isoformCountMatrix = merged_counts_iso,
  isoformRepExpression = cpm_iso,
  designMatrix = sex_design_combined,
  isoformExonAnnoation = here(
    "data", "nextflow", "bambu",
    "extended_annotations.gtf"
  ),
  isoformNtFasta = here("data", "gffread", "isoform_sequences.fa"),
  reduceToSwitchingGenes = FALSE
)

extractSwitchSummary(sex_switchlist)
```

Expected results: F vs M	5	5	4	

Only 4 genes were considered switching, so I will also try splitting by brain region.

One of the genes Atp6v0e2, had downregulation in females with AD in this study https://www.frontiersin.org/articles/10.3389/fnagi.2021.735611/full

### DTU in males and females split by brain region

Now, we can also split out switchlists for each region. This creates smaller numbers of replicates so we want to run dexseq instead of satuRn.

```{r make switchlists and run dexseq for sex by region}
counts_list <- list()
for (tissue in tissues) {
  subset_counts <- subset(
    merged_counts_iso,
    select = c(
      "isoform_id",
      sample_collection_metadata$sample_id[
        sample_collection_metadata$tissue == tissue
      ]
    )
  )
  counts_list[[tissue]] <- subset_counts
}
cpm_list <- list()
for (tissue in tissues) {
  # subset cpm
  subset_cpm <- subset(
    cpm_iso,
    select = c(
      "isoform_id",
      sample_collection_metadata$sample_id[
        sample_collection_metadata$tissue == tissue
      ]
    )
  )
  # drop rows with 0 cpm
  subset_cpm <- subset_cpm[rowSums(subset_cpm[, -1]) != 0, ]
  cpm_list[[tissue]] <- subset_cpm
}
region_sex_switchlist_list <- list()
for (tissue in tissues) {
  tryCatch(
    {
      switchlist_obj <- make_switchlist_dexseq(
        isoformCountMatrix = counts_list[[tissue]],
        isoformRepExpression = cpm_list[[tissue]],
        designMatrix = designs[[tissue]],
        isoformExonAnnoation = here(
          "data", "nextflow", "bambu",
          "extended_annotations.gtf"
        ),
        isoformNtFasta = here("data", "gffread", "isoform_sequences.fa"),
        reduceToSwitchingGenes = FALSE
      )
      region_sex_switchlist_list[[tissue]] <- switchlist_obj
    },
    error = function(e) {
      cat("ERROR: ", tissue, conditionMessage(e), "\n")
    }
  )
}

extractSwitchSummary(region_sex_switchlist_list[["cerebellum"]])
extractSwitchSummary(region_sex_switchlist_list[["cortex"]])
extractSwitchSummary(region_sex_switchlist_list[["striatum"]])
```

expected cerebellum results: F vs M	4	3	2	

expected cortex results: F vs M	22	19	14	

expected striatum results: F vs M	11	7	7	

Now, we want to save our data.

```{r save sex combined switchlist}
saveRDS(
  sex_switchlist,
  here("data", "switchlist_objects", "raw", "sex_switchlist.Rds")
)
```

Also save the brain-region-specific data.

```{r save sex by region switchlists}
saveRDS(
  region_sex_switchlist_list,
  here("data", "switchlist_objects", "raw", "region_sex_switchlist_list.Rds")
)
```

Next, we filter these same genes for significance.

```{r filter for significance}
# sex combined
sex_switchlist_sig_features <- dplyr::filter(
  sex_switchlist$isoformFeatures,
  abs(dIF) > 0.1 & isoform_switch_q_value < 0.05
)
# sex_region
region_sex_sig_features_list <- list()
for (tissue in tissues) {
  tryCatch(
    {
      temp_sig_features <- dplyr::filter(
        region_sex_switchlist_list[[tissue]]$isoformFeatures,
        abs(dIF) > 0.1 & isoform_switch_q_value < 0.05
      )
      region_sex_sig_features_list[[tissue]] <- temp_sig_features
    },
    error = function(e) {
      cat("ERROR: ", tissue, conditionMessage(e), "\n")
    }
  )
}
```

Save these features in results.

```{r save sex combined sig features}
write.csv(
  sex_switchlist_sig_features,
  here(
    "results",
    "dtu_transcripts",
    "sex_switchlist_sig_features.csv"
  ),
  row.names = FALSE
)
```

Do the same for tissue-specific sex switchlists.

```{r save sex region sig features}
for (i in seq_along(region_sex_sig_features_list)) {
  tissue_name <- names(region_sex_sig_features_list)[i]
  write.csv(
    region_sex_sig_features_list[[tissue_name]],
    here(
      "results",
      "dtu_transcripts",
      paste0(
        tissue_name,
        "_sex_sig_features.csv"
      )
    ),
    row.names = FALSE
  )
}
```

Now, we can collapse these counts to the gene level.

```{r collapse to gene level}
# sex combined
sex_switchlist_sig_genes <- unique(sex_switchlist_sig_features$gene_name)

# sex_region
region_sex_sig_genes_list <- list()
for (i in seq_along(region_sex_sig_features_list)) {
  tryCatch(
    {
      tissue_name <- names(region_sex_sig_features_list)[i]
      tissue_sig_genes <- unique(region_sex_sig_features_list[[i]]$gene_name)
      region_sex_sig_genes_list[[tissue_name]] <- tissue_sig_genes
    },
    error = function(e) {
      cat("ERROR: ", tissue, conditionMessage(e), "\n")
    }
  )
}
```

Save the combined genes only into results.

```{r save sig genes sex combined}
write.csv(
  sex_switchlist_sig_genes,
  here(
    "results",
    "dtu_genes",
    "sex_switchlist_sig_genes.csv"
  ),
  row.names = FALSE
)
```

Also save the genes for each region.

```{r save sig genes sex regions}
for (i in seq_along(region_sex_sig_genes_list)) {
  tissue_name <- names(region_sex_sig_genes_list)[i]
  write.csv(
    region_sex_sig_genes_list[[tissue_name]],
    here(
      "results",
      "dtu_genes",
      paste0(
        tissue_name,
        "_sex_sig_genes.csv"
      )
    ),
    row.names = FALSE
  )
}
```

Alright, now I have some differential used isoforms/genes saved. Let's visualize them with a volcano plot, although with how few genes there are it feels silly.

#### Make volcano plots

```{r volcano plots for sex split by brain region}
# use function
create_sex_volcano_plot(
  tissue_obj = stri_sex_switchlist_analyzed,
  save_path = here("results", "plots", "satuRn_volcano")
)
```

I removed the UpSetPlot code because there were only 3 genes considered switching in a single brain region (striatum).

#### GO Analysis using gprofiler2

Before doing GO analysis, I need to pull out significant genes. GO analysis does not currently work at the isoform level (at least for gprofiler2).

```{r pull out signficant genes}
get_cut_sig_genes(tissue = "striatum")
```

Now, perform go analysis with gprofiler2, but this gives no results.

```{r go analysis}
# run gprofiler function on striatum
run_gprofiler(tissue = "striatum") # no results to show
```

#### Clean up script

```{r tidy script}
style_file("07_dtu_region_sex.Rmd")
lint("07_dtu_region_sex.Rmd", linters = linters_with_defaults(
  object_length_linter = NULL,
  object_name_linter = NULL,
  object_usage_linter = NULL
))
```

Get processing time.

```{r proc time finish}
fptm <- proc.time() - ptm
fptm[3] / 60
```

Estimated processing time is 22 minutes.

#### Software versions

```{r versions}
sessionInfo()
```
R version 4.3.0 beta (2023-04-12 r84240)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.2 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
 [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] grid      stats4    stats     graphics  grDevices utils     datasets  methods  
[9] base     

other attached packages:
 [1] here_1.0.1                    lintr_3.0.2                  
 [3] styler_1.9.1                  gprofiler2_0.2.1             
 [5] org.Mm.eg.db_3.17.0           viridis_0.6.2                
 [7] viridisLite_0.4.2             ComplexHeatmap_2.16.0        
 [9] IsoformSwitchAnalyzeR_1.99.17 pfamAnalyzeR_1.0.0           
[11] sva_3.48.0                    genefilter_1.82.0            
[13] mgcv_1.8-42                   nlme_3.1-162                 
[15] satuRn_1.8.0                  DEXSeq_1.46.0                
[17] RColorBrewer_1.1-3            AnnotationDbi_1.62.0         
[19] DESeq2_1.40.0                 SummarizedExperiment_1.30.0  
[21] GenomicRanges_1.52.0          GenomeInfoDb_1.36.0          
[23] IRanges_2.34.0                S4Vectors_0.38.0             
[25] MatrixGenerics_1.12.0         matrixStats_0.63.0           
[27] Biobase_2.60.0                BiocGenerics_0.46.0          
[29] BiocParallel_1.34.0           limma_3.56.0                 
[31] lubridate_1.9.2               forcats_1.0.0                
[33] stringr_1.5.0                 dplyr_1.1.2                  
[35] purrr_1.0.1                   readr_2.1.4                  
[37] tidyr_1.3.0                   tibble_3.2.1                 
[39] ggplot2_3.4.2                 tidyverse_2.0.0              

loaded via a namespace (and not attached):
  [1] splines_4.3.0                 later_1.3.0                  
  [3] BiocIO_1.10.0                 bitops_1.0-7                 
  [5] filelock_1.0.2                R.oo_1.25.0                  
  [7] rex_1.2.1                     XML_3.99-0.14                
  [9] lifecycle_1.0.3               rprojroot_2.0.3              
 [11] edgeR_3.42.0                  doParallel_1.0.17            
 [13] MASS_7.3-58.4                 processx_3.8.0               
 [15] lattice_0.21-8                ensembldb_2.24.0             
 [17] magrittr_2.0.3                plotly_4.10.1                
 [19] rmarkdown_2.21                remotes_2.4.2                
 [21] yaml_2.3.7                    httpuv_1.6.9                 
 [23] pbapply_1.7-0                 DBI_1.1.3                    
 [25] zlibbioc_1.46.0               R.cache_0.16.0               
 [27] R.utils_2.12.2                AnnotationFilter_1.24.0      
 [29] RCurl_1.98-1.12               rappdirs_0.3.3               
 [31] circlize_0.4.15               GenomeInfoDbData_1.2.10      
 [33] annotate_1.78.0               codetools_0.2-19             
 [35] DelayedArray_0.25.0           xml2_1.3.4                   
 [37] tidyselect_1.2.0              shape_1.4.6                  
 [39] futile.logger_1.4.3           locfdr_1.1-8                 
 [41] BiocFileCache_2.8.0           GenomicAlignments_1.36.0     
 [43] jsonlite_1.8.4                GetoptLong_1.0.5             
 [45] ellipsis_0.3.2                survival_3.5-5               
 [47] iterators_1.0.14              foreach_1.5.2                
 [49] tools_4.3.0                   progress_1.2.2               
 [51] Rcpp_1.0.10                   glue_1.6.2                   
 [53] gridExtra_2.3                 xfun_0.39                    
 [55] withr_2.5.0                   formatR_1.14                 
 [57] BiocManager_1.30.20           fastmap_1.1.1                
 [59] boot_1.3-28.1                 fansi_1.0.4                  
 [61] callr_3.7.3                   digest_0.6.31                
 [63] timechange_0.2.0              R6_2.5.1                     
 [65] mime_0.12                     colorspace_2.1-0             
 [67] biomaRt_2.56.0                RSQLite_2.3.1                
 [69] R.methodsS3_1.8.2             utf8_1.2.3                   
 [71] generics_0.1.3                data.table_1.14.8            
 [73] tximeta_1.18.0                rtracklayer_1.60.0           
 [75] htmlwidgets_1.6.2             prettyunits_1.1.1            
 [77] httr_1.4.5                    pkgconfig_2.0.3              
 [79] gtable_0.3.3                  blob_1.2.4                   
 [81] hwriter_1.3.2.1               XVector_0.40.0               
 [83] htmltools_0.5.5               geneplotter_1.78.0           
 [85] ProtGenerics_1.32.0           clue_0.3-64                  
 [87] scales_1.2.1                  cyclocomp_1.1.0              
 [89] png_0.1-8                     knitr_1.42                   
 [91] lambda.r_1.2.4                rstudioapi_0.14              
 [93] tzdb_0.3.0                    reshape2_1.4.4               
 [95] rjson_0.2.21                  curl_5.0.0                   
 [97] cachem_1.0.8                  GlobalOptions_0.1.2          
 [99] BiocVersion_3.17.1            parallel_4.3.0               
[101] desc_1.4.2                    restfulr_0.0.15              
[103] pillar_1.9.0                  vctrs_0.6.3                  
[105] promises_1.2.0.1              dbplyr_2.3.2                 
[107] xtable_1.8-4                  cluster_2.1.4                
[109] tximport_1.28.0               evaluate_0.20                
[111] VennDiagram_1.7.3             GenomicFeatures_1.52.0       
[113] cli_3.6.1                     locfit_1.5-9.7               
[115] compiler_4.3.0                futile.options_1.0.1         
[117] Rsamtools_2.16.0              rlang_1.1.1                  
[119] crayon_1.5.2                  ps_1.7.4                     
[121] plyr_1.8.8                    stringi_1.7.12               
[123] munsell_0.5.0                 Biostrings_2.68.0            
[125] lazyeval_0.2.2                Matrix_1.5-4                 
[127] BSgenome_1.68.0               hms_1.1.3                    
[129] bit64_4.0.5                   KEGGREST_1.40.0              
[131] statmod_1.5.0                 shiny_1.7.4                  
[133] interactiveDisplayBase_1.38.0 AnnotationHub_3.8.0          
[135] memoise_2.0.1                 bit_4.0.5                    
[137] xmlparsedata_1.0.5    
