---
title: 'DTU Analysis: Brain Region and Sex'
author: "Emma Jones"
date: "2023-02-27"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Differential Transcript Usage (DTU) Analysis across Sex in Mouse Brain lrRNA-Seq Data

For code review, I am tracking the amount of time that this analysis will take.

```{r proc time}
ptm <- proc.time()
```


#### Load in packages

```{r load packages}
suppressPackageStartupMessages({library(tidyverse)
library(IsoformSwitchAnalyzeR)
library(ComplexHeatmap)
library(viridis)
library(org.Mm.eg.db)
library(gprofiler2)
library(styler)
library(lintr)
library(here)})

set.seed(123)
```

#### Load in data

Next, I want to read in data and get cpm values. Since I'll be doing this for all 3 analyses, I made it an R script and load in the output.

```{r read in data and get cpm values}
load(here("data", "cpm_out", "cpm_counts_metadata.RData"))
source("functions.R", local = knitr::knit_global())
```

### Get DTU by sex

Now that I have shown DTU differences across tissue, I am going to look at differences across sex. I'll need to create a new switchAnalyzeRlist object with metadata for sex.

```{r create metadata and create object for sex}
# make sex design
sex_design <- data.frame(
  sampleID = sample_collection_metadata$sample_id,
  condition = sample_collection_metadata$sex
)
# make switchlist
sex_switchlist <- importRdata(
  isoformCountMatrix = merged_counts_iso,
  isoformRepExpression = cpm_iso,
  designMatrix = sex_design,
  isoformExonAnnoation = here(
    "data", "nextflow", "bambu", "extended_annotations.gtf"
  ),
  isoformNtFasta = here("data", "gffread", "isoform_sequences.fa"),
  showProgress = FALSE
)
```

Now that I have my object, I can run a DTU analysis.

```{r run satuRn}
# filter data
sex_switchlist <- preFilter(sex_switchlist, geneExpressionCutoff = NULL)
# run satuRn
sex_switchlist_analyzed <- isoformSwitchTestSatuRn(
  switchAnalyzeRlist = sex_switchlist,
  reduceToSwitchingGenes = TRUE
)
extractSwitchSummary(sex_switchlist_analyzed)
```

Only 5 genes were considered switching, so I will also try splitting by brain region.

One of the genes Atp6v0e2, had downregulation in females with AD in this study https://www.frontiersin.org/articles/10.3389/fnagi.2021.735611/full

### Get sex DTU for each brain region

Before analysis, I need to subset the data for each brain region.

```{r split out brain regions}
# make vector of brain regions
all_regions <- c("cerebellum", "cortex", "hippocampus", "striatum")

# split each region
lapply(all_regions, split_region)
```

Now that the data is split out, I can make switchAnalyzeRlist objects for analysis.

```{r make design and switchlists}
# use function on all brain regions
lapply(all_regions, make_switchlist)
```

Once these are made, I can run satuRn.

#### Run satuRn

```{r saturn by sex and brain region}
# use function on all brain regions
lapply(all_regions, filter_run_saturn) # this won't run all samples

filter_run_saturn("cortex") # no significant genes
filter_run_saturn("cerebellum") # no significant genes
filter_run_saturn("hippocampus") # no significant genes
filter_run_saturn("striatum") # this finds 3
```

Alright, now I have some differential used isoforms/genes. Let's visualize them with a volcano plot, although not including background genes in the analysis might be a bit much.

#### Make volcano plots

```{r volcano plots for sex split by brain region}
# use function
create_sex_volcano_plot(stri_sex_switchlist_analyzed)
```

These volcano plots are very sparse. I'm going to add the argument to include all genes and see if it is overwhelming. Note: I don't know if this is necessary or helpful.

```{r redo satuRn split by tissue and sex}
# run new function on all regions
lapply(all_regions, filter_run_saturn_noreduce)
```

Okay, now time to plot again.

```{r volcano plot again}
# use function on all switchlists
create_sex_volcano_plot_rm(cere_sex_switchlist_analyzed)
create_sex_volcano_plot_rm(cort_sex_switchlist_analyzed)
create_sex_volcano_plot_rm(hipp_sex_switchlist_analyzed)
create_sex_volcano_plot_rm(stri_sex_switchlist_analyzed)
```

This is slightly better, I guess. Now the non-significant genes are included at the bottom. There are very few significant genes, but I'll still make an UpSet plot to compare them.

```{r sex significant isoforms upset plot}
# run new function on all regions
lapply(all_regions, get_sig_isoforms)

# make all into list
all_sex_isoform_list <- list(
  cerebellum = cere_sex_sig_isoforms,
  cortex = cort_sex_sig_isoforms,
  hippocampus = hipp_sex_sig_isoforms,
  striatum = stri_sex_sig_isoforms
)

# make matrix
all_sex_isoform_list_mat <- list_to_matrix(all_sex_isoform_list)

# create combination matrix
all_sex_isoform_list_comb_mat <- make_comb_mat(all_sex_isoform_list_mat)

# make UpSet plot (ComplexHeatmap)
upset_sex_isoform <- UpSet(
  all_sex_isoform_list_comb_mat,
  comb_order = order(-comb_size(all_sex_isoform_list_comb_mat)),
  comb_col = viridis(4)[comb_degree(all_sex_isoform_list_comb_mat)],
  top_annotation = upset_top_annotation(all_sex_isoform_list_comb_mat,
    add_numbers = TRUE
  ),
  right_annotation = upset_right_annotation(all_sex_isoform_list_comb_mat,
    add_numbers = TRUE
  )
)

draw(upset_sex_isoform,
  column_title = "Overlap of sex-DTU isoforms by brain region"
)
```

There is no overlap! An UpSet plot might not be the best visualization if there is no overlap. I find this interesting since that means they are all brain-region specific sex differences. So what are they? Are they enriched for anything?

#### GO Analysis using gprofiler2

Before doing GO analysis, I need to pull out significant genes. GO analysis does not currently work at the isoform level (at least for gprofiler2).

```{r pull out signficant genes}
# run  on all brain regions
lapply(all_regions, get_cut_sig_genes)
```

Now, perform go analysis with gprofiler2.

```{r go analysis}
# run new function on all brain regions
lapply(all_regions, run_gprofiler)

run_gprofiler("striatum")
```

It does not look like there is a lot of enrichment here. There are only a few genes so it makes sense to me that there is very little enrichment.

```{r interactive gprofiler}
# get name
name <- substr("striatum", 1, 4)
# assign name
assign(
  "temp_sex_sig_genes",
  get(paste0(name, "_sex_sig_genes"))
)
# run gprofiler
temp_sex_gostres <- gost(
  query = temp_sex_sig_genes, organism = "mmusculus", ordered_query = FALSE,
  multi_query = FALSE, significant = TRUE, exclude_iea = FALSE,
  measure_underrepresentation = FALSE, evcodes = FALSE,
  user_threshold = 0.05, correction_method = "g_SCS", domain_scope = "custom",
  custom_bg = str_extract(
    unique(cere_sex_switchlist_analyzed$isoformFeatures$gene_id),
    pattern = "ENSMUSG..........."
  ),
  numeric_ns = "", sources = NULL, as_short_link = FALSE
)
# only save and plot if there are significant results
if (!is.null(temp_sex_gostres)) {
  # save object
  assign(paste0(name, "_sex_gostres"),
    temp_sex_gostres,
    envir = .GlobalEnv
  )
  # plot object
  gostplot(temp_sex_gostres, capped = TRUE, interactive = TRUE)
}
```

Lastly, I want to get the gene symbols.

```{r get top sex gene symbols}
# run new function on all regions to get symbols for each list
lapply(all_regions, get_gene_symbols_sex)
```

#### Clean up script

```{r tidy script}
style_file("07_dtu_region_sex.Rmd")
lint("07_dtu_region_sex.Rmd", linters = linters_with_defaults(
  object_length_linter = NULL,
  object_name_linter = NULL,
  object_usage_linter = NULL
))
```

Get processing time.

```{r proc time finish}
fptm <- proc.time() - ptm
fptm[3] / 60
```

Estimated processing time was 181 minutes but I went to seminar and was troubleshootingso does not take that long.

#### Software versions

```{r versions}
sessionInfo()
```

R version 4.3.0 beta (2023-04-12 r84240)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.2 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
 [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] grid      stats4    stats     graphics  grDevices utils     datasets  methods  
[9] base     

other attached packages:
 [1] here_1.0.1                    lintr_3.0.2                  
 [3] styler_1.9.1                  gprofiler2_0.2.1             
 [5] org.Mm.eg.db_3.17.0           viridis_0.6.2                
 [7] viridisLite_0.4.1             ComplexHeatmap_2.16.0        
 [9] IsoformSwitchAnalyzeR_1.99.17 pfamAnalyzeR_1.0.0           
[11] sva_3.48.0                    genefilter_1.82.0            
[13] mgcv_1.8-42                   nlme_3.1-162                 
[15] satuRn_1.8.0                  DEXSeq_1.46.0                
[17] RColorBrewer_1.1-3            AnnotationDbi_1.62.0         
[19] DESeq2_1.40.0                 SummarizedExperiment_1.30.0  
[21] GenomicRanges_1.52.0          GenomeInfoDb_1.36.0          
[23] IRanges_2.34.0                S4Vectors_0.38.0             
[25] MatrixGenerics_1.12.0         matrixStats_0.63.0           
[27] Biobase_2.60.0                BiocGenerics_0.46.0          
[29] BiocParallel_1.34.0           limma_3.56.0                 
[31] lubridate_1.9.2               forcats_1.0.0                
[33] stringr_1.5.0                 dplyr_1.1.2                  
[35] purrr_1.0.1                   readr_2.1.4                  
[37] tidyr_1.3.0                   tibble_3.2.1                 
[39] ggplot2_3.4.2                 tidyverse_2.0.0              

loaded via a namespace (and not attached):
  [1] splines_4.3.0                 later_1.3.0                  
  [3] BiocIO_1.10.0                 bitops_1.0-7                 
  [5] filelock_1.0.2                R.oo_1.25.0                  
  [7] rex_1.2.1                     XML_3.99-0.14                
  [9] lifecycle_1.0.3               rprojroot_2.0.3              
 [11] edgeR_3.42.0                  doParallel_1.0.17            
 [13] MASS_7.3-58.4                 processx_3.8.0               
 [15] lattice_0.21-8                ensembldb_2.24.0             
 [17] crosstalk_1.2.0               magrittr_2.0.3               
 [19] rmarkdown_2.21                plotly_4.10.1                
 [21] remotes_2.4.2                 yaml_2.3.7                   
 [23] httpuv_1.6.9                  pbapply_1.7-0                
 [25] DBI_1.1.3                     zlibbioc_1.46.0              
 [27] R.cache_0.16.0                R.utils_2.12.2               
 [29] AnnotationFilter_1.24.0       RCurl_1.98-1.12              
 [31] rappdirs_0.3.3                circlize_0.4.15              
 [33] GenomeInfoDbData_1.2.10       annotate_1.78.0              
 [35] codetools_0.2-19              DelayedArray_0.25.0          
 [37] xml2_1.3.4                    tidyselect_1.2.0             
 [39] shape_1.4.6                   futile.logger_1.4.3          
 [41] locfdr_1.1-8                  farver_2.1.1                 
 [43] BiocFileCache_2.8.0           GenomicAlignments_1.36.0     
 [45] jsonlite_1.8.4                GetoptLong_1.0.5             
 [47] ellipsis_0.3.2                survival_3.5-5               
 [49] iterators_1.0.14              systemfonts_1.0.4            
 [51] foreach_1.5.2                 tools_4.3.0                  
 [53] progress_1.2.2                ragg_1.2.5                   
 [55] Rcpp_1.0.10                   glue_1.6.2                   
 [57] gridExtra_2.3                 xfun_0.39                    
 [59] withr_2.5.0                   formatR_1.14                 
 [61] BiocManager_1.30.20           fastmap_1.1.1                
 [63] boot_1.3-28.1                 fansi_1.0.4                  
 [65] callr_3.7.3                   digest_0.6.31                
 [67] timechange_0.2.0              R6_2.5.1                     
 [69] mime_0.12                     textshaping_0.3.6            
 [71] colorspace_2.1-0              biomaRt_2.56.0               
 [73] RSQLite_2.3.1                 R.methodsS3_1.8.2            
 [75] utf8_1.2.3                    generics_0.1.3               
 [77] data.table_1.14.8             tximeta_1.18.0               
 [79] rtracklayer_1.60.0            htmlwidgets_1.6.2            
 [81] prettyunits_1.1.1             httr_1.4.5                   
 [83] pkgconfig_2.0.3               gtable_0.3.3                 
 [85] blob_1.2.4                    hwriter_1.3.2.1              
 [87] XVector_0.40.0                htmltools_0.5.5              
 [89] geneplotter_1.78.0            ProtGenerics_1.32.0          
 [91] clue_0.3-64                   scales_1.2.1                 
 [93] cyclocomp_1.1.0               png_0.1-8                    
 [95] knitr_1.42                    lambda.r_1.2.4               
 [97] rstudioapi_0.14               tzdb_0.3.0                   
 [99] reshape2_1.4.4                rjson_0.2.21                 
[101] curl_5.0.0                    cachem_1.0.8                 
[103] GlobalOptions_0.1.2           BiocVersion_3.17.1           
[105] parallel_4.3.0                desc_1.4.2                   
[107] restfulr_0.0.15               pillar_1.9.0                 
[109] vctrs_0.6.2                   promises_1.2.0.1             
[111] dbplyr_2.3.2                  xtable_1.8-4                 
[113] cluster_2.1.4                 tximport_1.28.0              
[115] evaluate_0.20                 VennDiagram_1.7.3            
[117] GenomicFeatures_1.52.0        cli_3.6.1                    
[119] locfit_1.5-9.7                compiler_4.3.0               
[121] futile.options_1.0.1          Rsamtools_2.16.0             
[123] rlang_1.1.1                   crayon_1.5.2                 
[125] labeling_0.4.2                ps_1.7.4                     
[127] plyr_1.8.8                    stringi_1.7.12               
[129] munsell_0.5.0                 Biostrings_2.68.0            
[131] lazyeval_0.2.2                Matrix_1.5-4                 
[133] BSgenome_1.68.0               hms_1.1.3                    
[135] bit64_4.0.5                   KEGGREST_1.40.0              
[137] statmod_1.5.0                 shiny_1.7.4                  
[139] interactiveDisplayBase_1.38.0 AnnotationHub_3.8.0          
[141] memoise_2.0.1                 bit_4.0.5                    
[143] xmlparsedata_1.0.5  