---
title: "Neurological Disease Gene Enrichment"
author: "Emma Jones"
date: "2023-03-14"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Differential Transcript Usage (DTU) Analysis across Sex in Mouse Brain lrRNA-Seq Data

The goal of this script is to compare my DTU genes to known Alzheimer's disease, psychiatric disorder, and CPAM case genes. It is dependent on scripts 01 - 07.

I am going to split this analysis by brain region, and am using the region_others isoform switch test results.

For code review, I am tracking the amount of time that this analysis will take.

```{r proc time}
ptm <- proc.time()
```

#### Load in packages

```{r load packages}
suppressPackageStartupMessages({
  library(tidyverse)
  library(IsoformSwitchAnalyzeR)
  library(viridis)
  library(styler)
  library(lintr)
  library(here)
  library(openxlsx)
  library(org.Hs.eg.db)
  library(org.Mm.eg.db)
  library(biomaRt)
  library(readxl)
  library(ComplexHeatmap)
})
```

Get functions into your environment

```{r read in functions}
source("functions.R", local = knitr::knit_global())
```

#### Load in data

Load in switchlist objects.

```{r load in switchlist objects}
# load object for region vs others
region_all_switchlist_analyzed <- readRDS(
  here("data", "switchlist_objects", "raw", "region_all_switchlist_list.Rds")
)

# load object for region vs region comparisons
region_region_switchlist_analyzed <- readRDS(
  here("data", "switchlist_objects", "raw", "region_region_switchlist_list.Rds")
)
```

#### AD Genes

Load in AD gene set, from this paper https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9005347/ 

```{r read AD gene set}
# read in from web
ad_genes <- read.xlsx("https://static-content.springer.com/esm/art%3A10.1038%2Fs41588-022-01024-z/MediaObjects/41588_2022_1024_MOESM4_ESM.xlsx",
  sheet = 6, rows = 5:87, cols = 4, colNames = FALSE
)

colnames(ad_genes) <- "symbol"

ad_genes <- unique(ad_genes)

mapped_ad_genes <- convert_human_to_mouse(ad_genes)

ad_mouse <- mapped_ad_genes$Gene.stable.ID.1
```

#### Psych Genes

Load in mood disorders gene set. This table does not have individual genes split out, so I need to do that. the paper we pull this gene set from is available here https://www.sciencedirect.com/science/article/pii/S0092867419312760 

```{r read in mood disorders gene sets}
# read in from web
mood_genes <- read.xlsx("https://ars.els-cdn.com/content/image/1-s2.0-S0092867419312760-mmc4.xlsx",
  sheet = 1, rows = 4:149, cols = 8, colNames = FALSE
)

colnames(mood_genes) <- "symbol"
# separate out the columns
new_col <- mood_genes %>%
  separate_wider_delim(symbol, ";",
    names = paste0("gene", 1:36),
    too_many = "debug", too_few = "debug"
  )

# clear up formatting
vals <- unlist(new_col[, 1:36], use.names = FALSE)
vals <- vals[!is.na(vals)]
vals <- vals[!vals == "" & !vals == "-"]
vals <- gsub("\\s*\\([^\\)]+\\)", "", vals)

mood_genes <- vals

mapped_genes <- convert_human_to_mouse(mood_genes)

psychiatric_mouse <- mapped_genes$Gene.stable.ID.1
```

#### C-PAM genes

Now, I am going to pull in C-PAM case genes from UAB's center for precision animal modeling. This information is all publicly available at https://uabcpam.softr.app/full-model-progress

```{r pull in cpam genes}
cpam_cases <- read_excel(here("doc", "CPAM", "CPAM_Cases.xlsx"))

cpam_genes <- cpam_cases$Gene[-c(1, length(cpam_cases))]

mapped_genes <- convert_human_to_mouse(cpam_genes)

cpam_mouse <- mapped_genes$Gene.stable.ID.1
```

### Run Comparisons

Now, I can use the results from this object to get my significant gene lists, Although they should all be saved already as outputs. To keep the ensembl ids, I will be doing this. I made a function to pull out the gene lists, compare, and save objects.

```{r run analysis}
# compare gene lists 
all_ad_genes <- lapply(
  region_all_switchlist_analyzed, 
  compare_switching_genes, 
  ad_mouse
)
all_cpam_genes <- lapply(
  region_all_switchlist_analyzed,
  compare_switching_genes,
  cpam_mouse
)
all_psych_genes <- lapply(
  region_all_switchlist_analyzed,
  compare_switching_genes,
  psychiatric_mouse
)
```

Now, save your data.

```{r save compare lists}
saveRDS(
  all_psych_genes,
  here("results", "neuro_dtu_genes", "all_psych_genes.Rds")
)

saveRDS(
  all_cpam_genes,
  here("results", "neuro_dtu_genes", "all_cpam_genes.Rds")
)
```

So, from these, there are no switching genes that are known ad risk genes. In addition, the striatum has no overlap with any gene lists, which makes sense because I only see 2 significant switching genes. For hippocampus and cortex, there is only one psych genes compared to the 7 for cerebellum, likely more reflective of having more genes in the cerebellum list (129 vs less than 50).

Either way, I can still make an upset plot for psychiatric genes at minimum.

```{r make upset plots}
# make matrix
all_psych_genes_mat <- list_to_matrix(all_psych_genes)

# create combination matrix
all_psych_genes_comb_mat <- make_comb_mat(all_psych_genes_mat)

# save
png(here("results", "psych_dtu_genes", "psych_genes_region_upset.png"))

# make UpSet plot (ComplexHeatmap)
upset_psych_genes <- UpSet(
  all_psych_genes_comb_mat,
  comb_order = order(-comb_size(all_psych_genes_comb_mat)),
  comb_col = viridis(4)[comb_degree(all_psych_genes_comb_mat)],
  top_annotation = upset_top_annotation(all_psych_genes_comb_mat,
    add_numbers = TRUE
  ),
  right_annotation = upset_right_annotation(all_psych_genes_comb_mat,
    add_numbers = TRUE
  )
)

draw(upset_psych_genes,
  column_title = "Overlap of DTU psychiatric risk genes by brain region"
)

dev.off()
```

It looks like there is no overlap of psychiatric risk genes.

## Repeat for region vs region comparisons

First, I need to pull out the individual dtu comparisons.

```{r pull out individual dtu comparisons}
sig_region_region <- dplyr::filter(
  region_region_switchlist_analyzed$isoformFeatures,
  abs(dIF) > 0.1 & isoform_switch_q_value < 0.05
)

# cerebellum cortex
cerebellum_cortex_subset <- filter(
  sig_region_region,
  condition_1 == "cerebellum" & condition_2 == "cortex"
)
cerebellum_cortex_dtu_genes <- cerebellum_cortex_subset$gene_id

# cerebellum hippocampus
cerebellum_hippocampus_subset <- filter(
  sig_region_region,
  condition_1 == "cerebellum" & condition_2 == "hippocampus"
)
cerebellum_hippocampus_dtu_genes <- cerebellum_hippocampus_subset$gene_id

# cerebellum striatum
cerebellum_striatum_subset <- filter(
  sig_region_region,
  condition_1 == "cerebellum" & condition_2 == "striatum"
)
cerebellum_striatum_dtu_genes <- cerebellum_striatum_subset$gene_id

# cortex hippocampus
cortex_hippocampus_subset <- filter(
  sig_region_region,
  condition_1 == "cortex" & condition_2 == "hippocampus"
)
cortex_hippocampus_dtu_genes <- cortex_hippocampus_subset$gene_id

# cortex striatum
cortex_striatum_subset <- filter(
  sig_region_region,
  condition_1 == "cortex" & condition_2 == "striatum"
)
cortex_striatum_dtu_genes <- cortex_striatum_subset$gene_id

# hippocampus striatum
hippocampus_striatum_subset <- filter(
  sig_region_region,
  condition_1 == "hippocampus" & condition_2 == "striatum"
)
hippocampus_striatum_dtu_genes <- hippocampus_striatum_subset$gene_id
```

We want to be able to apply our function again, so I will make a list of all these comparisons

```{r make list of gene lists}

all_comparisons <- list(cerebellum_cortex_dtu_genes = cerebellum_cortex_dtu_genes,
                        cerebellum_hippocampus_dtu_genes = cerebellum_hippocampus_dtu_genes,
                        cerebellum_striatum_dtu_genes = cerebellum_striatum_dtu_genes,
                        cortex_hippocampus_dtu_genes = cortex_hippocampus_dtu_genes,
                        cortex_striatum_dtu_genes = cortex_striatum_dtu_genes,
                        hippocampus_striatum_dtu_genes = hippocampus_striatum_dtu_genes)

```

I want to run the "simple" version of this function to preserve the gene lists and skip the step of using a switchlist object.

```{r run another analysis}
# compare gene lists 
region_region_ad_genes <- lapply(
  all_comparisons, 
  compare_switching_genes_simple, 
  ad_mouse
)
region_region_cpam_genes <- lapply(
  all_comparisons,
  compare_switching_genes_simple,
  cpam_mouse
)
region_region_psych_genes <- lapply(
  all_comparisons,
  compare_switching_genes_simple,
  psychiatric_mouse
)
```

Great, looks like we have genes for all three of these lists, even if there are not many for ad or cpam compared to the psychiatric list.

Now, save your data.

```{r save compare lists}
saveRDS(
  region_region_ad_genes,
  here("results", "neuro_dtu_genes", "region_region_ad_genes.Rds")
)
saveRDS(
  region_region_cpam_genes,
  here("results", "neuro_dtu_genes", "region_region_cpam_genes.Rds")
)
saveRDS(
  region_region_psych_genes,
  here("results", "neuro_dtu_genes", "region_region_psych_genes.Rds")
)
```

#### Clean up script

```{r tidy script}
style_file("08_dtu_neuro_diseases.Rmd")
lint("08_dtu_neuro_diseases.Rmd", linters = linters_with_defaults(
  object_length_linter = NULL,
  object_name_linter = NULL,
  object_usage_linter = NULL
))
```

Get processing time.

```{r proc time finish}
fptm <- proc.time() - ptm
fptm[3] / 60
```

This script takes 4 minutes to run.

#### Software versions

```{r versions}
sessionInfo()
```
R version 4.3.0 beta (2023-04-12 r84240)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.2 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
 [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] grid      stats4    stats     graphics  grDevices utils     datasets  methods  
[9] base     

other attached packages:
 [1] ComplexHeatmap_2.16.0         readxl_1.4.2                 
 [3] biomaRt_2.56.0                org.Mm.eg.db_3.17.0          
 [5] org.Hs.eg.db_3.17.0           openxlsx_4.2.5.2             
 [7] here_1.0.1                    lintr_3.0.2                  
 [9] styler_1.9.1                  viridis_0.6.2                
[11] viridisLite_0.4.1             IsoformSwitchAnalyzeR_1.99.17
[13] pfamAnalyzeR_1.0.0            sva_3.48.0                   
[15] genefilter_1.82.0             mgcv_1.8-42                  
[17] nlme_3.1-162                  satuRn_1.8.0                 
[19] DEXSeq_1.46.0                 RColorBrewer_1.1-3           
[21] AnnotationDbi_1.62.0          DESeq2_1.40.0                
[23] SummarizedExperiment_1.30.0   GenomicRanges_1.52.0         
[25] GenomeInfoDb_1.36.0           IRanges_2.34.0               
[27] S4Vectors_0.38.0              MatrixGenerics_1.12.0        
[29] matrixStats_0.63.0            Biobase_2.60.0               
[31] BiocGenerics_0.46.0           BiocParallel_1.34.0          
[33] limma_3.56.0                  lubridate_1.9.2              
[35] forcats_1.0.0                 stringr_1.5.0                
[37] dplyr_1.1.2                   purrr_1.0.1                  
[39] readr_2.1.4                   tidyr_1.3.0                  
[41] tibble_3.2.1                  ggplot2_3.4.2                
[43] tidyverse_2.0.0              

loaded via a namespace (and not attached):
  [1] splines_4.3.0                 later_1.3.0                  
  [3] BiocIO_1.10.0                 bitops_1.0-7                 
  [5] filelock_1.0.2                cellranger_1.1.0             
  [7] R.oo_1.25.0                   XML_3.99-0.14                
  [9] rex_1.2.1                     lifecycle_1.0.3              
 [11] doParallel_1.0.17             rprojroot_2.0.3              
 [13] edgeR_3.42.0                  processx_3.8.0               
 [15] lattice_0.21-8                ensembldb_2.24.0             
 [17] magrittr_2.0.3                rmarkdown_2.21               
 [19] remotes_2.4.2                 yaml_2.3.7                   
 [21] httpuv_1.6.9                  zip_2.2.2                    
 [23] pbapply_1.7-0                 DBI_1.1.3                    
 [25] zlibbioc_1.46.0               R.cache_0.16.0               
 [27] R.utils_2.12.2                AnnotationFilter_1.24.0      
 [29] RCurl_1.98-1.12               rappdirs_0.3.3               
 [31] circlize_0.4.15               GenomeInfoDbData_1.2.10      
 [33] annotate_1.78.0               codetools_0.2-19             
 [35] DelayedArray_0.25.0           xml2_1.3.4                   
 [37] shape_1.4.6                   tidyselect_1.2.0             
 [39] futile.logger_1.4.3           locfdr_1.1-8                 
 [41] BiocFileCache_2.8.0           GenomicAlignments_1.36.0     
 [43] jsonlite_1.8.4                GetoptLong_1.0.5             
 [45] ellipsis_0.3.2                iterators_1.0.14             
 [47] survival_3.5-5                foreach_1.5.2                
 [49] tools_4.3.0                   progress_1.2.2               
 [51] Rcpp_1.0.10                   glue_1.6.2                   
 [53] gridExtra_2.3                 xfun_0.39                    
 [55] withr_2.5.0                   formatR_1.14                 
 [57] BiocManager_1.30.20           fastmap_1.1.1                
 [59] boot_1.3-28.1                 fansi_1.0.4                  
 [61] callr_3.7.3                   digest_0.6.31                
 [63] timechange_0.2.0              R6_2.5.1                     
 [65] mime_0.12                     colorspace_2.1-0             
 [67] RSQLite_2.3.1                 R.methodsS3_1.8.2            
 [69] utf8_1.2.3                    generics_0.1.3               
 [71] tximeta_1.18.0                rtracklayer_1.60.0           
 [73] prettyunits_1.1.1             httr_1.4.5                   
 [75] pkgconfig_2.0.3               gtable_0.3.3                 
 [77] blob_1.2.4                    hwriter_1.3.2.1              
 [79] XVector_0.40.0                htmltools_0.5.5              
 [81] geneplotter_1.78.0            clue_0.3-64                  
 [83] ProtGenerics_1.32.0           scales_1.2.1                 
 [85] cyclocomp_1.1.0               png_0.1-8                    
 [87] knitr_1.42                    lambda.r_1.2.4               
 [89] rstudioapi_0.14               tzdb_0.3.0                   
 [91] reshape2_1.4.4                rjson_0.2.21                 
 [93] curl_5.0.0                    GlobalOptions_0.1.2          
 [95] cachem_1.0.8                  BiocVersion_3.17.1           
 [97] parallel_4.3.0                desc_1.4.2                   
 [99] restfulr_0.0.15               pillar_1.9.0                 
[101] vctrs_0.6.2                   promises_1.2.0.1             
[103] dbplyr_2.3.2                  cluster_2.1.4                
[105] xtable_1.8-4                  tximport_1.28.0              
[107] evaluate_0.20                 VennDiagram_1.7.3            
[109] GenomicFeatures_1.52.0        cli_3.6.1                    
[111] locfit_1.5-9.7                compiler_4.3.0               
[113] futile.options_1.0.1          Rsamtools_2.16.0             
[115] rlang_1.1.1                   crayon_1.5.2                 
[117] ps_1.7.4                      plyr_1.8.8                   
[119] stringi_1.7.12                munsell_0.5.0                
[121] Biostrings_2.68.0             lazyeval_0.2.2               
[123] Matrix_1.5-4                  BSgenome_1.68.0              
[125] hms_1.1.3                     bit64_4.0.5                  
[127] KEGGREST_1.40.0               statmod_1.5.0                
[129] shiny_1.7.4                   interactiveDisplayBase_1.38.0
[131] AnnotationHub_3.8.0           memoise_2.0.1                
[133] bit_4.0.5
