---
title: "View DTU Isoform Consequences"
author: "Emma Jones"
date: "2023-03-02"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This script depends on scripts 01 through 07, which identifies significantly switching genes.
For these genes, we can look at individual isoform switches.


```{r read in functions}
source("functions.R", local = knitr::knit_global())
```

read in data (might replace with function later)

```{r load in switchlist objects}
# load object comparing all regions to each other
region_region_switchlist_analyzed <- readRDS(
  here("data", "switchlist_objects", "region_region_switchlist_saturn.Rds")
  )
# load object for cerebellum
cerebellum_switchlist_analyzed <- readRDS(
  here("data", "switchlist_objects", "cerebellum_switchlist_saturn.Rds")
  )
# load object for cortex
cortex_switchlist_analyzed <- readRDS(
  here("data", "switchlist_objects", "cortex_switchlist_saturn.Rds")
  )
# load object for hippocampus
hippocampus_switchlist_analyzed <- readRDS(
  here("data", "switchlist_objects", "hippocampus_switchlist_saturn.Rds")
  )
# load object for cortex
striatum_switchlist_analyzed <- readRDS(
  here("data", "switchlist_objects", "striatum_switchlist_saturn.Rds")
  )
```

also, need to read in the sex-specific objects

```{r read in sex specific objects}
# load object for cerebellum
cerebellum_sex_switchlist_analyzed <- readRDS(
  here("data", "switchlist_objects", "cerebellum_sex_switchlist_saturn.Rds")
  )
# load object for cortex
cortex_sex_switchlist_analyzed <- readRDS(
  here("data", "switchlist_objects", "cortex_sex_switchlist_saturn.Rds")
  )
# load object for hippocampus
hippocampus_sex_switchlist_analyzed <- readRDS(
  here("data", "switchlist_objects", "hippocampus_sex_switchlist_saturn.Rds")
  )
# load object for cortex
striatum_sex_switchlist_analyzed <- readRDS(
  here("data", "switchlist_objects", "striatum_sex_switchlist_saturn.Rds")
  )
```



```{r examine isoform switch events}
# apply function for creating open reading frames and saving them
all_regions <- c("cerebellum", "cortex", "hippocampus", "striatum")
lapply(all_regions, add_save_orfs_sex)
```


```{r plot isoform switch events}
switchPlot(cerebellum_sex_switchlist_orf, gene = 'ENSMUSG00000024758.16')
switchPlot(cerebellum_sex_switchlist_orf, gene = 'ENSMUSG00000053819.17')
switchPlot(cerebellum_sex_switchlist_orf, gene = 'ENSMUSG00000072501.6')

switchPlot(cortex_sex_switchlist_orf, gene = 'ENSMUSG00000020056.17')
switchPlot(cortex_sex_switchlist_orf, gene = 'ENSMUSG00000035212.15')
switchPlot(cortex_sex_switchlist_orf, gene = 'ENSMUSG00000061479.16')

switchPlot(hippocampus_sex_switchlist_orf, gene = 'ENSMUSG00000025264.13')

switchPlot(striatum_sex_switchlist_orf, gene = 'ENSMUSG00000002107.19')
switchPlot(striatum_sex_switchlist_orf, gene = 'ENSMUSG00000022210.8')
switchPlot(striatum_sex_switchlist_orf, gene = 'ENSMUSG00000029128.13')
switchPlot(striatum_sex_switchlist_orf, gene = 'ENSMUSG00000034544.18')
switchPlot(striatum_sex_switchlist_orf, gene = 'ENSMUSG00000038365.9')
switchPlot(striatum_sex_switchlist_orf, gene = 'ENSMUSG00000040591.19')
switchPlot(striatum_sex_switchlist_orf, gene = 'ENSMUSG00000057914.16')
```
want to to the analyzeDeepHMHMM.

First I am going to analyze isoform switching by brain region

```{r edit function for orf}
# apply function for adding and saving orf
all_regions <- c("cortex", "cerebellum", "hippocampus", "striatum")

lapply(all_regions, add_save_orfs)

```

Great, now we have all the orfs annotated for these objects.


```{r plot sig orfs}

gene_for_plotting <- dplyr::filter(cerebellum_switchlist_analyzed$isoformFeatures,
    abs(dIF) > 0.1 & isoform_switch_q_value < 0.05)$gene_id

for(i in gene_for_plotting[1:50]) {
  switchPlot(cerebellum_switchlist_analyzed, gene = i)
}


cerebellum_switchlist_analyzed[["isoformFeatures"]]$condition_1 <- "other_regions"
cerebellum_switchlist_analyzed[["isoformFeatures"]]$condition_2 <- "cerebellum"

switchPlot(cerebellum_switchlist_orf, gene = "ENSMUSG00000004892.14", 
           localTheme = theme_bw(base_size = 10))

```

## Use External Tools

Now, I want to try using external tools for analysis. To do this I need to extract nucleotide and amino acid sequences. Isoform SwitchAnalyzeR has a function for this.

```{r extract sequences}

striatum_sex_switchlist_orf <- extractSequence(
  striatum_sex_switchlist_orf,
  pathToOutput = here("data", "switchlist_fasta"),
  writeToFile = TRUE
)

```

add external data

```{r add external sequence analysis}

striatum_sex_switchlist_more <- analyzeCPC2(
  switchAnalyzeRlist = striatum_sex_switchlist_orf,
  pathToCPC2resultFile = here("data", "CPC2", "result_cpc2.txt"),
  removeNoncodinORFs = TRUE
)

striatum_sex_switchlist_more <- analyzeSignalP(
  switchAnalyzeRlist = striatum_sex_switchlist_more,
  pathToSignalPresultFile = here("data", "SignalP6", "prediction_results.txt")
)

striatum_sex_switchlist_more <- analyzeDeepLoc2(
  switchAnalyzeRlist = striatum_sex_switchlist_more,
  pathToDeepLoc2resultFile = here("data", "DeepLoc2", "results_640660B000004A492D07E377.csv")
)

striatum_sex_switchlist_more <- analyzeDeepTMHMM(
  switchAnalyzeRlist = striatum_sex_switchlist_more,
  pathToDeepTMHMMresultFile = here("data", "DeepTMHMM", "TMRs.gff3")
)

striatum_sex_switchlist_more <- analyzePFAM(
  switchAnalyzeRlist = striatum_sex_switchlist_more,
  pathToPFAMresultFile = here("data", "Pfam", "pfam_results.txt")
)

striatum_sex_switchlist_more

switchPlot(striatum_sex_switchlist_more, gene = 'ENSMUSG00000002107.19')
switchPlot(striatum_sex_switchlist_more, gene = 'ENSMUSG00000022210.8')
switchPlot(striatum_sex_switchlist_more, gene = 'ENSMUSG00000029128.13')
switchPlot(striatum_sex_switchlist_more, gene = 'ENSMUSG00000034544.18')
switchPlot(striatum_sex_switchlist_more, gene = 'ENSMUSG00000038365.9')
switchPlot(striatum_sex_switchlist_more, gene = 'ENSMUSG00000040591.19')
switchPlot(striatum_sex_switchlist_more, gene = 'ENSMUSG00000057914.16')
```


```{r predict alternative splicing}

striatum_sex_switchlist_more <- analyzeAlternativeSplicing(
    switchAnalyzeRlist = striatum_sex_switchlist_more,
    quiet = TRUE
)

table(striatum_sex_switchlist_more$AlternativeSplicingAnalysis$IR)

consequencesOfInterest <- c('intron_retention','coding_potential','NMD_status','domains_identified','ORF_seq_similarity')

striatum_sex_switchlist_more <- analyzeSwitchConsequences(
    striatum_sex_switchlist_more,
    consequencesToAnalyze = consequencesOfInterest, 
    dIFcutoff = 0.1,
    showProgress = FALSE
)

extractSwitchSummary(striatum_sex_switchlist_more, dIFcutoff = 0.1, filterForConsequences = TRUE)

#this is cool and I want to show for all brain regions I think
extractSplicingSummary(
    striatum_sex_switchlist_more,
    asFractionTotal = FALSE,
    plotGenes=FALSE
)

splicingEnrichment <- extractSplicingEnrichment(
    striatum_sex_switchlist_more,
    splicingToAnalyze='all',
    returnResult=TRUE,
    returnSummary=TRUE
)


```

#### Do external analyisis on all comparisons (region region object)

```{r export all region data}

region_region_switchlist_orf <- extractSequence(
  region_region_switchlist_orf,
  pathToOutput = here("data", "switchlist_fasta"),
  writeToFile = TRUE,
  outputPrefix = "all_region"
)

```


pull in data

```{r all region import pfam}

region_region_switchlist_more <- analyzePFAM(
  switchAnalyzeRlist = region_region_switchlist_orf,
  pathToPFAMresultFile = here("data", "Pfam", "all_regions", "pfamscan_out.txt")
)

```

```{r analyze alternative splicing for all brain regions}

region_region_switchlist_more <- analyzeAlternativeSplicing(
    switchAnalyzeRlist = region_region_switchlist_more,
    quiet = TRUE
)

extractSplicingSummary(
    region_region_switchlist_more,
    asFractionTotal = FALSE,
    plotGenes=FALSE
)

splicingEnrichment <- extractSplicingEnrichment(
    region_region_switchlist_more,
    splicingToAnalyze='all',
    returnResult=TRUE,
    returnSummary=TRUE
)
```

