---
title: "Protein Domain Information"
author: "Emma Jones"
date: "2023-05-15"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


## Use Pfam to annotate protein domains

Now, I want to try using external tools for analysis. To do this I need to extract nucleotide and amino acid sequences. IsoformSwitchAnalyzeR has a function for this.

This script depends on scripts 00-13, and also includes run_pfam.sh.

#### Load packages

```{r load packages and set seed}
suppressPackageStartupMessages({
  library(tidyverse)
  library(IsoformSwitchAnalyzeR)
  library(viridis)
  library(styler)
  library(lintr)
  library(here)
})

set.seed(123)
```

#### Load in data

```{r load in data}
files <- list.files(here("data", "switchlist_objects", "de_added"),
  pattern = ".Rds"
)
# loop in files
for (i in files) {
  temp <- readRDS(
    here("data", "switchlist_objects", "de_added", i)
  )
  assign(paste0(str_sub(i, end = -11), "analyzed"),
    temp,
    envir = .GlobalEnv
  )
}
```

Now that the files are read in, we can extract the amino acid sequences

```{r extract sequences}
striatum_sex_analyzed <- extractSequence(
  striatum_sex_analyzed,
  pathToOutput = here("data", "switchlist_fasta"),
  outputPrefix = "striatum_sex",
  writeToFile = TRUE
)

region_region_analyzed <- extractSequence(
  region_region_analyzed,
  pathToOutput = here("data", "switchlist_fasta"),
  outputPrefix = "region_region",
  writeToFile = TRUE
)

cerebellum_analyzed <- extractSequence(
  cerebellum_analyzed,
  pathToOutput = here("data", "switchlist_fasta"),
  outputPrefix = "cerebellum",
  writeToFile = TRUE
)

cortex_analyzed <- extractSequence(
  cortex_analyzed,
  pathToOutput = here("data", "switchlist_fasta"),
  outputPrefix = "cortex",
  writeToFile = TRUE
)

hippocampus_analyzed <- extractSequence(
  hippocampus_analyzed,
  pathToOutput = here("data", "switchlist_fasta"),
  outputPrefix = "hippocampus",
  writeToFile = TRUE
)

striatum_analyzed <- extractSequence(
  striatum_analyzed,
  pathToOutput = here("data", "switchlist_fasta"),
  outputPrefix = "striatum",
  writeToFile = TRUE
)
```

PFAM!!! Start with region_region comparison switchlist

```{bash run pfan on region region AA}
cd /opt/PfamScan/

perl -I /opt/PfamScan/ pfam_scan.pl \
-fasta /data/user/efjones/230227_EJ_MouseBrainIsoDiv/data/switchlist_fasta/region_region_AA.fasta \
-dir /PfamScan/PfamFiles \
| tee /data/user/efjones/230227_EJ_MouseBrainIsoDiv/results/PfamScan/region_region_pfamscan_out.txt
```

Do the same thing for each region vs others

cerebellum

```{bash run pfam on each region - cerebellum}
cd /opt/PfamScan/

perl -I /opt/PfamScan/ pfam_scan.pl \
-fasta /data/user/efjones/230227_EJ_MouseBrainIsoDiv/data/switchlist_fasta/cerebellum_AA.fasta \
-dir /PfamScan/PfamFiles \
| tee /data/user/efjones/230227_EJ_MouseBrainIsoDiv/results/PfamScan/cerebellum_pfamscan_out.txt
```

cortex

```{bash run pfam on each region - cortex}
cd /opt/PfamScan/

perl -I /opt/PfamScan/ pfam_scan.pl \
-fasta /data/user/efjones/230227_EJ_MouseBrainIsoDiv/data/switchlist_fasta/cortex_AA.fasta \
-dir /PfamScan/PfamFiles \
| tee /data/user/efjones/230227_EJ_MouseBrainIsoDiv/results/PfamScan/cortex_pfamscan_out.txt
```

hippocampus

```{bash run pfam on each region - hippocampus}
cd /opt/PfamScan/

perl -I /opt/PfamScan/ pfam_scan.pl \
-fasta /data/user/efjones/230227_EJ_MouseBrainIsoDiv/data/switchlist_fasta/hippocampus_AA.fasta \
-dir /PfamScan/PfamFiles \
| tee /data/user/efjones/230227_EJ_MouseBrainIsoDiv/results/PfamScan/hippocampus_pfamscan_out.txt
```

striatum

```{bash run pfam on each region - striatum}
cd /opt/PfamScan/

perl -I /opt/PfamScan/ pfam_scan.pl \
-fasta /data/user/efjones/230227_EJ_MouseBrainIsoDiv/data/switchlist_fasta/striatum_AA.fasta \
-dir /PfamScan/PfamFiles \
| tee /data/user/efjones/230227_EJ_MouseBrainIsoDiv/results/PfamScan/striatum_pfamscan_out.txt
```
finally, do striatum split by sex

```{bash run pfamscan on striatum split by sex}
cd /opt/PfamScan/

perl -I /opt/PfamScan/ pfam_scan.pl \
-fasta /data/user/efjones/230227_EJ_MouseBrainIsoDiv/data/switchlist_fasta/striatum_sex_AA.fasta \
-dir /PfamScan/PfamFiles \
| tee /data/user/efjones/230227_EJ_MouseBrainIsoDiv/results/PfamScan/striatum_sex_pfamscan_out.txt
```

#### Pull pfam results back into objects

Then add this back to your object.

```{r annotate protein domains}
# region region
region_region_switchlist_pfam <- analyzePFAM(
  switchAnalyzeRlist = region_region_analyzed,
  pathToPFAMresultFile = 
    here("results", "PfamScan", "region_region_pfamscan_out.txt")
)
# cerebellum other
cerebellum_switchlist_pfam <- analyzePFAM(
  switchAnalyzeRlist = cerebellum_analyzed,
  pathToPFAMresultFile = here("results", "PfamScan", "cerebellum_pfamscan_out.txt")
)
# cortex other
cortex_switchlist_pfam <- analyzePFAM(
  switchAnalyzeRlist = cortex_analyzed,
  pathToPFAMresultFile = here("results", "PfamScan", "cortex_pfamscan_out.txt")
)
# hippocampus other
hippocampus_switchlist_pfam <- analyzePFAM(
  switchAnalyzeRlist = hippocampus_analyzed,
  pathToPFAMresultFile = here("results", "PfamScan", "hippocampus_pfamscan_out.txt")
)
# striatum other
striatum_switchlist_pfam <- analyzePFAM(
  switchAnalyzeRlist = striatum_analyzed,
  pathToPFAMresultFile = here("results", "PfamScan", "striatum_pfamscan_out.txt")
)
# striatum sex
striatum_sex_switchlist_pfam <- analyzePFAM(
  switchAnalyzeRlist = striatum_sex_analyzed,
  pathToPFAMresultFile = here("results", "PfamScan", "striatum_sex_pfamscan_out.txt")
)

```

plot everything


```{r plot with protein domain info}
# region region
switchPlotTopSwitches(region_region_switchlist_pfam,
  pathToOutput = here(
    "results", "plots", "switch_plots", "pfam_added",
    "region_region"
  ),
  splitFunctionalConsequences = FALSE
)
# cerebellum
switchPlotTopSwitches(cerebellum_switchlist_pfam,
  pathToOutput = here(
    "results", "plots", "switch_plots", "pfam_added",
    "cerebellum"
  ),
  splitFunctionalConsequences = FALSE
)
# cortex
switchPlotTopSwitches(cortex_switchlist_pfam,
  pathToOutput = here(
    "results", "plots", "switch_plots", "pfam_added",
    "cortex"
  ),
  splitFunctionalConsequences = FALSE
)
# hippocampus
switchPlotTopSwitches(hippocampus_switchlist_pfam,
  pathToOutput = here(
    "results", "plots", "switch_plots", "pfam_added",
    "hippocampus"
  ),
  splitFunctionalConsequences = FALSE
)
# striatum
switchPlotTopSwitches(striatum_switchlist_pfam,
  pathToOutput = here(
    "results", "plots", "switch_plots", "pfam_added",
    "striatum"
  ),
  splitFunctionalConsequences = FALSE
)
# striatum - sex
switchPlotTopSwitches(striatum_sex_switchlist_pfam,
  pathToOutput = here(
    "results", "plots", "switch_plots", "pfam_added",
    "striatum"
  ),
  splitFunctionalConsequences = FALSE
)
```
