---
title: "compare results"
author: "Emma Jones"
date: "2023-05-05"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Comparing results of satuRn and DEXSeq

This script is for comparing the results of isoform switch tests with satuRn and DEXSeq (with and without SVA)

```{r load in packages}
library(tidyverse)
library(IsoformSwitchAnalyzeR)
library(ComplexHeatmap)
library(viridis)
library(org.Mm.eg.db)
library(gprofiler2)
library(styler)
library(lintr)
library(here)
```
### Load in data


```{r load in data}
# load in switchlist comparisons
load(file = here("data", "dtu_versions", "switchlists.RData"))
```

Since I want to make dfs 4 times, I'm going to make a function that creates the df for me


```{r make dfs for dtu with pvalues}
# make function for creating dataframes that extract the DTUs
make_sig_df <- function(object = x, name = y) {
  # pull features
  sig_features_temp <- object$isoformFeatures
  # filter for signifciant events
  sig_features_temp <- dplyr::filter(
    sig_features_temp, abs(dIF) > 0.1 & isoform_switch_q_value < 0.05
    )
  # get cols of interest
  temp_subset <- 
    sig_features_temp[,c("isoform_id", "gene_id", "condition_1", "condition_2",
                         "gene_name", "isoform_switch_q_value",
                         "gene_switch_q_value")]
  # export subset to global environment
  assign(paste0(name, "_results"), temp_subset, envir = .GlobalEnv)
}

# apply function 
make_sig_df(object = region_region_switchlist_analyzed, name = "dexseq_sva") 
make_sig_df(object = region_region_switchlist_saturn, name = "saturn_sva")
make_sig_df(object = region_region_switchlist_analyzed_nobatch, name = "dexseq_no_sva")
make_sig_df(object = region_region_switchlist_saturn_nobatch, name = "saturn_no_sva")

```

Now that I have separate data frames for each condition, I need to keep the DTUs that actually overlap. I think that the best way to do this is to add suffixes to the columns so it is clear which qvalue comes from where with inner_join.

```{r inner join}
# do dexseq vs saturn for sva
sva_joined <- inner_join(dexseq_sva_results, saturn_sva_results, by = c("isoform_id", "gene_id", "condition_1", "condition_2"), suffix = c("_dexseq", "_saturn"))
# do dexseq vs saturn for no sva
no_sva_joined <- inner_join(dexseq_no_sva_results, saturn_no_sva_results, by = c("isoform_id", "gene_id", "condition_1", "condition_2"), suffix = c("_dexseq", "_saturn"))

# compare dexseq with and without sva
dexseq_joined <- inner_join(dexseq_sva_results, dexseq_no_sva_results, by = c("isoform_id", "gene_id", "condition_1", "condition_2"), suffix = c("_sva", "_no"))
# compare saturn with and without sva
saturn_joined <- inner_join(saturn_sva_results, saturn_no_sva_results, by = c("isoform_id", "gene_id", "condition_1", "condition_2"), suffix = c("_sva", "_no"))

```


```{r plot and get cor}

# plot log because so many values are very low
plot(log(sva_joined$isoform_switch_q_value_dexseq),
     log(sva_joined$isoform_switch_q_value_saturn)
     )
# get pearson correlation - .6859
cor(log(sva_joined$isoform_switch_q_value_dexseq),
    log(sva_joined$isoform_switch_q_value_saturn)
    )
# get spearman correlation - .6271
cor(log(sva_joined$isoform_switch_q_value_dexseq),
    log(sva_joined$isoform_switch_q_value_saturn), method = "spearman"
    )

# plot the log because so many values are very low
plot(log(no_sva_joined$isoform_switch_q_value_dexseq),
     log(no_sva_joined$isoform_switch_q_value_saturn)
     )
# get pearson correlation - .6711
cor(log(no_sva_joined$isoform_switch_q_value_dexseq),
    log(no_sva_joined$isoform_switch_q_value_saturn)
    )
# get spearman correlation - .5796
cor(log(no_sva_joined$isoform_switch_q_value_dexseq),
    log(no_sva_joined$isoform_switch_q_value_saturn), method = "spearman"
    )

#do for dexseq
# plot log because so many values are very low
plot(log(dexseq_joined$isoform_switch_q_value_sva),
     log(dexseq_joined$isoform_switch_q_value_no)
     )
# get pearson correlation - .8888
cor(log(dexseq_joined$isoform_switch_q_value_sva),
    log(dexseq_joined$isoform_switch_q_value_no)
    )
# get spearman correlation - .6524
cor(log(dexseq_joined$isoform_switch_q_value_sva),
    log(dexseq_joined$isoform_switch_q_value_no), method = "spearman"
    )

## plot log because so many values are very low
plot(log(saturn_joined$isoform_switch_q_value_sva),
     log(saturn_joined$isoform_switch_q_value_no)
     )
# get pearson correlation - .8877
cor(log(saturn_joined$isoform_switch_q_value_sva),
    log(saturn_joined$isoform_switch_q_value_no)
    )
# get spearman correlation -.7598
cor(log(saturn_joined$isoform_switch_q_value_sva),
    log(saturn_joined$isoform_switch_q_value_no), method = "spearman"
    )
```

Overall, with spearman correlation, SVA helps satuRn and DEXSeq results to correlate more, and satuRn is more correlated with itself than DEXSeq. DEXSeq seems to have the biggest fluctuations in pvalue, suggesting that it was possibly previously inflated. (like e-279).

Because each list contains results for all 6 comparisons, I'm going to filter down so that the DTU genes are only from one of those comparisons (some repeat, avoiding problems there)

```{r filter down to one region region comparison}

dexseq_sva_results_small <- filter(dexseq_sva_results, condition_1 == "cerebellum" &
                             condition_2 == "striatum")

dexseq_no_sva_results_small <- filter(dexseq_no_sva_results, condition_1 == "cerebellum" &
                             condition_2 == "striatum")

saturn_sva_results_small <- filter(saturn_sva_results, condition_1 == "cerebellum" &
                             condition_2 == "striatum")

saturn_no_sva_results_small <- filter(saturn_no_sva_results, condition_1 == "cerebellum" &
                             condition_2 == "striatum")

```


I installed the ggvenn package to make a Venn diagram.

```{r plot venn diagrams}
# install potnential package; need to add to docker later if used in paper
if (!require(devtools)) install.packages("devtools")
devtools::install_github("yanlinlin82/ggvenn")
library(ggvenn)

# get colors
install.packages("rcartocolor")
my_colors <- carto_pal(7, "Vivid")
my_colors

# make results into a list
for_plot <- list(dexseq_sva = dexseq_sva_results_small$isoform_id,
           saturn_sva = saturn_sva_results_small$isoform_id,
           dexseq_no_sva = dexseq_sva_no_results_small$isoform_id,
           saturn_no_sva = saturn_sva_no_results_small$isoform_id)
# plot venn diagram
ggvenn(for_plot,
       fill_color = c("#E58606", "#5D69B1", "#52BCA3", "#99C945"),
       stroke_size = 0.5, set_name_size = 4
       )

```


Another way to visualize this is an upset plot. I'll use the complexheatmap version so it can have numbers on the plot.

```{r make upsetplot}

# create matrix
for_plot_mat <- list_to_matrix(for_plot)

# generate combination matrix
for_plot_comb_mat <- make_comb_mat(for_plot_mat)

# make UpSet plot (ComplexHeatmap)
UpSet(for_plot_comb_mat,
  comb_order = order(-comb_size(for_plot_comb_mat)),
  comb_col = c("#E58606", "#5D69B1", "#52BCA3", "#99C945")[comb_degree(for_plot_comb_mat)],
  top_annotation = upset_top_annotation(for_plot_comb_mat,
    add_numbers = TRUE
  ),
  right_annotation = upset_right_annotation(for_plot_comb_mat,
    add_numbers = TRUE
  )
)

```

Next, I going to make some switchplots. I want to get the most unqiue saturn genes, aka not overlapping with dexseq.

```{r narrow transcript lists}
shared_results <- intersect(dexseq_sva_results_small$isoform_id,
                        saturn_sva_results_small$isoform_id)

unique_transcripts <- saturn_sva_results_small$isoform_id[!saturn_sva_results_small$isoform_id %in% shared_results]

unique_results <- saturn_sva_results[saturn_sva_results$isoform_id %in% unique_transcripts,]

unique_results_ordered <- unique_results[order(unique_results$isoform_switch_q_value),]

top_saturn_genes <- unique(head(unique_results_ordered$gene_id, 20))

# do the same for dexseq
unique_dexseq_transcripts <- dexseq_sva_results_small$isoform_id[!dexseq_sva_results_small$isoform_id %in% shared_results]

dexseq_unique_results <- dexseq_sva_results[dexseq_sva_results$isoform_id %in% unique_dexseq_transcripts,]

dexseq_unique_results_ordered <- dexseq_unique_results[order(dexseq_unique_results$isoform_switch_q_value),]

top_dexseq_genes <- unique(head(dexseq_unique_results_ordered$gene_id, 20))


```

Now that I have my top_saturn_genes, I will make some switch plots. Actually, I dont have the orfs anotated so it will just be that bottom portion.

```{r plot plot plot}
# plot saturn
plot_iso_expression <- function(gene_id){
  plot <- switchPlotIsoUsage(
    region_region_switchlist_saturn,
    gene = gene_id,
    condition1 = "cerebellum",
    condition2 = "striatum"
    )
  plot + ggtitle(gene_id)
}

mapply(plot_iso_expression, top_saturn_genes, SIMPLIFY = FALSE)

# plot dexseq
plot_iso_expression <- function(gene_id){
  plot <- switchPlotIsoUsage(
    region_region_switchlist_analyzed,
    gene = gene_id,
    condition1 = "cerebellum",
    condition2 = "striatum"
    )
  plot + ggtitle(gene_id)
}

mapply(plot_iso_expression, top_dexseq_genes, SIMPLIFY = FALSE)

```

