---
title: "View DTU Isoform Consequences"
author: "Emma Jones"
date: "2023-03-02"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This script depends on the dtu_region_sex script, which identified a few significantly switched genes.

For these few genes, we can look at individual isoform switches.

```{r examine isoform switch events}
# make function to examine isoform switch events
add_save_orfs <- function(x) {
  # pull in switchlist
  assign("switchlist_analyzed", get(paste0(x, "_sex_switchlist_analyzed")))
  # add open reading frames
  switchlist_analyzed <- addORFfromGTF(
    switchAnalyzeRlist = switchlist_analyzed,
    pathToGTF = here("data", "gencode_annotations",
                     "gencode.vM31.primary_assembly.annotation.gtf"))
  # add novel isoform orfs
  switchlist_analyzed <- analyzeNovelIsoformORF(
  switchlist_analyzed, analysisAllIsoformsWithoutORF = TRUE)
  # save
  saveRDS(switchlist_analyzed, here(
    "data", "switchlist_objects","orf_added", paste0(
      x,"_sex_switchlist_orf.Rds"
      )
    )
    )
  # assign object
  assign(paste0(name, "_sex_switchlist_analyzed"), switchlist_analyzed,
         envir = .GlobalEnv)
}
# apply new function
all_regions <- c("cerebellum", "cortex", "hippocampus", "striatum")
lapply(all_regions[2:4], add_save_orfs)
```


```{r plot isoform switch events}
switchPlot(cerebellum_sex_switchlist_orf, gene = 'ENSMUSG00000024758.16')
switchPlot(cerebellum_sex_switchlist_orf, gene = 'ENSMUSG00000053819.17')
switchPlot(cerebellum_sex_switchlist_orf, gene = 'ENSMUSG00000072501.6')

switchPlot(cortex_sex_switchlist_orf, gene = 'ENSMUSG00000020056.17')
switchPlot(cortex_sex_switchlist_orf, gene = 'ENSMUSG00000035212.15')
switchPlot(cortex_sex_switchlist_orf, gene = 'ENSMUSG00000061479.16')

switchPlot(hippocampus_sex_switchlist_orf, gene = 'ENSMUSG00000025264.13')

switchPlot(striatum_sex_switchlist_orf, gene = 'ENSMUSG00000002107.19')
switchPlot(striatum_sex_switchlist_orf, gene = 'ENSMUSG00000022210.8')
switchPlot(striatum_sex_switchlist_orf, gene = 'ENSMUSG00000029128.13')
switchPlot(striatum_sex_switchlist_orf, gene = 'ENSMUSG00000034544.18')
switchPlot(striatum_sex_switchlist_orf, gene = 'ENSMUSG00000038365.9')
switchPlot(striatum_sex_switchlist_orf, gene = 'ENSMUSG00000040591.19')
switchPlot(striatum_sex_switchlist_orf, gene = 'ENSMUSG00000057914.16')
```
For plotting, want to add gene names and make function. Also want to to the analyzeDeepHMHMM.

First I am going to analyze isoform switching by brain region

```{r edit function for orf}
add_save_orfs <- function(x) {
  # pull in switchlist
  assign("switchlist_analyzed", get(paste0(x, "_switchlist_analyzed")))
  # add open reading frames
  switchlist_analyzed <- addORFfromGTF(
    switchAnalyzeRlist = switchlist_analyzed,
    pathToGTF = here("data", "gencode_annotations",
                     "gencode.vM31.primary_assembly.annotation.gtf"))
  # add novel isoform orfs
  switchlist_analyzed <- analyzeNovelIsoformORF(
  switchlist_analyzed, analysisAllIsoformsWithoutORF = TRUE)
  # save
  saveRDS(switchlist_analyzed, here(
    "data", "switchlist_objects","orf_added", paste0(
      x,"_switchlist_orf.Rds"
      )
    )
    )
  # assign object
  assign(paste0(x, "_switchlist_analyzed"), switchlist_analyzed,
         envir = .GlobalEnv)
}
# apply new function
all_regions <- c("cortex", "cerebellum", "hippocampus", "striatum")

lapply(all_regions, add_save_orfs)

```

Great, now we have all the orfs annotated for these objects.


```{r plot sig orfs}

gene_for_plotting <- dplyr::filter(cerebellum_switchlist_analyzed$isoformFeatures,
    abs(dIF) > 0.1 & isoform_switch_q_value < 0.05)$gene_id

for(i in gene_for_plotting[1:50]) {
  switchPlot(cerebellum_switchlist_analyzed, gene = i)
}


cerebellum_switchlist_analyzed[["isoformFeatures"]]$condition_1 <- "other_regions"
cerebellum_switchlist_analyzed[["isoformFeatures"]]$condition_2 <- "cerebellum"

switchPlot(cerebellum_switchlist_orf, gene = "ENSMUSG00000004892.14", 
           localTheme = theme_bw(base_size = 10))

```

## Use External Tools

Now, I want to try using external tools for analysis. To do this I need to extract nucleotide and amino acid sequences. Isoform SwitchAnalyzeR has a function for this.

```{r extract sequences}

striatum_sex_switchlist_orf <- extractSequence(
  striatum_sex_switchlist_orf,
  pathToOutput = here("data", "switchlist_fasta"),
  writeToFile = TRUE
)

```
add external data

```{r add external sequence analysis}

striatum_sex_switchlist_more <- analyzeCPC2(
  switchAnalyzeRlist = striatum_sex_switchlist_orf,
  pathToCPC2resultFile = here("data", "CPC2", "result_cpc2.txt"),
  removeNoncodinORFs = TRUE
)

striatum_sex_switchlist_more <- analyzeSignalP(
  switchAnalyzeRlist = striatum_sex_switchlist_more,
  pathToSignalPresultFile = here("data", "SignalP6", "prediction_results.txt")
)

striatum_sex_switchlist_more <- analyzeDeepLoc2(
  switchAnalyzeRlist = striatum_sex_switchlist_more,
  pathToDeepLoc2resultFile = here("data", "DeepLoc2", "results_640660B000004A492D07E377.csv")
)

striatum_sex_switchlist_more <- analyzeDeepTMHMM(
  switchAnalyzeRlist = striatum_sex_switchlist_more,
  pathToDeepTMHMMresultFile = here("data", "DeepTMHMM", "TMRs.gff3")
)

striatum_sex_switchlist_more <- analyzePFAM(
  switchAnalyzeRlist = striatum_sex_switchlist_more,
  pathToPFAMresultFile = here("data", "Pfam", "pfam_results.txt")
)

striatum_sex_switchlist_more

switchPlot(striatum_sex_switchlist_more, gene = 'ENSMUSG00000002107.19')
switchPlot(striatum_sex_switchlist_more, gene = 'ENSMUSG00000022210.8')
switchPlot(striatum_sex_switchlist_more, gene = 'ENSMUSG00000029128.13')
switchPlot(striatum_sex_switchlist_more, gene = 'ENSMUSG00000034544.18')
switchPlot(striatum_sex_switchlist_more, gene = 'ENSMUSG00000038365.9')
switchPlot(striatum_sex_switchlist_more, gene = 'ENSMUSG00000040591.19')
switchPlot(striatum_sex_switchlist_more, gene = 'ENSMUSG00000057914.16')
```
I needed to do this earlier, I will edit the scripts in order to add this.

```{r annotate genes!}
get_gene_symbols <- function(x) {
  # assign name
  assign("temp_switchlist", get(paste0(x, "_sex_switchlist_analyzed")))
  # pull shorter gene IDs
  temp_switchlist[["isoformFeatures"]]$shorter <-
    str_extract(temp_switchlist[["isoformFeatures"]]$gene_id,
      pattern = "ENSMUSG..........."
    )
  # get gene symbols from annotation dbi
  temp_gene_symbols <- AnnotationDbi::select(
    org.Mm.eg.db,
    keys = unique(temp_switchlist[["isoformFeatures"]]$shorter),
    columns = c("SYMBOL", "ENSEMBL", "GENETYPE"), keytype = "ENSEMBL"
  )
  # add symbols and biotypes to object
  temp_switchlist[["isoformFeatures"]] <-
    left_join(temp_switchlist[["isoformFeatures"]],
      temp_gene_symbols,
      by = c("shorter" = "ENSEMBL")
    )
  # add to correct columns
  temp_switchlist[["isoformFeatures"]]$gene_name <-
    temp_switchlist[["isoformFeatures"]]$SYMBOL
  temp_switchlist[["isoformFeatures"]]$gene_biotype <-
    temp_switchlist[["isoformFeatures"]]$GENETYPE
  # remove extra columns
  temp_switchlist[["isoformFeatures"]]$shorter <- NULL
  temp_switchlist[["isoformFeatures"]]$SYMBOL <- NULL
  temp_switchlist[["isoformFeatures"]]$GENETYPE <- NULL
  # rename object
  assign(paste0(x, "_sex_switchlist_analyzed"),
    temp_switchlist,
    envir = .GlobalEnv
  )
  saveRDS(temp_switchlist, here(
    "data", "switchlist_objects",
    paste0(x, "_sex_switchlist_analyzed.Rds")
  ))
}

# apply function to all brain regions
lapply(all_regions, get_gene_symbols)

```


```{r predict alternative splicing}

striatum_sex_switchlist_more <- analyzeAlternativeSplicing(
    switchAnalyzeRlist = striatum_sex_switchlist_more,
    quiet = TRUE
)

table(striatum_sex_switchlist_more$AlternativeSplicingAnalysis$IR)

consequencesOfInterest <- c('intron_retention','coding_potential','NMD_status','domains_identified','ORF_seq_similarity')

striatum_sex_switchlist_more <- analyzeSwitchConsequences(
    striatum_sex_switchlist_more,
    consequencesToAnalyze = consequencesOfInterest, 
    dIFcutoff = 0.1,
    showProgress = FALSE
)

extractSwitchSummary(striatum_sex_switchlist_more, dIFcutoff = 0.1, filterForConsequences = TRUE)

#this is cool and I want to show for all brain regions I think
extractSplicingSummary(
    striatum_sex_switchlist_more,
    asFractionTotal = FALSE,
    plotGenes=FALSE
)

splicingEnrichment <- extractSplicingEnrichment(
    striatum_sex_switchlist_more,
    splicingToAnalyze='all',
    returnResult=TRUE,
    returnSummary=TRUE
)


```

#### Do external analyisis on all comparisons (region region object)

```{r export all region data}

region_region_switchlist_orf <- extractSequence(
  region_region_switchlist_orf,
  pathToOutput = here("data", "switchlist_fasta"),
  writeToFile = TRUE,
  outputPrefix = "all_region"
)

```


pull in data

```{r all region import pfam}

region_region_switchlist_more <- analyzePFAM(
  switchAnalyzeRlist = region_region_switchlist_orf,
  pathToPFAMresultFile = here("data", "Pfam", "all_regions", "pfamscan_out.txt")
)

```

```{r analyze alternative splicing for all brain regions}

region_region_switchlist_more <- analyzeAlternativeSplicing(
    switchAnalyzeRlist = region_region_switchlist_more,
    quiet = TRUE
)

extractSplicingSummary(
    region_region_switchlist_more,
    asFractionTotal = FALSE,
    plotGenes=FALSE
)

splicingEnrichment <- extractSplicingEnrichment(
    region_region_switchlist_more,
    splicingToAnalyze='all',
    returnResult=TRUE,
    returnSummary=TRUE
)
```

