---
title: 'DTU Analysis: Brain Region and Sex'
author: "Emma Jones"
date: "2023-02-27"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Differential Transcript Usage (DTU) Analysis across Sex in Mouse Brain lrRNA-Seq Data

#### Load in packages

```{r load packages}
library(tidyverse)
library(IsoformSwitchAnalyzeR)
library(ComplexHeatmap)
library(viridis)
library(org.Mm.eg.db)
library(gprofiler2)
library(styler)
library(lintr)
library(here)
```

#### Load in data

Next, I want to read in data and get cpm values. Since I'll be doing this for all 3 analyses, I made it an R script that can be sourced.

```{r read in data and get cpm values}
source("calculate_cpm.R", local = knitr::knit_global())
source("functions.R", local = knitr::knit_global())
```

### Get DTU by sex

Now that I have shown DTU differences across tissue, I am going to look at differences across sex. I'll need to create a new switchAnalyzeRlist object with metadata for sex.

```{r create metadata and create object for sex}
# make sex design
sex_design <- data.frame(
  sampleID = sample_collection_metadata$sample_id,
  condition = sample_collection_metadata$sex
)
# make switchlist
sex_switchlist <- importRdata(
  isoformCountMatrix = merged_counts_iso,
  isoformRepExpression = cpm_iso,
  designMatrix = sex_design,
  isoformExonAnnoation = here(
    "data", "nextflow", "results", "bambu",
    "extended_annotations.gtf"
  ),
  isoformNtFasta = here("data", "gffread", "isoform_sequences.fa"),
  showProgress = FALSE
)
```

Now that I have my object, I can run a DTU analysis.

```{r run DEXSeq}
# filter data
sex_switchlist <- preFilter(sex_switchlist, geneExpressionCutoff = NULL)
# run DEXSeq
sex_switchlist_analyzed <- isoformSwitchTestDEXSeq(
  switchAnalyzeRlist = sex_switchlist,
  reduceToSwitchingGenes = TRUE
)
extractSwitchSummary(sex_switchlist_analyzed)
```

No genes were considered switching, so I will try splitting by brain region.

### Get sex DTU for each brain region

Before analysis, I need to subset the data for each brain region.

```{r split out brain regions}
# make vector of brain regions
all_regions <- c("cerebellum", "cortex", "hippocampus", "striatum")

# split each region
lapply(all_regions, split_region)
```

Now that the data is split out, I can make switchAnalyzeRlist objects for analysis.

```{r make design and switchlists}
# use function on all brain regions
lapply(all_regions, make_switchlist)
```

Once these are made, I can run DEXSeq.

#### Run DEXSeq

```{r DEXSeq by sex and brain region}
# use function on all brain regions
lapply(all_regions, filter_run_dexseq)
```

Alright, now I have some differential used isoforms/genes. Let's visualize them with a volcano plot, although not including background genes in the analysis might be a bit much.

#### Make volcano plots

```{r volcano plots for sex split by brain region}
# use function on all switchlists
create_sex_volcano_plot(cere_sex_switchlist_analyzed)
create_sex_volcano_plot(cort_sex_switchlist_analyzed)
create_sex_volcano_plot(hipp_sex_switchlist_analyzed)
create_sex_volcano_plot(stri_sex_switchlist_analyzed)
```

These volcano plots are very sparse. I'm going to add the argument to include all genes and see if it is overwhelming.

```{r redo DEXSEq split by tissue and sex}
# run new function on all regions
lapply(all_regions, filter_run_dexseq_noreduce)
```

Okay, now time to plot again.

```{r volcano plot again}
# use function on all switchlists
create_sex_volcano_plot_rm(cere_sex_switchlist_analyzed)
create_sex_volcano_plot_rm(cort_sex_switchlist_analyzed)
create_sex_volcano_plot_rm(hipp_sex_switchlist_analyzed)
create_sex_volcano_plot_rm(stri_sex_switchlist_analyzed)
```

This is slightly better, I guess. Now the non-significant genes are included at the bottom. There are very few significant genes, but I'll still make an UpSet plot to compare them.

```{r sex significant isoforms upset plot}
# run new function on all regions
lapply(all_regions, get_sig_isoforms)

# make all into list
all_sex_isoform_list <- list(
  cerebellum = cere_sex_sig_isoforms,
  cortex = cort_sex_sig_isoforms,
  hippocampus = hipp_sex_sig_isoforms,
  striatum = stri_sex_sig_isoforms
)

# make matrix
all_sex_isoform_list_mat <- list_to_matrix(all_sex_isoform_list)

# create combination matrix
all_sex_isoform_list_comb_mat <- make_comb_mat(all_sex_isoform_list_mat)

# make UpSet plot (ComplexHeatmap)
upset_sex_isoform <- UpSet(
  all_sex_isoform_list_comb_mat,
  comb_order = order(-comb_size(all_sex_isoform_list_comb_mat)),
  comb_col = viridis(4)[comb_degree(all_sex_isoform_list_comb_mat)],
  top_annotation = upset_top_annotation(all_sex_isoform_list_comb_mat,
    add_numbers = TRUE
  ),
  right_annotation = upset_right_annotation(all_sex_isoform_list_comb_mat,
    add_numbers = TRUE
  )
)

draw(upset_sex_isoform,
  column_title = "Overlap of sex-DTU isoforms by brain region"
)
```

There is no overlap! An UpSet plot might not be the best visualization if there is no overlap. I find this interesting since that means they are all brain-region specific sex differences. So what are they? Are they enriched for anything?

#### GO Analysis using gprofiler2

Before doing GO analysis, I need to pull out significant genes. GO analysis does not currently work at the isoform level (at least for gprofiler2).

```{r pull out signficant genes}
# run  on all brain regions
lapply(all_regions, get_cut_sig_genes)
```

Now, perform go analysis with gprofiler2.

```{r go analysis}
# run new function on all brain regions
lapply(all_regions, run_gprofiler)
```

It does not look like there is a lot of enrichment here. There are only a few genes so it makes sense to me that there is very little enrichment.

Lastly, I want to get the gene symbols.

```{r get top sex gene symbols}
# run new function on all regions to get symbols for each list
lapply(all_regions, get_gene_symbols_sex)
```

#### Clean up script

```{r tidy script}
style_file("dtu_region_sex.Rmd")
lint("dtu_region_sex.Rmd", linters = linters_with_defaults(
  object_length_linter = NULL,
  object_name_linter = NULL,
  object_usage_linter = NULL
))
```

#### Software versions

```{r versions}
sessionInfo()
```
