---
title: 'DTU Analysis: Brain Region and Sex'
author: "Emma Jones"
date: "2023-02-27"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Differential Transcript Usage (DTU) Analysis across Sex in Mouse Brain lrRNA-Seq Data

#### Load in packages

```{r load packages}
library(tidyverse)
library(IsoformSwitchAnalyzeR)
library(ComplexHeatmap)
library(viridis)
library(org.Mm.eg.db)
library(gprofiler2)
library(styler)
library(lintr)
library(here)
```

#### Load in data

Next, I want to read in data and get cpm values. Since I'll be doing this for all 3 analyses, I made it an R script that can be sourced.

```{r read in data and get cpm values}
source("calculate_cpm.R", local = knitr::knit_global())
```

### Get DTU by sex

Now that I have shown DTU differences across tissue, I am going to look at differences across sex. I'll need to create a new switchAnalyzeRlist object with metadata for sex.

```{r create metadata and create object for sex}
sex_design <- data.frame(
  sampleID = sample_collection_metadata$sample_id,
  condition = sample_collection_metadata$sex
)

sex_switchlist <- importRdata(
  isoformCountMatrix = merged_counts_iso,
  isoformRepExpression = cpm_iso,
  designMatrix = sex_design,
  isoformExonAnnoation = here(
    "data", "nextflow", "results", "bambu",
    "extended_annotations.gtf"
  ),
  isoformNtFasta = here("data", "gffread", "isoform_sequences.fa"),
  showProgress = FALSE
)
```

Now that I have my object, I can run a DTU analysis.

```{r run DEXSeq}
sex_switchlist <- preFilter(sex_switchlist, geneExpressionCutoff = NULL)
# run DEXSeq
sex_switchlist_analyzed <- isoformSwitchTestDEXSeq(
  switchAnalyzeRlist = sex_switchlist,
  reduceToSwitchingGenes = TRUE
)
extractSwitchSummary(sex_switchlist_analyzed)
```

No genes were considered switching, so I will try splitting by brain region.

### Get sex DTU for each brain region

Before analysis, I need to subset the data for each brain region.

```{r split out brain regions}
# make function to split out brain regions
split_region <- function(x) {
  # subset counts
  subset_counts <- subset(
    merged_counts_iso,
    select = c(
      "isoform_id",
      sample_collection_metadata$sample_id[
        sample_collection_metadata$tissue == x
      ]
    )
  )
  # subset cpm
  subset_cpm <- subset(
    cpm_iso,
    select = c(
      "isoform_id",
      sample_collection_metadata$sample_id[
        sample_collection_metadata$tissue == x
      ]
    )
  )
  # drop rows with 0 cpm
  subset_cpm <- subset_cpm[rowSums(subset_cpm[, -1]) != 0, ]
  # get name
  name <- substr(x, 1, 4)
  # assign objects
  assign(paste0(name, "_counts"), subset_counts, envir = .GlobalEnv)
  assign(paste0(name, "_cpm"), subset_cpm, envir = .GlobalEnv)
}

# split each region
split_region("cerebellum")
split_region("cortex")
split_region("hippocampus")
split_region("striatum")
```

Now that the data is split out, I can make switchAnalyzeRlist objects for analysis.

```{r make design and switchlists}
# make function
make_switchlist <- function(x) {
  # get name
  name <- substr(x, 1, 4)
  # get objects for function
  assign("name_counts", get(paste0(name, "_counts")))
  assign("name_cpm", get(paste0(name, "_cpm")))
  # make design
  temp_design <- data.frame(
    sampleID = sample_collection_metadata$sample_id[
      sample_collection_metadata$tissue == x
    ],
    condition = sample_collection_metadata$sex[
      sample_collection_metadata$tissue == x
    ]
  )
  # make switchlist
  temp_switchlist <- importRdata(
    isoformCountMatrix = name_counts,
    isoformRepExpression = name_cpm,
    designMatrix = temp_design,
    isoformExonAnnoation = here(
      "data", "nextflow", "results", "bambu", "extended_annotations.gtf"
    ),
    isoformNtFasta = here("data", "gffread", "isoform_sequences.fa"),
    showProgress = FALSE
  )
  # assign variable
  assign(paste0(name, "_sex_switchlist"), temp_switchlist, envir = .GlobalEnv)
}

# use function on brain regions
make_switchlist("cerebellum")
make_switchlist("cortex")
make_switchlist("hippocampus")
make_switchlist("striatum")
```

Once these are made, I can run DEXSeq.

#### Run DEXSeq

```{r DEXSeq by sex and brain region}
# cerebellum
cere_sex_switchlist <- preFilter(cere_sex_switchlist,
  geneExpressionCutoff = NULL
)
# run DEXSeq
cere_sex_switchlist_analyzed <- isoformSwitchTestDEXSeq(
  switchAnalyzeRlist = cere_sex_switchlist,
  reduceToSwitchingGenes = TRUE
)
extractSwitchSummary(cere_sex_switchlist_analyzed)

# cortex
cort_sex_switchlist <- preFilter(cort_sex_switchlist,
  geneExpressionCutoff = NULL
)
# run DEXSeq
cort_sex_switchlist_analyzed <- isoformSwitchTestDEXSeq(
  switchAnalyzeRlist = cort_sex_switchlist,
  reduceToSwitchingGenes = TRUE
)
extractSwitchSummary(cort_sex_switchlist_analyzed)

# hippocampus
hipp_sex_switchlist <- preFilter(hipp_sex_switchlist,
  geneExpressionCutoff = NULL
)
# run DEXSeq
hipp_sex_switchlist_analyzed <- isoformSwitchTestDEXSeq(
  switchAnalyzeRlist = hipp_sex_switchlist,
  reduceToSwitchingGenes = TRUE
)
extractSwitchSummary(hipp_sex_switchlist_analyzed)

# striatum
stri_sex_switchlist <- preFilter(stri_sex_switchlist,
  geneExpressionCutoff = NULL
)
# run DEXSeq
stri_sex_switchlist_analyzed <- isoformSwitchTestDEXSeq(
  switchAnalyzeRlist = stri_sex_switchlist,
  reduceToSwitchingGenes = TRUE
)
extractSwitchSummary(stri_sex_switchlist_analyzed)
```

Alright, now I have some differential used isoforms/genes. Let's visualize them with a volcano plot, although not including background genes in the analysis might be a bit much.

#### Make volcano plots

```{r volcano plots for sex split by brain region}
# cerebellum
ggplot(
  data = cere_sex_switchlist_analyzed$isoformFeatures,
  aes(x = dIF, y = -log10(isoform_switch_q_value))
) +
  geom_point(aes(color = abs(dIF) > 0.1 & isoform_switch_q_value < 0.05),
    size = 2,
    alpha = 0.5,
    stroke = NA
  ) +
  geom_hline(
    yintercept = -log10(0.05),
    linetype = "dashed",
    color = "magenta",
    linewidth = .7
  ) +
  geom_vline(
    xintercept = c(-0.1, 0.1),
    linetype = "dashed",
    color = "turquoise3",
    linewidth = .7
  ) +
  scale_color_manual("Signficant\nIsoform Switch",
    labels = c("not significant", "significant"),
    values = c("gray40", "limegreen")
  ) +
  labs(
    x = "dIF (Differential Isoform Fraction)",
    y = "-log10 (Isoform Switch q Value)"
  ) +
  theme_light() +
  theme(
    legend.text = element_text(size = 11),
    axis.text = element_text(size = 10)
  ) +
  guides(colour = guide_legend(override.aes = list(size = 4))) +
  ggtitle("Cerebellum Differentially Used Isoforms - by sex")

# cortex
ggplot(
  data = cort_sex_switchlist_analyzed$isoformFeatures,
  aes(x = dIF, y = -log10(isoform_switch_q_value))
) +
  geom_point(aes(color = abs(dIF) > 0.1 & isoform_switch_q_value < 0.05),
    size = 2,
    alpha = 0.5,
    stroke = NA
  ) +
  geom_hline(
    yintercept = -log10(0.05),
    linetype = "dashed",
    color = "magenta",
    linewidth = .7
  ) +
  geom_vline(
    xintercept = c(-0.1, 0.1),
    linetype = "dashed",
    color = "turquoise3",
    linewidth = .7
  ) +
  scale_color_manual("Signficant\nIsoform Switch",
    labels = c("not significant", "significant"),
    values = c("gray40", "limegreen")
  ) +
  labs(
    x = "dIF (Differential Isoform Fraction)",
    y = "-log10 (Isoform Switch q Value)"
  ) +
  theme_light() +
  theme(
    legend.text = element_text(size = 11),
    axis.text = element_text(size = 10)
  ) +
  guides(colour = guide_legend(override.aes = list(size = 4))) +
  ggtitle("Cortex Differentially Used Isoforms - by sex")

# hippocampus
ggplot(
  data = hipp_sex_switchlist_analyzed$isoformFeatures,
  aes(x = dIF, y = -log10(isoform_switch_q_value))
) +
  geom_point(aes(color = abs(dIF) > 0.1 & isoform_switch_q_value < 0.05),
    size = 2,
    alpha = 0.5,
    stroke = NA
  ) +
  geom_hline(
    yintercept = -log10(0.05),
    linetype = "dashed",
    color = "magenta",
    linewidth = .7
  ) +
  geom_vline(
    xintercept = c(-0.1, 0.1),
    linetype = "dashed",
    color = "turquoise3",
    linewidth = .7
  ) +
  scale_color_manual("Signficant\nIsoform Switch",
    labels = c("not significant", "significant"),
    values = c("gray40", "limegreen")
  ) +
  labs(
    x = "dIF (Differential Isoform Fraction)",
    y = "-log10 (Isoform Switch q Value)"
  ) +
  theme_light() +
  theme(
    legend.text = element_text(size = 11),
    axis.text = element_text(size = 10)
  ) +
  guides(colour = guide_legend(override.aes = list(size = 4))) +
  ggtitle("Hippocampus Differentially Used Isoforms - by sex")

# striatum
ggplot(
  data = stri_sex_switchlist_analyzed$isoformFeatures,
  aes(x = dIF, y = -log10(isoform_switch_q_value))
) +
  geom_point(aes(color = abs(dIF) > 0.1 & isoform_switch_q_value < 0.05),
    size = 2,
    alpha = 0.5,
    stroke = NA
  ) +
  geom_hline(
    yintercept = -log10(0.05),
    linetype = "dashed",
    color = "magenta",
    linewidth = .7
  ) +
  geom_vline(
    xintercept = c(-0.1, 0.1),
    linetype = "dashed",
    color = "turquoise3",
    linewidth = .7
  ) +
  scale_color_manual("Signficant\nIsoform Switch",
    labels = c("not significant", "significant"),
    values = c("gray40", "limegreen")
  ) +
  labs(
    x = "dIF (Differential Isoform Fraction)",
    y = "-log10 (Isoform Switch q Value)"
  ) +
  theme_light() +
  theme(
    legend.text = element_text(size = 11),
    axis.text = element_text(size = 10)
  ) +
  guides(colour = guide_legend(override.aes = list(size = 4))) +
  ggtitle("Striatum Differentially Used Isoforms - by sex")
```

Yeah, these volcano plots are awful. I'm going to add the argument to include all genes and see if it is overwhelming.

```{r redo DEXSEq split by tissue and sex}
# cerebellum
# run DEXSeq
cere_sex_switchlist_analyzed <- isoformSwitchTestDEXSeq(
  switchAnalyzeRlist = cere_sex_switchlist,
  reduceToSwitchingGenes = FALSE
)
extractSwitchSummary(cere_sex_switchlist_analyzed)

# cortex
# run DEXSeq
cort_sex_switchlist_analyzed <- isoformSwitchTestDEXSeq(
  switchAnalyzeRlist = cort_sex_switchlist,
  reduceToSwitchingGenes = FALSE
)
extractSwitchSummary(cort_sex_switchlist_analyzed)

# hippocampus
# run DEXSeq
hipp_sex_switchlist_analyzed <- isoformSwitchTestDEXSeq(
  switchAnalyzeRlist = hipp_sex_switchlist,
  reduceToSwitchingGenes = FALSE
)
extractSwitchSummary(hipp_sex_switchlist_analyzed)

# striatum
# run DEXSeq
stri_sex_switchlist_analyzed <- isoformSwitchTestDEXSeq(
  switchAnalyzeRlist = stri_sex_switchlist,
  reduceToSwitchingGenes = FALSE
)
extractSwitchSummary(stri_sex_switchlist_analyzed)
```

Okay, now time to plot again.

```{r volcano plot again}
# cerebellum
ggplot(
  data = cere_sex_switchlist_analyzed$isoformFeatures,
  aes(x = dIF, y = -log10(isoform_switch_q_value))
) +
  geom_point(aes(color = abs(dIF) > 0.1 & isoform_switch_q_value < 0.05),
    size = 2,
    alpha = 0.5,
    stroke = NA
  ) +
  geom_hline(
    yintercept = -log10(0.05),
    linetype = "dashed",
    color = "magenta",
    linewidth = .7
  ) +
  geom_vline(
    xintercept = c(-0.1, 0.1),
    linetype = "dashed",
    color = "turquoise3",
    linewidth = .7
  ) +
  scale_color_manual("Signficant\nIsoform Switch",
    labels = c("not significant", "significant"),
    values = c("gray40", "limegreen")
  ) +
  labs(
    x = "dIF (Differential Isoform Fraction)",
    y = "-log10 (Isoform Switch q Value)"
  ) +
  theme_light() +
  theme(
    legend.text = element_text(size = 11),
    axis.text = element_text(size = 10)
  ) +
  guides(colour = guide_legend(override.aes = list(size = 4))) +
  ggtitle("Cerebellum Differentially Used Isoforms - by sex")

# cortex
ggplot(
  data = cort_sex_switchlist_analyzed$isoformFeatures,
  aes(x = dIF, y = -log10(isoform_switch_q_value))
) +
  geom_point(aes(color = abs(dIF) > 0.1 & isoform_switch_q_value < 0.05),
    size = 2,
    alpha = 0.5,
    stroke = NA
  ) +
  geom_hline(
    yintercept = -log10(0.05),
    linetype = "dashed",
    color = "magenta",
    linewidth = .7
  ) +
  geom_vline(
    xintercept = c(-0.1, 0.1),
    linetype = "dashed",
    color = "turquoise3",
    linewidth = .7
  ) +
  scale_color_manual("Signficant\nIsoform Switch",
    labels = c("not significant", "significant"),
    values = c("gray40", "limegreen")
  ) +
  labs(
    x = "dIF (Differential Isoform Fraction)",
    y = "-log10 (Isoform Switch q Value)"
  ) +
  theme_light() +
  theme(
    legend.text = element_text(size = 11),
    axis.text = element_text(size = 10)
  ) +
  guides(colour = guide_legend(override.aes = list(size = 4))) +
  ggtitle("Cortex Differentially Used Isoforms - by sex")

# hippocampus
ggplot(
  data = hipp_sex_switchlist_analyzed$isoformFeatures,
  aes(x = dIF, y = -log10(isoform_switch_q_value))
) +
  geom_point(aes(color = abs(dIF) > 0.1 & isoform_switch_q_value < 0.05),
    size = 2,
    alpha = 0.5,
    stroke = NA
  ) +
  geom_hline(
    yintercept = -log10(0.05),
    linetype = "dashed",
    color = "magenta",
    linewidth = .7
  ) +
  geom_vline(
    xintercept = c(-0.1, 0.1),
    linetype = "dashed",
    color = "turquoise3",
    linewidth = .7
  ) +
  scale_color_manual("Signficant\nIsoform Switch",
    labels = c("not significant", "significant"),
    values = c("gray40", "limegreen")
  ) +
  labs(
    x = "dIF (Differential Isoform Fraction)",
    y = "-log10 (Isoform Switch q Value)"
  ) +
  theme_light() +
  theme(
    legend.text = element_text(size = 11),
    axis.text = element_text(size = 10)
  ) +
  guides(colour = guide_legend(override.aes = list(size = 4))) +
  ggtitle("Hippocampus Differentially Used Isoforms - by sex")

# striatum
ggplot(
  data = stri_sex_switchlist_analyzed$isoformFeatures,
  aes(x = dIF, y = -log10(isoform_switch_q_value))
) +
  geom_point(aes(color = abs(dIF) > 0.1 & isoform_switch_q_value < 0.05),
    size = 2,
    alpha = 0.5,
    stroke = NA
  ) +
  geom_hline(
    yintercept = -log10(0.05),
    linetype = "dashed",
    color = "magenta",
    linewidth = .7
  ) +
  geom_vline(
    xintercept = c(-0.1, 0.1),
    linetype = "dashed",
    color = "turquoise3",
    linewidth = .7
  ) +
  scale_color_manual("Signficant\nIsoform Switch",
    labels = c("not significant", "significant"),
    values = c("gray40", "limegreen")
  ) +
  labs(
    x = "dIF (Differential Isoform Fraction)",
    y = "-log10 (Isoform Switch q Value)"
  ) +
  theme_light() +
  theme(
    legend.text = element_text(size = 11),
    axis.text = element_text(size = 10)
  ) +
  guides(colour = guide_legend(override.aes = list(size = 4))) +
  ggtitle("Striatum Differentially Used Isoforms - by sex")
```

This is slightly better, I guess. There are very few genes, but I'll still make an UpSet plot to compare them.

```{r sex upset}
# cerebellum
cere_sex_sig_isoforms <- unique(
  dplyr::filter(
    cere_sex_switchlist_analyzed$isoformFeatures,
    abs(dIF) > 0.1 &
      isoform_switch_q_value < 0.05
  )$isoform_id
)

# cortex
cort_sex_sig_isoforms <- unique(
  dplyr::filter(
    cort_sex_switchlist_analyzed$isoformFeatures,
    abs(dIF) > 0.1 & isoform_switch_q_value < 0.05
  )$isoform_id
)

# hippocampus isoforms
hipp_sex_sig_isoforms <- unique(
  dplyr::filter(
    hipp_sex_switchlist_analyzed$isoformFeatures,
    abs(dIF) > 0.1 & isoform_switch_q_value < 0.05
  )$isoform_id
)

# striatum isoforms
stri_sex_sig_isoforms <- unique(
  dplyr::filter(
    stri_sex_switchlist_analyzed$isoformFeatures,
    abs(dIF) > 0.1 & isoform_switch_q_value < 0.05
  )$isoform_id
)

# make all into list
all_sex_isoform_list <- list(
  cerebellum = cere_sex_sig_isoforms,
  cortex = cort_sex_sig_isoforms,
  hippocampus = hipp_sex_sig_isoforms,
  striatum = stri_sex_sig_isoforms
)

# make matrix
all_sex_isoform_list_mat <- list_to_matrix(all_sex_isoform_list)

# create combination matrix
all_sex_isoform_list_comb_mat <- make_comb_mat(all_sex_isoform_list_mat)

# make UpSet plot (ComplexHeatmap)
upset_sex_isoform <- UpSet(
  all_sex_isoform_list_comb_mat,
  comb_order = order(-comb_size(all_sex_isoform_list_comb_mat)),
  comb_col = viridis(4)[comb_degree(all_sex_isoform_list_comb_mat)],
  top_annotation = upset_top_annotation(all_sex_isoform_list_comb_mat,
    add_numbers = TRUE
  ),
  right_annotation = upset_right_annotation(all_sex_isoform_list_comb_mat,
    add_numbers = TRUE
  )
)

draw(upset_sex_isoform,
  column_title = "Overlap of sex-DTU isoforms by brain region"
)
```

There is no overlap! I find this interesting since that means they are brain-region specific sex differences. So what are they? Are they enriched for anything?

#### GO Analysis using gprofiler2

```{r go analysis}
# cerebellum
cere_sex_sig_genes <- unique(
  dplyr::filter(
    cere_sex_switchlist_analyzed$isoformFeatures,
    abs(dIF) > 0.1 & isoform_switch_q_value < 0.05
  )$gene_id
)

cere_sex_sig_genes <- str_extract(cere_sex_sig_genes, "ENSMUSG...........")

cere_sexgostres <- gost(
  query = cere_sex_sig_genes,
  organism = "mmusculus", ordered_query = FALSE,
  multi_query = FALSE, significant = TRUE, exclude_iea = FALSE,
  measure_underrepresentation = FALSE, evcodes = FALSE,
  user_threshold = 0.05, correction_method = "g_SCS",
  domain_scope = "custom",
  custom_bg = str_extract(
    unique(cere_sex_switchlist_analyzed$isoformFeatures$gene_id),
    pattern = "ENSMUSG..........."
  ),
  numeric_ns = "", sources = NULL, as_short_link = FALSE
)

gostplot(cere_sexgostres, capped = TRUE, interactive = FALSE)

# cortex
cort_sex_sig_genes <- unique(
  dplyr::filter(
    cort_sex_switchlist_analyzed$isoformFeatures,
    abs(dIF) > 0.1 & isoform_switch_q_value < 0.05
  )$gene_id
)

cort_sex_sig_genes <- str_extract(cort_sex_sig_genes, "ENSMUSG...........")

cort_sexgostres <- gost(
  organism = "mmusculus", ordered_query = FALSE,
  multi_query = FALSE, significant = TRUE, exclude_iea = FALSE,
  measure_underrepresentation = FALSE, evcodes = FALSE,
  user_threshold = 0.05, correction_method = "g_SCS",
  domain_scope = "custom",
  custom_bg = str_extract(
    unique(cere_sex_switchlist_analyzed$isoformFeatures$gene_id),
    pattern = "ENSMUSG..........."
  ),
  numeric_ns = "", sources = NULL, as_short_link = FALSE
)

gostplot(cort_sexgostres, capped = TRUE, interactive = FALSE)

# hippocampus
hipp_sex_sig_genes <- unique(
  dplyr::filter(
    hipp_sex_switchlist_analyzed$isoformFeatures,
    abs(dIF) > 0.1 & isoform_switch_q_value < 0.05
  )$gene_id
)

hipp_sex_sig_genes <- str_extract(hipp_sex_sig_genes, "ENSMUSG...........")

hipp_sexgostres <- gost(
  query = hipp_sex_sig_genes,
  organism = "mmusculus", ordered_query = FALSE,
  multi_query = FALSE, significant = TRUE, exclude_iea = FALSE,
  measure_underrepresentation = FALSE, evcodes = FALSE,
  user_threshold = 0.05, correction_method = "g_SCS",
  domain_scope = "custom",
  custom_bg = str_extract(
    unique(cere_sex_switchlist_analyzed$isoformFeatures$gene_id),
    pattern = "ENSMUSG..........."
  ),
  numeric_ns = "", sources = NULL, as_short_link = FALSE
)

# not significant

# striatum
stri_sex_sig_genes <- unique(
  dplyr::filter(
    stri_sex_switchlist_analyzed$isoformFeatures,
    abs(dIF) > 0.1 & isoform_switch_q_value < 0.05
  )$gene_id
)

stri_sex_sig_genes <- str_extract(stri_sex_sig_genes, "ENSMUSG...........")

stri_sexgostres <- gost(
  query = stri_sex_sig_genes,
  organism = "mmusculus", ordered_query = FALSE,
  multi_query = FALSE, significant = TRUE, exclude_iea = FALSE,
  measure_underrepresentation = FALSE, evcodes = FALSE,
  user_threshold = 0.05, correction_method = "g_SCS",
  domain_scope = "custom",
  custom_bg = str_extract(
    unique(cere_sex_switchlist_analyzed$isoformFeatures$gene_id),
    pattern = "ENSMUSG..........."
  ),
  numeric_ns = "", sources = NULL, as_short_link = FALSE
)

# not significant
```

It does not look like there is a lot of enrichment here. There are only a few genes so it makes sense to me that there is very little enrichment.

Now, I want to get the gene symbols. Honestly, I should have done this way before in the analysis. I should consider doing that later.

```{r get top sex gene symbols}
# get symbols for each list
# cerebellum
cere_sex_sig_genes
cere_sex_sig_genes_df <- AnnotationDbi::select(
  org.Mm.eg.db,
  keys = cere_sex_sig_genes,
  columns = c("SYMBOL", "GENENAME", "ENSEMBL"), keytype = "ENSEMBL"
)

# cortex
cort_sex_sig_genes
cort_sex_sig_genes_df <- AnnotationDbi::select(
  org.Mm.eg.db,
  keys = cort_sex_sig_genes,
  columns = c("SYMBOL", "GENENAME", "ENSEMBL"), keytype = "ENSEMBL"
)

# hippocampus
hipp_sex_sig_genes
hipp_sex_sig_genes_df <- AnnotationDbi::select(
  org.Mm.eg.db,
  keys = hipp_sex_sig_genes,
  columns = c("SYMBOL", "GENENAME", "ENSEMBL"), keytype = "ENSEMBL"
)

# striatum
stri_sex_sig_genes
stri_sex_sig_genes_df <- AnnotationDbi::select(
  org.Mm.eg.db,
  keys = stri_sex_sig_genes,
  columns = c("SYMBOL", "GENENAME", "ENSEMBL"), keytype = "ENSEMBL"
)
```

#### Clean up script

```{r tidy script}
style_file("dtu_region_sex.Rmd")
lint("dtu_region_sex.Rmd")
```

#### Software versions

```{r versions}
sessionInfo()
```
