---
title: "Neurological Disease Gene Enrichment"
author: "Emma Jones"
date: "2023-03-14"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Differential Transcript Usage (DTU) Analysis across Sex in Mouse Brain lrRNA-Seq Data

#### Load in packages

```{r load packages}
library(tidyverse)
library(IsoformSwitchAnalyzeR)
library(viridis)
library(styler)
library(lintr)
library(here)
library(openxlsx)
library(org.Hs.eg.db)
```

#### Load in data

Load in switchlist object. it needs to have the orf added, if not already.

```{r load in switchlist objects}
# load object comparing all regions to each other
region_region_switchlist_analyzed <- readRDS(
  here("data","switchlist_objects","region_region_switchlist_analyzed.Rds")
  )

add_save_orfs("region_region")

region_region_switchlist_analyzed <- analyzePFAM(
  switchAnalyzeRlist = region_region_switchlist_analyzed,
  pathToPFAMresultFile = here("data", "Pfam", "all_regions", "pfamscan_out.txt")
)
```

Load in AD gene set

```{r read AD gene set}
# read in from web
AD_genes <- read.xlsx("https://static-content.springer.com/esm/art%3A10.1038%2Fs41588-022-01024-z/MediaObjects/41588_2022_1024_MOESM4_ESM.xlsx", sheet = 6, rows = 5:87, cols = 4, colNames = FALSE)

colnames(AD_genes) <- "symbol"

AD_genes <- unique(AD_genes)

AD_ensembl <- AnnotationDbi::select(org.Hs.eg.db,
    keys = AD_genes$symbol,
    columns = c("SYMBOL", "ENSEMBL"), keytype = "SYMBOL"
  )

AD_ensembl <- AD_ensembl[!is.na(AD_ensembl$ENSEMBL),]

copy <- AD_ensembl[,2]
```

Load in mood disorders gene set. This table does not have individual genes split out, so I need to do that.

```{r read in mood disorders gene sets}
# read in from web
mood_genes <- read.xlsx("https://ars.els-cdn.com/content/image/1-s2.0-S0092867419312760-mmc4.xlsx", sheet = 1, rows = 4:149, cols = 8, colNames = FALSE)

colnames(mood_genes) <- "symbol"

new_col <- mood_genes %>% separate_wider_delim(mood, ";", names = paste0("gene", 1:36), too_many = "debug", too_few = "debug")

vals <- unlist(new_col[,1:36], use.names = FALSE)

vals <- vals[!is.na(vals)]

vals <- vals[!vals == "" & !vals == "-"]

vals <- gsub("\\s*\\([^\\)]+\\)", "", vals)

mood_genes$symbol <- vals
```

#### Visualize!

```{r visualize}

mouse_AD <- convertHumanGeneList(copy)

switchPlot(region_region_switchlist_analyzed, gene = 'ENSMUSG00000068747')
```

#### Clean up script

```{r tidy script}
style_file("dtu_neuro_disease.Rmd")
lint("dtu_neuro_disease.Rmd", linters = linters_with_defaults(
  object_length_linter = NULL,
  object_name_linter = NULL,
  object_usage_linter = NULL
))
```

#### Software versions

```{r versions}
sessionInfo()
```
