---
title: 'DTU Analysis: Pairwise Brain Regions'
author: "Emma Jones"
date: "2023-02-23"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


## Differential Transcript Usage (DTU) Analysis in Mouse Brain lrRNA-Seq Data

#### Load in packages

```{r load packages}
library(tidyverse)
library(IsoformSwitchAnalyzeR)
library(ComplexHeatmap)
library(viridis)
library(org.Mm.eg.db)
library(gprofiler2)
library(styler)
library(lintr)
library(here)
```

#### Load in data

Next, I want to read in data and get tpm values. Since I'll be doing this for all 3 analyses, I made it an R script that can be sourced.

```{r read in data and get tpm values}
source("calculate_cpm.R", local = knitr::knit_global())
```

## Begin DTU analysis for pairwise brain region comparisons

The package requires a design matrix in specific format.

```{r make design matrix}
exp_design <- data.frame(
  sampleID = sample_collection_metadata$sample_id,
  condition = sample_collection_metadata$tissue
)
```

Make a switchlist, which is an object required for running isoformswitchanalyzer.

```{r add fasta of transcripts to object}
cpm_switchlist <- importRdata(
  isoformCountMatrix = merged_counts_iso,
  isoformRepExpression = cpm_iso,
  designMatrix = exp_design,
  isoformExonAnnoation = here(
    "data", "nextflow", "results", "bambu",
    "extended_annotations.gtf"
  ),
  isoformNtFasta = here("data", "gffread", "isoform_sequences.fa"),
  showProgress = FALSE
)
```

Now, I can run DEXSeq. Because my data includes counts of less than 1, I still want to include them in my analysis.

```{r run dexseq}
# filter data and disabling gene expression cutoff
cpm_switchlist <- preFilter(cpm_switchlist, geneExpressionCutoff = NULL)
# now run DEXSeq
cpm_switchlist_dexseq <- isoformSwitchTestDEXSeq(
  switchAnalyzeRlist = cpm_switchlist,
  reduceToSwitchingGenes = TRUE
)

extractSwitchSummary(cpm_switchlist_dexseq)
```

### Make UpSet Plots

The goal of these plots is to show overlap between significantly differentially used isoforms across conditions.
First, I need to get significant genes and split them out by brain region.

```{r make upset plots for overlap by brain region}
# first, get significant isoform switches
sig_isoform_features <- cpm_switchlist_dexseq$isoformFeatures
sig_isoform_features <- dplyr::filter(
  sig_isoform_features, abs(dIF) > 0.1 & isoform_switch_q_value < 0.05
)

# pull just genes
sig_isoform_genes <- sig_isoform_features %>%
  distinct(gene_id, condition_1, condition_2, .keep_all = TRUE)

# make function for filtering genes
filter_genes <- function(x) {
  # subset genes of interest
  sig_genes_subset <- filter(
    sig_isoform_genes,
    condition_1 == x | condition_2 == x
  )
  sig_genes_subset <- unique(sig_genes_subset$gene_id)
  # name object
  name <- substr(x, 1, 4)
  assign(paste0("sig_isoform_genes_", name),
    sig_genes_subset,
    envir = .GlobalEnv
  )
}

filter_genes("cerebellum")
filter_genes("cortex")
filter_genes("hippocampus")
filter_genes("striatum")

# make into list
sig_isoform_gene_list <- list(
  cerebellum = sig_isoform_genes_cere,
  cortex = sig_isoform_genes_cort,
  hippocampus = sig_isoform_genes_hipp,
  striatum = sig_isoform_genes_stri
)
```

I also can indicate not just overlapping, but specifically which genes are DTU and have overlap?

I want to get a list of genes for each comparison.

```{r get list of genes for each comparison}
filter_genes_comp <- function(x, y) {
  # subset genes of interest
  sig_genes_subset <- filter(
    sig_isoform_genes,
    condition_1 == x & condition_2 == y
  )
  sig_genes_subset <- unique(sig_genes_subset$gene_id)
  # name object
  name_1 <- substr(x, 1, 4)
  name_2 <- substr(y, 1, 4)
  assign(paste0("sig_isoform_genes_", name_1, "_", name_2),
    sig_genes_subset,
    envir = .GlobalEnv
  )
}

filter_genes_comp("cerebellum", "cortex")
filter_genes_comp("cerebellum", "hippocampus")
filter_genes_comp("cerebellum", "striatum")
filter_genes_comp("cortex", "hippocampus")
filter_genes_comp("cortex", "striatum")
filter_genes_comp("hippocampus", "striatum")
```

Once I have these, I can combine them into a list for plotting.

```{r setup overlap of comparisons}
# make into list
sig_overlap_list <- list(
  cere_cort = sig_isoform_genes_cere_cort,
  cere_hipp = sig_isoform_genes_cere_hipp,
  cere_stri = sig_isoform_genes_cere_stri,
  cort_hipp = sig_isoform_genes_cort_hipp,
  cort_stri = sig_isoform_genes_cort_stri,
  hipp_stri = sig_isoform_genes_hipp_stri
)
```

Now that the data is prepped, I can plot a complexheatmap UpSet plot. Here, color indicates degree of overlap being examined.

```{r make complexheatmap UpSet plot for all comparisons}
# make UpSet plot (ComplexHeatmap)
# create matrix
sig_overlap_list_mat <- list_to_matrix(sig_overlap_list)

# generate combination matrix
sig_overlap_list_comb_mat <- make_comb_mat(sig_overlap_list_mat)

# make UpSet plot (ComplexHeatmap)
UpSet(sig_overlap_list_comb_mat,
  comb_order = order(-comb_size(sig_overlap_list_comb_mat)),
  comb_col = viridis(6)[comb_degree(sig_overlap_list_comb_mat)],
  top_annotation = upset_top_annotation(sig_overlap_list_comb_mat,
    add_numbers = TRUE
  ),
  right_annotation = upset_right_annotation(sig_overlap_list_comb_mat,
    add_numbers = TRUE
  )
)
```

I think this is the most informative use of an UpSet plot. There are a lot of overlap that only have 1 gene in common, so maybe later I want to only include the comparisons with the most overlap (like top 10).

### Volcano Plots

Making volcano plots is a classic way to show the relationship between fold changes and p values. I will make some here with the differential isoform fraction, instead of the more common differential expression (log2foldchange) seen in DE experiments.

```{r make function for volcano plot}
# create function
create_volcano_plot <- function(x, y) {
  # create subset
  analyzed_subset <- dplyr::filter(
    cpm_switchlist_dexseq$isoformFeatures,
    condition_1 == x & condition_2 == y
  )
  # plot
  volcano <- ggplot(
    data = analyzed_subset,
    aes(x = dIF, y = -log10(isoform_switch_q_value))
  ) +
    geom_point(aes(color = abs(dIF) > 0.1 & isoform_switch_q_value < 0.05),
      size = 2,
      alpha = 0.5,
      stroke = NA
    ) +
    geom_hline(
      yintercept = -log10(0.05),
      linetype = "dashed",
      color = "magenta",
      linewidth = .7
    ) +
    geom_vline(
      xintercept = c(-0.1, 0.1),
      linetype = "dashed",
      color = "turquoise3",
      linewidth = .7
    ) +
    scale_color_manual("Signficant\nIsoform Switch",
      labels = c("not significant", "significant"),
      values = c("gray40", "limegreen")
    ) +
    labs(
      x = "dIF (Differential Isoform Fraction)",
      y = "-log10 (Isoform Switch q Value)"
    ) +
    theme_light() +
    theme(
      legend.text = element_text(size = 11),
      axis.text = element_text(size = 10)
    ) +
    guides(colour = guide_legend(override.aes = list(size = 4))) +
    ggtitle(paste0(
      "cerebellum", " vs. ", "cortex",
      " differentially used isoforms"
    ))
  # save
  ggsave(
    paste0(
      here("results", "plots", "DEXSeq_volcano"),
      "/", x, "_", y, "_volcano.png"
    ),
    plot = volcano, width = 6, height = 4
  )
}

# now plot for each comparison
create_volcano_plot("cerebellum", "cortex")
create_volcano_plot("cerebellum", "hippocampus")
create_volcano_plot("cerebellum", "striatum")
create_volcano_plot("cortex", "hippocampus")
create_volcano_plot("cortex", "striatum")
create_volcano_plot("hippocampus", "striatum")
```

## GO Anlaysis of Gene Sets

The nest step of doing DTU is a GO analysis with my gene sets. I am going to use the package gprofiler2. I guess I want to do GO for every comparison list and then see if they have similar processes.

I need to get the correct number of background genes, so this is all genes that were tested for significance, whether significant or not.

```{r get correct list of background genes}
bg_genes <- unique(cpm_switchlist$isoformFeatures$gene_id)
bg_genes <- str_extract(bg_genes, "ENSMUSG...........")
```

Now I can run gprofiler.

```{r gprofiler2}
# create function
run_plot_gprofiler <- function(x) {
  # get names
  name <- deparse(substitute(x))
  name <- substr(name, nchar(name) - 9 + 1, nchar(name))
  # remove numbers after decimal point
  genes_remove_decimal <- str_extract(
    x, "ENSMUSG..........."
  )
  # run gprofiler2
  gostres <- gost(
    query = genes_remove_decimal,
    organism = "mmusculus", ordered_query = FALSE,
    multi_query = FALSE, significant = TRUE, exclude_iea = FALSE,
    measure_underrepresentation = FALSE, evcodes = FALSE,
    user_threshold = 0.05, correction_method = "g_SCS",
    domain_scope = "custom", custom_bg = bg_genes,
    numeric_ns = "", sources = NULL, as_short_link = FALSE
  )
  # plot
  gostplot(gostres, capped = TRUE, interactive = FALSE)
  # save
  ggsave(paste0(
    here("results", "plots", "gprofiler2"),
    "/", name, "_gostres.png"
  ), width = 6, height = 4)
}

run_plot_gprofiler(sig_isoform_genes_cere_cort)
run_plot_gprofiler(sig_isoform_genes_cere_hipp)
run_plot_gprofiler(sig_isoform_genes_cere_stri)
run_plot_gprofiler(sig_isoform_genes_cort_hipp) # not significant
run_plot_gprofiler(sig_isoform_genes_cort_stri)
run_plot_gprofiler(sig_isoform_genes_hipp_stri)
```

There are not a lot of results, and most things are not significant. This amount of GO is not really seeming beneficial to me, even with hundreds of DTU genes. 

### Final thoughts

Overall, in this analysis I show that the largest differences in DTU are across cerebellum and the other brain regions.

#### Clean up script

```{r tidy script}
style_file("dtu_region_region.Rmd")
lint("dtu_region_region.Rmd",
  linters = linters_with_defaults(
    object_length_linter = NULL,
    object_name_linter = NULL,
    object_usage_linter = NULL
  )
)
```

#### Software versions

```{r versions}
sessionInfo()
```
