---
title: 'DTU Analysis: Brain Region and Sex'
author: "Emma Jones"
date: "2023-02-27"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Differential Transcript Usage (DTU) Analysis across Sex in Mouse Brain lrRNA-Seq Data

#### Load in packages

```{r load packages}
library(tidyverse)
library(IsoformSwitchAnalyzeR)
library(ComplexHeatmap)
library(viridis)
library(org.Mm.eg.db)
library(gprofiler2)
library(styler)
library(lintr)
library(here)
```

#### Load in data

Next, I want to read in data and get cpm values. Since I'll be doing this for all 3 analyses, I made it an R script that can be sourced.

```{r read in data and get cpm values}
source("calculate_cpm.R", local = knitr::knit_global())
```

### Get DTU by sex

Now that I have shown DTU differences across tissue, I am going to look at differences across sex. I'll need to create a new switchAnalyzeRlist object with metadata for sex.

```{r create metadata and create object for sex}
# make sex design
sex_design <- data.frame(
  sampleID = sample_collection_metadata$sample_id,
  condition = sample_collection_metadata$sex
)
# make switchlist
sex_switchlist <- importRdata(
  isoformCountMatrix = merged_counts_iso,
  isoformRepExpression = cpm_iso,
  designMatrix = sex_design,
  isoformExonAnnoation = here(
    "data", "nextflow", "results", "bambu",
    "extended_annotations.gtf"
  ),
  isoformNtFasta = here("data", "gffread", "isoform_sequences.fa"),
  showProgress = FALSE
)
```

Now that I have my object, I can run a DTU analysis.

```{r run DEXSeq}
# filter data
sex_switchlist <- preFilter(sex_switchlist, geneExpressionCutoff = NULL)
# run DEXSeq
sex_switchlist_analyzed <- isoformSwitchTestDEXSeq(
  switchAnalyzeRlist = sex_switchlist,
  reduceToSwitchingGenes = TRUE
)
extractSwitchSummary(sex_switchlist_analyzed)
```

No genes were considered switching, so I will try splitting by brain region.

### Get sex DTU for each brain region

Before analysis, I need to subset the data for each brain region.

```{r split out brain regions}
# make vector of brain regions
all_regions <- c("cerebellum", "cortex", "hippocampus", "striatum")
# make function to split out brain regions
split_region <- function(x) {
  # subset counts
  subset_counts <- subset(
    merged_counts_iso,
    select = c(
      "isoform_id",
      sample_collection_metadata$sample_id[
        sample_collection_metadata$tissue == x
      ]
    )
  )
  # subset cpm
  subset_cpm <- subset(
    cpm_iso,
    select = c(
      "isoform_id",
      sample_collection_metadata$sample_id[
        sample_collection_metadata$tissue == x
      ]
    )
  )
  # drop rows with 0 cpm
  subset_cpm <- subset_cpm[rowSums(subset_cpm[, -1]) != 0, ]
  # get name
  name <- substr(x, 1, 4)
  # assign objects
  assign(paste0(name, "_counts"), subset_counts, envir = .GlobalEnv)
  assign(paste0(name, "_cpm"), subset_cpm, envir = .GlobalEnv)
}

# split each region
lapply(all_regions, split_region)
```

Now that the data is split out, I can make switchAnalyzeRlist objects for analysis.

```{r make design and switchlists}
# make function
make_switchlist <- function(x) {
  # get name
  name <- substr(x, 1, 4)
  # get objects for function
  assign("name_counts", get(paste0(name, "_counts")))
  assign("name_cpm", get(paste0(name, "_cpm")))
  # make design
  temp_design <- data.frame(
    sampleID = sample_collection_metadata$sample_id[
      sample_collection_metadata$tissue == x
    ],
    condition = sample_collection_metadata$sex[
      sample_collection_metadata$tissue == x
    ]
  )
  # make switchlist
  temp_switchlist <- importRdata(
    isoformCountMatrix = name_counts,
    isoformRepExpression = name_cpm,
    designMatrix = temp_design,
    isoformExonAnnoation = here(
      "data", "nextflow", "results", "bambu", "extended_annotations.gtf"
    ),
    isoformNtFasta = here("data", "gffread", "isoform_sequences.fa"),
    showProgress = FALSE
  )
  # assign variable
  assign(paste0(name, "_sex_switchlist"), temp_switchlist, envir = .GlobalEnv)
}

# use function on all brain regions
lapply(all_regions, make_switchlist)
```

Once these are made, I can run DEXSeq.

#### Run DEXSeq

```{r DEXSeq by sex and brain region}
# make function to filter and run DEXSeq
filter_run_dexseq <- function(x) {
  # get name
  name <- substr(x, 1, 4)
  # assign name
  assign("temp_switchlist", get(paste0(name, "_sex_switchlist")))
  # filter switchlist
  temp_switchlist <- preFilter(temp_switchlist, geneExpressionCutoff = NULL)
  # run DEXSeq
  switchlist_analyzed <- isoformSwitchTestDEXSeq(
    switchAnalyzeRlist = temp_switchlist,
    reduceToSwitchingGenes = TRUE
  )
  # rename object
  assign(paste0(name, "_sex_switchlist_analyzed"),
    switchlist_analyzed,
    envir = .GlobalEnv
  )
}
# use function on all brain regions
lapply(all_regions, filter_run_dexseq)
```

Alright, now I have some differential used isoforms/genes. Let's visualize them with a volcano plot, although not including background genes in the analysis might be a bit much.

#### Make volcano plots

```{r volcano plots for sex split by brain region}
# create function
create_sex_volcano_plot <- function(x) {
  # get name
  name <- deparse(substitute(x))
  name <- substr(name, 1, 4)
  # plot
  volcano_plot <- ggplot(
    data = x$isoformFeatures,
    aes(x = dIF, y = -log10(isoform_switch_q_value))
  ) +
    geom_point(aes(color = abs(dIF) > 0.1 & isoform_switch_q_value < 0.05),
      size = 2,
      alpha = 0.5,
      stroke = NA
    ) +
    geom_hline(
      yintercept = -log10(0.05),
      linetype = "dashed",
      color = "magenta",
      linewidth = .7
    ) +
    geom_vline(
      xintercept = c(-0.1, 0.1),
      linetype = "dashed",
      color = "turquoise3",
      linewidth = .7
    ) +
    scale_color_manual("Signficant\nIsoform Switch",
      labels = c("not significant", "significant"),
      values = c("gray40", "limegreen")
    ) +
    labs(
      x = "dIF (Differential Isoform Fraction)",
      y = "-log10 (Isoform Switch q Value)"
    ) +
    theme_light() +
    theme(
      legend.text = element_text(size = 11),
      axis.text = element_text(size = 10)
    ) +
    guides(colour = guide_legend(override.aes = list(size = 4))) +
    ggtitle(paste0(name, " sex differentially used isoforms"))
  # save
  ggsave(
    paste0(
      here("results", "plots", "DEXSeq_volcano"),
      "/", name, "_sex_volcano.png"
    ),
    volcano_plot,
    width = 6, height = 4
  )
}
# use function on all switchlists
create_sex_volcano_plot(cere_sex_switchlist_analyzed)
create_sex_volcano_plot(cort_sex_switchlist_analyzed)
create_sex_volcano_plot(hipp_sex_switchlist_analyzed)
create_sex_volcano_plot(stri_sex_switchlist_analyzed)
```

These volcano plots are very sparse. I'm going to add the argument to include all genes and see if it is overwhelming.

```{r redo DEXSEq split by tissue and sex}
# make different function with reducing argument
filter_run_dexseq_noreduce <- function(x) {
  # get name
  name <- substr(x, 1, 4)
  # assign name
  assign("temp_switchlist", get(paste0(name, "_sex_switchlist")))
  # filter switchlist
  temp_switchlist <- preFilter(temp_switchlist, geneExpressionCutoff = NULL)
  # run DEXSeq
  switchlist_analyzed <- isoformSwitchTestDEXSeq(
    switchAnalyzeRlist = temp_switchlist,
    reduceToSwitchingGenes = FALSE
  )
  # rename object
  assign(paste0(name, "_sex_switchlist_analyzed"),
    switchlist_analyzed,
    envir = .GlobalEnv
  )
}
# run new function on all regions
lapply(all_regions, filter_run_dexseq_noreduce)
```

Okay, now time to plot again.

```{r volcano plot again}
# make volcano plot function, but remove NAs
create_sex_volcano_plot <- function(x) {
  # get name
  name <- deparse(substitute(x))
  name <- substr(name, 1, 4)
  # plot
  volcano_plot <- ggplot(
    data = x$isoformFeatures,
    aes(x = dIF, y = -log10(isoform_switch_q_value))
  ) +
    geom_point(aes(color = abs(dIF) > 0.1 & isoform_switch_q_value < 0.05),
      size = 2,
      alpha = 0.5,
      stroke = NA
    ) +
    geom_hline(
      yintercept = -log10(0.05),
      linetype = "dashed",
      color = "magenta",
      linewidth = .7
    ) +
    geom_vline(
      xintercept = c(-0.1, 0.1),
      linetype = "dashed",
      color = "turquoise3",
      linewidth = .7
    ) +
    scale_color_manual("Signficant\nIsoform Switch",
      labels = c("not significant", "significant"),
      values = c("gray40", "limegreen"),
      na.translate = FALSE
    ) +
    labs(
      x = "dIF (Differential Isoform Fraction)",
      y = "-log10 (Isoform Switch q Value)"
    ) +
    theme_light() +
    theme(
      legend.text = element_text(size = 11),
      axis.text = element_text(size = 10)
    ) +
    guides(colour = guide_legend(override.aes = list(size = 4))) +
    ggtitle(paste0(name, " sex differentially used isoforms"))
  # save
  ggsave(
    paste0(
      here("results", "plots", "DEXSeq_volcano"),
      "/", name, "_sex_volcano.png"
    ),
    volcano_plot,
    width = 6, height = 4
  )
}
# use function on all switchlists
create_sex_volcano_plot(cere_sex_switchlist_analyzed)
create_sex_volcano_plot(cort_sex_switchlist_analyzed)
create_sex_volcano_plot(hipp_sex_switchlist_analyzed)
create_sex_volcano_plot(stri_sex_switchlist_analyzed)
```

This is slightly better, I guess. Now the non-significant genes are included at the bottom. There are very few significant genes, but I'll still make an UpSet plot to compare them.

```{r sex significant isoforms upset plot}
# write function to get significant isoforms
get_sig_isoforms <- function(x) {
  # get name of object
  name <- substr(x, 1, 4)
  # assign name
  assign(
    "temp_sex_switchlist_analyzed",
    get(paste0(name, "_sex_switchlist_analyzed"))
  )
  # get unqiue isoforms for that gene
  temp_sex_sig_isoforms <- unique(
    dplyr::filter(
      temp_sex_switchlist_analyzed$isoformFeatures,
      abs(dIF) > 0.1 & isoform_switch_q_value < 0.05
    )$isoform_id
  )
  # rename object
  assign(paste0(name, "_sex_sig_isoforms"),
    temp_sex_sig_isoforms,
    envir = .GlobalEnv
  )
}
# run new function on all regions
lapply(all_regions, get_sig_isoforms)

# make all into list
all_sex_isoform_list <- list(
  cerebellum = cere_sex_sig_isoforms,
  cortex = cort_sex_sig_isoforms,
  hippocampus = hipp_sex_sig_isoforms,
  striatum = stri_sex_sig_isoforms
)

# make matrix
all_sex_isoform_list_mat <- list_to_matrix(all_sex_isoform_list)

# create combination matrix
all_sex_isoform_list_comb_mat <- make_comb_mat(all_sex_isoform_list_mat)

# make UpSet plot (ComplexHeatmap)
upset_sex_isoform <- UpSet(
  all_sex_isoform_list_comb_mat,
  comb_order = order(-comb_size(all_sex_isoform_list_comb_mat)),
  comb_col = viridis(4)[comb_degree(all_sex_isoform_list_comb_mat)],
  top_annotation = upset_top_annotation(all_sex_isoform_list_comb_mat,
    add_numbers = TRUE
  ),
  right_annotation = upset_right_annotation(all_sex_isoform_list_comb_mat,
    add_numbers = TRUE
  )
)

draw(upset_sex_isoform,
  column_title = "Overlap of sex-DTU isoforms by brain region"
)
```

There is no overlap! An UpSet plot might not be the best visualization if there is no overlap. I find this interesting since that means they are all brain-region specific sex differences. So what are they? Are they enriched for anything?

#### GO Analysis using gprofiler2

Before doing GO analysis, I need to pull out signficant genes. GO analysis does not surrently work at the isoform level (at least for gprofiler2).

```{r pull out signficant genes}
# make simple function to get significant genes and cutting off decimals
get_cut_sig_genes <- function(x) {
  # get name
  name <- substr(x, 1, 4)
  # assign name
  assign(
    "temp_sex_switchlist_analyzed",
    get(paste0(name, "_sex_switchlist_analyzed"))
  )
  # subset features
  temp_sex_sig_features <- dplyr::filter(
    temp_sex_switchlist_analyzed$isoformFeatures,
    abs(dIF) > 0.1 & isoform_switch_q_value < 0.05
  )
  # assign object
  assign(paste0(name, "_sig_features"),
    temp_sex_sig_features,
    envir = .GlobalEnv
  )
  # pull genes
  sig_genes <- unique(temp_sex_sig_features$gene_id)
  sig_genes <- str_extract(sig_genes, "ENSMUSG...........")
  # assign object
  assign(paste0(name, "_sex_sig_genes"),
    sig_genes,
    envir = .GlobalEnv
  )
}
# run new function on all brain regions
lapply(all_regions, get_cut_sig_genes)
```

```{r go analysis}
# make go analysis function
run_gprofiler <- function(x) {
  # get name
  name <- substr(x, 1, 4)
  # assign name
  assign(
    "temp_sex_sig_genes",
    get(paste0(name, "_sex_sig_genes"))
  )
  # run gpfrofiler
  temp_sex_gostres <- gost(
    query = temp_sex_sig_genes, organism = "mmusculus", ordered_query = FALSE,
    multi_query = FALSE, significant = TRUE, exclude_iea = FALSE,
    measure_underrepresentation = FALSE, evcodes = FALSE,
    user_threshold = 0.05, correction_method = "g_SCS", domain_scope = "custom",
    custom_bg = str_extract(
      unique(cere_sex_switchlist_analyzed$isoformFeatures$gene_id),
      pattern = "ENSMUSG..........."
    ),
    numeric_ns = "", sources = NULL, as_short_link = FALSE
  )
  # only save and plot if there are significant results
  if (!is.null(temp_sex_gostres)) {
    # save object
    assign(paste0(name, "_sex_gostres"),
      temp_sex_gostres,
      envir = .GlobalEnv
    )
    # plot object
    gostplot(temp_sex_gostres, capped = TRUE, interactive = FALSE)
  }
}
# run new function on all brain regions
lapply(all_regions, run_gprofiler)
```

It does not look like there is a lot of enrichment here. There are only a few genes so it makes sense to me that there is very little enrichment.

Now, I want to get the gene symbols.

```{r get top sex gene symbols}
# write function to get symbols for each list
get_gene_symbols <- function(x) {
  # get name of object
  name <- substr(x, 1, 4)
  # assign name
  assign("temp_sex_sig_genes", get(paste0(name, "_sex_sig_genes")))
  # get gene symbols from annotation dbi
  temp_sex_sig_gene_symbols <- AnnotationDbi::select(
    org.Mm.eg.db,
    keys = temp_sex_sig_genes,
    columns = c("SYMBOL", "GENENAME", "ENSEMBL"), keytype = "ENSEMBL"
  )
  # rename object
  assign(paste0(name, "_sex_sig_gene_symbols"),
    temp_sex_sig_gene_symbols,
    envir = .GlobalEnv
  )
}
# run new function on all regions
lapply(all_regions, get_gene_symbols)
```

#### Clean up script

```{r tidy script}
style_file("dtu_region_sex.Rmd")
lint("dtu_region_sex.Rmd", linters = linters_with_defaults(
  object_length_linter = NULL,
  object_name_linter = NULL,
  object_usage_linter = NULL
  )
)
```

#### Software versions

```{r versions}
sessionInfo()
```
