---
title: "Victoria Please Run"
author: "Emma Jones"
date: "2023-06-29"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Differential Transcript Usage (DTU) Analysis in Mouse Brain lrRNA-Seq Data

For code review, I am tracking the amount of time that this analysis will take.

```{r proc time}
ptm <- proc.time()
```

First, we need to load in all packages.

#### Load in packages

```{r load packages and set seed}
suppressPackageStartupMessages({
  library(tidyverse)
  library(IsoformSwitchAnalyzeR)
  library(ComplexHeatmap)
  library(viridis)
  library(org.Mm.eg.db)
  library(gprofiler2)
  library(styler)
  library(lintr)
  library(here)
})

set.seed(123)
```

#### Load in data

Next, I want to read in data and get cpm values. Since I'll be doing this for all 3 analyses, I made it an R script that can be run beforehand and saves the data to an RData file. I will also source all of the functions I wrote.

```{r read in data and get cpm values}
load(here("data", "cpm_out", "cpm_counts_metadata.RData"))
source("functions.R", local = knitr::knit_global())
```

## Begin DTU analysis for pairwise brain region comparisons

The package requires a design matrix in specific format.

```{r make design matrix}
exp_design <- data.frame(
  sampleID = sample_collection_metadata$sample_id,
  condition = sample_collection_metadata$tissue
)
```

### Run satuRn

Now, I can make a switchlists and run satuRn. Because my data includes counts of less than 1, I still want to include them in my analysis. Therefore, I made the geneExpressionCutoff = NULL. This is a custom function from the functions script.

```{r make switchlist and run saturn}

# create switchlist
region_region_switchlist <- importRdata(
  isoformCountMatrix = merged_counts_iso,
  isoformRepExpression = cpm_iso,
  designMatrix = exp_design,
  isoformExonAnnoation = here(
    "data", "nextflow", "bambu",
    "extended_annotations.gtf"
  ),
  isoformNtFasta = here("data", "gffread", "isoform_sequences.fa")
)

# filter switchlist
region_region_switchlist <- preFilter(region_region_switchlist,
                                      geneExpressionCutoff = NULL)

# run satuRn, with reduce to switching genes equal to true
switchlist_analyzed_true <- isoformSwitchTestSatuRn(
  switchAnalyzeRlist = region_region_switchlist,
  reduceToSwitchingGenes = TRUE
)

extractSwitchSummary(switchlist_analyzed_true)

# run satuRn, with reduce to switching genes equal to false
switchlist_analyzed_false <- isoformSwitchTestSatuRn(
  switchAnalyzeRlist = region_region_switchlist,
  reduceToSwitchingGenes = FALSE
)

extractSwitchSummary(switchlist_analyzed_false)

```

Emma's expected results:

filtering:
The filtering removed 50748 ( 53.34% of ) transcripts. There is now 44386 isoforms left

when reduce to switching genes = TRUE

cerebellum vs cortex	444	331	288	
cerebellum vs hippocampus	169	129	114	
cerebellum vs striatum	514	419	321	
cortex vs hippocampus	40	32	30	
cortex vs striatum	253	182	167	
hippocampus vs striatum	70	67	48	
Combined	924	819	570	

when reduce to switching genes = FALSE

cerebellum vs cortex	475	331	319	
cerebellum vs hippocampus	175	129	120	
cerebellum vs striatum	548	419	355	
cortex vs hippocampus	41	32	31	
cortex vs striatum	273	182	187	
hippocampus vs striatum	72	67	50	
Combined	1002	819	648	