---
title: "Comparing DTU, DTE, and DGE"
author: "Emma Jones"
date: "2023-06-01"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Comparing DTU, DTE, and DGE

The purpose of this script is to compare genes with differential gene expression, differential transcript usage, and differential transcript expression. You must be using docker 1.7. It makes eulerr and upset plots for all overlapping comparisons. These eulerr plots are used in figure 3 of the manuscript.

For code review, I am tracking the amount of time that this analysis will take.

```{r proc time}
ptm <- proc.time()
```

#### Load in Packages

```{r load packages}
# load packages
suppressPackageStartupMessages({
  library(tidyverse)
  library(IsoformSwitchAnalyzeR)
  library(ComplexHeatmap)
  library(viridis)
  library(styler)
  library(lintr)
  library(here)
  library(eulerr)
  library(cowplot)
})

# source functions
source("de_functions.R")
```

### Set color palettes

I get my colors based on palettes from Paul Tol: https://personal.sron.nl/~pault/

My palette borrows other colors from tol_light that were less similar to the three I chose for my eulerr diagrams.

```{r get colors}
EJ_custom_palette <- c(
  "#EEDD88", "#77AADD", "#FFAABB", "#5D69B1", "#EE8866", "#52BCA3", "#99C945",
  "#99DDFF", "#DDDDDD"
)
```

## Load in Data

#### Load in DGE data

First, we want to load in the DGE data

```{r load in DGE data}
# DGE
dge_files <- list.files(here("results", "de_genes"))
# loop in files
for (i in dge_files) {
  temp <- read.csv(
    here("results", "de_genes", i),
    header = FALSE
  )
  genes <- temp$V1
  assign(paste0(str_sub(i, end = -5), "_dge_genes"),
    genes,
    envir = .GlobalEnv
  )
}
```

#### Load in DTE data

Next, we need to pull in the transcript counts files to get gene name information for all transcripts and load in the DTE lists.

```{r pull in DTE data}
transcript_counts <- read.table(
  here("data", "nextflow", "bambu", "counts_transcript.txt"),
  header = TRUE
)

tx_gene_names <- transcript_counts[, 1:2]

# DTE
dte_files <- list.files(here("results", "de_transcripts"))
# loop in files
for (i in dte_files) {
  temp <- read.csv(
    here("results", "de_transcripts", i),
    header = FALSE
  )
  transcripts <- temp$V1
  genes <- tx_gene_names$GENEID[tx_gene_names$TXNAME %in% transcripts]
  assign(paste0(str_sub(i, end = -5), "_dte_genes"),
    genes,
    envir = .GlobalEnv
  )
}
```

#### Load in DTU data

Finally, load in the DTU data for comparisons. It is important to note here that because some of it is stored in lists, I need two different ways of importing the data.

```{r load in DTU data}
# start with region region
region_region_switchlist <- readRDS(
  here("data", "switchlist_objects", "de_added", "region_region_orf_de.Rds")
)
sig_region_region <- dplyr::filter(
  region_region_switchlist$isoformFeatures,
  abs(dIF) > 0.1 & isoform_switch_q_value < 0.05
)
region_region_dtu_genes <- unique(sig_region_region$gene_id)

# region others
region_all_switchlist_list <- readRDS(
  here(
    "data", "switchlist_objects", "de_added",
    "region_all_switchlist_list_orf_de.Rds"
  )
)
for (i in names(region_all_switchlist_list)) {
  sig_temp <- dplyr::filter(
    region_all_switchlist_list[[i]]$isoformFeatures,
    abs(dIF) > 0.1 & isoform_switch_q_value < 0.05
  )
  genes <- sig_temp$gene_id
  genes <- unique(genes)
  assign(paste0(i, "_dtu_genes"),
    genes,
    envir = .GlobalEnv
  )
}

# region sex
region_sex_switchlist_list <- readRDS(
  here(
    "data", "switchlist_objects", "de_added",
    "region_sex_switchlist_list_orf_de.Rds"
  )
)
for (i in names(region_sex_switchlist_list)) {
  sig_temp <- dplyr::filter(
    region_sex_switchlist_list[[i]]$isoformFeatures,
    abs(dIF) > 0.1 & isoform_switch_q_value < 0.05
  )
  genes <- sig_temp$gene_id
  genes <- unique(genes)
  assign(paste0(i, "_sex_dtu_genes"),
    genes,
    envir = .GlobalEnv
  )
}
```

Next, we can pull what we need from each file to make single comparison lists.

```{r pull out individual dtu comparisons}
# cerebellum cortex
cerebellum_cortex_subset <- filter(
  sig_region_region,
  condition_1 == "cerebellum" & condition_2 == "cortex"
)
cerebellum_cortex_dtu_genes <- cerebellum_cortex_subset$gene_id

# cerebellum hippocampus
cerebellum_hippocampus_subset <- filter(
  sig_region_region,
  condition_1 == "cerebellum" & condition_2 == "hippocampus"
)
cerebellum_hippocampus_dtu_genes <- cerebellum_hippocampus_subset$gene_id

# cerebellum striatum
cerebellum_striatum_subset <- filter(
  sig_region_region,
  condition_1 == "cerebellum" & condition_2 == "striatum"
)
cerebellum_striatum_dtu_genes <- cerebellum_striatum_subset$gene_id

# cortex hippocampus
cortex_hippocampus_subset <- filter(
  sig_region_region,
  condition_1 == "cortex" & condition_2 == "hippocampus"
)
cortex_hippocampus_dtu_genes <- cortex_hippocampus_subset$gene_id

# cortex striatum
cortex_striatum_subset <- filter(
  sig_region_region,
  condition_1 == "cortex" & condition_2 == "striatum"
)
cortex_striatum_dtu_genes <- cortex_striatum_subset$gene_id

# hippocampus striatum
hippocampus_striatum_subset <- filter(
  sig_region_region,
  condition_1 == "hippocampus" & condition_2 == "striatum"
)
hippocampus_striatum_dtu_genes <- hippocampus_striatum_subset$gene_id
```

## Make comparison lists for stats and plotting

#### Make pairwise lists

Now that everything is extracted, we need to make it into lists. I'll start with region-region direct comparisons.

```{r prep region region lists}
cerebellum_cortex_list <- list(
  `DGE` = cerebellum_cortex_dge_genes,
  `DTE` = cerebellum_cortex_dte_genes,
  `DTU` = cerebellum_cortex_dtu_genes
)

cerebellum_hippocampus_list <- list(
  `DGE` = cerebellum_hippocampus_dge_genes,
  `DTE` = cerebellum_hippocampus_dte_genes,
  `DTU` = cerebellum_hippocampus_dtu_genes
)

cerebellum_striatum_list <- list(
  `DGE` = cerebellum_striatum_dge_genes,
  `DTE` = cerebellum_striatum_dte_genes,
  `DTU` = cerebellum_striatum_dtu_genes
)

cortex_hippocampus_list <- list(
  `DGE` = cortex_hippocampus_dge_genes,
  `DTE` = cortex_hippocampus_dte_genes,
  `DTU` = cortex_hippocampus_dtu_genes
)

cortex_striatum_list <- list(
  `DGE` = cortex_striatum_dge_genes,
  `DTE` = cortex_striatum_dte_genes,
  `DTU` = cortex_striatum_dtu_genes
)

hippocampus_striatum_list <- list(
  `DGE` = striatum_hippocampus_dge_genes,
  `DTE` = striatum_hippocampus_dte_genes,
  `DTU` = hippocampus_striatum_dtu_genes
)
```

#### Make single region lists 

Next, we can do for the individual regions.

```{r prep single region data}
# make results into a list
cerebellum_list <- list(
  `DGE` = cerebellum_dge_genes,
  `DTE` = cerebellum_dte_genes,
  `DTU` = cerebellum_dtu_genes
)

cortex_list <- list(
  `DGE` = cortex_dge_genes,
  `DTE` = cortex_dte_genes,
  `DTU` = cortex_dtu_genes
)

hippocampus_list <- list(
  `DGE` = hippocampus_dge_genes,
  `DTE` = hippocampus_dte_genes,
  `DTU` = hippocampus_dtu_genes
)

striatum_list <- list(
  `DGE` = striatum_dge_genes,
  `DTE` = striatum_dte_genes,
  `DTU` = striatum_dtu_genes
)
```

#### Make across sex within region lists

Finally, we want to compare to within brain regions across sex.

```{r make list for comparing within brain region across sex}
# make results into a list
cerebellum_sex_list <- list(
  `DGE` = cerebellum_sex_dge_genes,
  `DTE` = cerebellum_sex_dte_genes,
  `DTU` = cerebellum_sex_dtu_genes
)

cortex_sex_list <- list(
  `DGE` = cortex_sex_dge_genes,
  `DTE` = cortex_sex_dte_genes,
  `DTU` = cortex_sex_dtu_genes
)

hippocampus_sex_list <- list(
  `DGE` = hippocampus_sex_dge_genes,
  `DTE` = hippocampus_sex_dte_genes,
  `DTU` = hippocampus_sex_dtu_genes
)

striatum_sex_list <- list(
  `DGE` = striatum_sex_dge_genes,
  `DTE` = striatum_sex_dte_genes,
  `DTU` = striatum_sex_dtu_genes
)
```

#### Make compound lists

I also want to look at the overlap of DGE and DTE with themselves, as compared to the DTU in script 5. This code allows us to make lists, the corresponding matrices and combination matrices required for the upset plots. 

##### Pairwise comparision

I'll start with the pairwise genes identified, the largest gene lists.

```{r create pairwise lists of DGE and DTE overlap}
dge_main_list <- list(
  `Cerebellum vs. Cortex` = cerebellum_cortex_dge_genes,
  `Cerebellum vs. Hippocampus` = cerebellum_hippocampus_dge_genes,
  `Cerebellum vs. Striatum` = cerebellum_striatum_dge_genes,
  `Cortex vs. Hippocampus` = cortex_hippocampus_dge_genes,
  `Cortex vs. Striatum` = cortex_striatum_dge_genes,
  `Hippocampus vs. Striatum` = striatum_hippocampus_dge_genes
)

dte_main_list <- list(
  `Cerebellum vs. Cortex` = cerebellum_cortex_dte_genes,
  `Cerebellum vs. Hippocampus` = cerebellum_hippocampus_dte_genes,
  `Cerebellum vs. Striatum` = cerebellum_striatum_dte_genes,
  `Cortex vs. Hippocampus` = cortex_hippocampus_dte_genes,
  `Cortex vs. Striatum` = cortex_striatum_dte_genes,
  `Hippocampus vs. Striatum` = striatum_hippocampus_dte_genes
)

dtu_main_list <- list(
  `Cerebellum vs. Cortex` = cerebellum_cortex_dtu_genes,
  `Cerebellum vs. Hippocampus` = cerebellum_hippocampus_dtu_genes,
  `Cerebellum vs. Striatum` = cerebellum_striatum_dtu_genes,
  `Cortex vs. Hippocampus` = cortex_hippocampus_dtu_genes,
  `Cortex vs. Striatum` = cortex_striatum_dtu_genes,
  `Hippocampus vs. Striatum` = hippocampus_striatum_dtu_genes
)
```

##### Single region comparisons

Also, I want to make lists of all dge and dtu single region comparisons.

```{r make lists of all dge and dtu single region comparisons}
dge_single_list <- list(
  `Cerebellum` = cerebellum_dge_genes,
  `Hippocampus` = hippocampus_dge_genes,
  `Cortex` = cortex_dge_genes,
  `Striatum` = striatum_dge_genes
)

dte_single_list <- list(
  `Cerebellum` = cerebellum_dte_genes,
  `Hippocampus` = hippocampus_dte_genes,
  `Cortex` = cortex_dte_genes,
  `Striatum` = striatum_dte_genes
)

dtu_single_list <- list(
  `Cerebellum` = cerebellum_dtu_genes,
  `Hippocampus` = hippocampus_dtu_genes,
  `Cortex` = cortex_dtu_genes,
  `Striatum` = striatum_dtu_genes
)
```

##### Sex region comparisions

And, I want to make make lists of all dge and dtu sex comparisons.

```{r make lists of all dge and dtu sex comparisons}
dge_sex_list <- list(
  `Cerebellum` = cerebellum_sex_dge_genes,
  `Hippocampus` = hippocampus_sex_dge_genes,
  `Cortex` = cortex_sex_dge_genes,
  `Striatum` = striatum_sex_dge_genes
)

dte_sex_list <- list(
  `Cerebellum` = cerebellum_sex_dte_genes,
  `Hippocampus` = hippocampus_sex_dte_genes,
  `Cortex` = cortex_sex_dte_genes,
  `Striatum` = striatum_sex_dte_genes
)

dtu_sex_list <- list(
  `Cerebellum` = cerebellum_sex_dtu_genes,
  `Hippocampus` = hippocampus_sex_dtu_genes,
  `Cortex` = cortex_sex_dtu_genes,
  `Striatum` = striatum_sex_dtu_genes
)
```

We use all of these lists a lot, so I save them as an intermediate R object for simplicity's sake. You can now more easily load this data if you find you need to plot it again for any reason.

```{r save object lists}
save.image(file = here(
  "data", "comparison_gene_lists",
  "all_comparison_gene_lists.Rdata"
))

# you can also load the data here, if needed
```

## Plotting DGE, DTE, and DTU overlaps!

I created a make comparison plots function to create eulerr diagrams and upset plots for overlapping genes. We can save these all in the next few code chunks.

#### Pairwise region overlap plotting

We can run the function first on all pairwise brain region comparisons.

```{r plot comparisions for pairwise region}
make_comparison_plots(cerebellum_cortex_list,
  save_path = here("results", "plots")
)

make_comparison_plots(cerebellum_hippocampus_list,
  save_path = here("results", "plots")
)

make_comparison_plots(cerebellum_striatum_list,
  save_path = here("results", "plots")
)

make_comparison_plots(cortex_hippocampus_list,
  save_path = here("results", "plots")
)

make_comparison_plots(cortex_striatum_list,
  save_path = here("results", "plots")
)

make_comparison_plots(hippocampus_striatum_list,
  save_path = here("results", "plots")
)
```

#### Single region overlap plotting

I also want to run this function on the single region comparisons.

```{r run plotting function on single brain regions}
make_comparison_plots(cerebellum_list, save_path = here("results", "plots"))

make_comparison_plots(cortex_list, save_path = here("results", "plots"))

make_comparison_plots(hippocampus_list, save_path = here("results", "plots"))

make_comparison_plots(striatum_list, save_path = here("results", "plots"))
```

#### Sex with region overlap plotting

And finally, plot comparisons within brain regions across sex. This output is what is actually used for figure 3.

```{r plot comps within region across sex}
make_comparison_plots(cerebellum_sex_list,
  save_path = here("results", "plots")
)

make_comparison_plots(cortex_sex_list,
  save_path = here("results", "plots")
)


make_comparison_plots(hippocampus_sex_list,
  save_path = here("results", "plots")

)

make_comparison_plots(striatum_sex_list,
  save_path = here("results", "plots")
)
```

## Additional statistics

Once we have made these lists from earlier, we can also use them to get some basic statistics on how many genes overlap and how many are unique. This is mainly for the manuscript.

First I aggregate all dge and dte genes together into a long list.

```{r aggregate all genes together}
# aggregate all dge genes
all_pairwise_dge_genes <- unique(c(
  cerebellum_cortex_dge_genes, cerebellum_hippocampus_dge_genes,
  cerebellum_striatum_dge_genes, cortex_hippocampus_dge_genes,
  cortex_striatum_dge_genes,
  striatum_hippocampus_dge_genes
))


# aggregate all dte genes
all_pairwise_dte_genes <- unique(c(
  cerebellum_cortex_dte_genes, cerebellum_hippocampus_dte_genes,
  cerebellum_striatum_dte_genes, cortex_hippocampus_dte_genes,
  cortex_striatum_dte_genes,
  striatum_hippocampus_dte_genes
))


# aggregate all pairwise dtu genes
all_pairwise_dtu_genes <- unique(c(
  cerebellum_cortex_dtu_genes, cerebellum_hippocampus_dtu_genes,
  cerebellum_striatum_dtu_genes, cortex_hippocampus_dtu_genes,
  cortex_striatum_dtu_genes,
  hippocampus_striatum_dtu_genes
))
```

Now when can get the lengths of the unque genes in these lists.

```{r get pirwise comparisions stats part 1 - how many intersect?}
# how many pairwise dge and dte genes intersect?
length(intersect(all_pairwise_dge_genes, all_pairwise_dte_genes))
```

Expected results: 3042

```{r get pairwise comparisions stats part 2 - how much overlap}
# how many pairwise dtu genes overlap with dge
length(intersect(all_pairwise_dge_genes, all_pairwise_dtu_genes))

# and what percent dtu genes overlap with dge?
length(intersect(all_pairwise_dge_genes, all_pairwise_dtu_genes)) /
  length(all_pairwise_dtu_genes)

# how many pairwise dtu genes overlap with dte
length(intersect(all_pairwise_dte_genes, all_pairwise_dtu_genes))

# and what percent dtu genes overlap with dte?
length(intersect(all_pairwise_dte_genes, all_pairwise_dtu_genes)) /
  length(all_pairwise_dtu_genes)
```

Expected results:
[1] 328
[1] 0.5061728
[1] 504
[1] 0.7777778

What about any genes identified with DESeq2 at all?

```{r get pairwise comparisions stats part 3 - what about all DESeq2}
# all dge or dte genes
all_dge_dte <- union(all_pairwise_dge_genes, all_pairwise_dte_genes)

# how many pairwise dtu genes overlap with dge or dte
length(intersect(all_dge_dte, all_pairwise_dtu_genes))

# and what percent dtu genes overlap with dge or dte?
length(intersect(all_dge_dte, all_pairwise_dtu_genes)) /
  length(all_pairwise_dtu_genes)
```

Expected results:
[1] 512
[1] 0.7901235

How many DTU genes were unidentified by DGE or DTE?

```{r get more pairwise stats part 4 - how many unidentified}
# leaving how many unidentified?
length(all_pairwise_dtu_genes) - length(intersect(
  all_dge_dte,
  all_pairwise_dtu_genes
))

# and what percent?
(length(all_pairwise_dtu_genes) - length(intersect(
  all_dge_dte,
  all_pairwise_dtu_genes
))) /
  length(all_pairwise_dtu_genes)
```

Expected results:
[1] 136
[1] 0.2098765

This means that 21% of DTU genes would not have been identified without using satuRn.

#### Clean up script

```{r tidy script}
style_file("15_compare_DTU_DGE.Rmd")
lint("15_compare_DTU_DGE.Rmd",
  linters = linters_with_defaults(
    object_length_linter = NULL,
    object_name_linter = NULL,
    object_usage_linter = NULL
  )
)
```

Also, needed to get processing time.

```{r proc time finish}
fptm <- proc.time() - ptm
fptm[3] / 60
```

22 minutes to run, including the time it took to edit all text and comment.

#### Software versions

My software versions will be commented below.

```{r versions}
sessionInfo()
```

R version 4.3.1 (2023-06-16)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.3 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
 [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] grid      stats4    stats     graphics  grDevices utils     datasets  methods  
[9] base     

other attached packages:
 [1] cowplot_1.1.1               eulerr_7.0.0                here_1.0.1                 
 [4] lintr_3.1.0                 styler_1.10.2               viridis_0.6.4              
 [7] viridisLite_0.4.2           ComplexHeatmap_2.16.0       IsoformSwitchAnalyzeR_2.1.3
[10] pfamAnalyzeR_1.0.1          sva_3.48.0                  genefilter_1.82.1          
[13] mgcv_1.9-0                  nlme_3.1-163                satuRn_1.8.0               
[16] DEXSeq_1.46.0               RColorBrewer_1.1-3          AnnotationDbi_1.62.2       
[19] DESeq2_1.40.2               SummarizedExperiment_1.30.2 GenomicRanges_1.52.0       
[22] GenomeInfoDb_1.36.3         IRanges_2.34.1              S4Vectors_0.38.1           
[25] MatrixGenerics_1.12.3       matrixStats_1.0.0           Biobase_2.60.0             
[28] BiocGenerics_0.46.0         BiocParallel_1.34.2         limma_3.56.2               
[31] lubridate_1.9.2             forcats_1.0.0               stringr_1.5.0              
[34] dplyr_1.1.3                 purrr_1.0.2                 readr_2.1.4                
[37] tidyr_1.3.0                 tibble_3.2.1                ggplot2_3.4.3              
[40] tidyverse_2.0.0            

loaded via a namespace (and not attached):
  [1] splines_4.3.1                 later_1.3.1                  
  [3] BiocIO_1.10.0                 bitops_1.0-7                 
  [5] filelock_1.0.2                R.oo_1.25.0                  
  [7] polyclip_1.10-6               rex_1.2.1                    
  [9] XML_3.99-0.14                 lifecycle_1.0.3              
 [11] rprojroot_2.0.3               edgeR_3.42.4                 
 [13] doParallel_1.0.17             processx_3.8.2               
 [15] lattice_0.21-8                ensembldb_2.24.0             
 [17] backports_1.4.1               magrittr_2.0.3               
 [19] rmarkdown_2.24                remotes_2.4.2.1              
 [21] yaml_2.3.7                    httpuv_1.6.11                
 [23] pbapply_1.7-2                 DBI_1.1.3                    
 [25] abind_1.4-5                   zlibbioc_1.46.0              
 [27] R.cache_0.16.0                R.utils_2.12.2               
 [29] AnnotationFilter_1.24.0       RCurl_1.98-1.12              
 [31] rappdirs_0.3.3                circlize_0.4.15              
 [33] GenomeInfoDbData_1.2.10       annotate_1.78.0              
 [35] codetools_0.2-19              DelayedArray_0.26.7          
 [37] xml2_1.3.5                    tidyselect_1.2.0             
 [39] shape_1.4.6                   futile.logger_1.4.3          
 [41] locfdr_1.1-8                  farver_2.1.1                 
 [43] BiocFileCache_2.8.0           GenomicAlignments_1.36.0     
 [45] jsonlite_1.8.7                GetoptLong_1.0.5             
 [47] ellipsis_0.3.2                survival_3.5-7               
 [49] iterators_1.0.14              polylabelr_0.2.0             
 [51] foreach_1.5.2                 tools_4.3.1                  
 [53] progress_1.2.2                Rcpp_1.0.11                  
 [55] glue_1.6.2                    gridExtra_2.3                
 [57] xfun_0.40                     withr_2.5.0                  
 [59] formatR_1.14                  BiocManager_1.30.22          
 [61] fastmap_1.1.1                 boot_1.3-28.1                
 [63] fansi_1.0.4                   callr_3.7.3                  
 [65] digest_0.6.33                 timechange_0.2.0             
 [67] R6_2.5.1                      mime_0.12                    
 [69] colorspace_2.1-0              biomaRt_2.56.1               
 [71] RSQLite_2.3.1                 R.methodsS3_1.8.2            
 [73] utf8_1.2.3                    generics_0.1.3               
 [75] data.table_1.14.8             tximeta_1.18.3               
 [77] rtracklayer_1.60.1            prettyunits_1.1.1            
 [79] httr_1.4.7                    S4Arrays_1.0.6               
 [81] pkgconfig_2.0.3               gtable_0.3.4                 
 [83] blob_1.2.4                    hwriter_1.3.2.1              
 [85] XVector_0.40.0                htmltools_0.5.6              
 [87] geneplotter_1.78.0            ProtGenerics_1.32.0          
 [89] clue_0.3-64                   scales_1.2.1                 
 [91] cyclocomp_1.1.1               png_0.1-8                    
 [93] knitr_1.44                    lambda.r_1.2.4               
 [95] rstudioapi_0.15.0             tzdb_0.4.0                   
 [97] reshape2_1.4.4                rjson_0.2.21                 
 [99] curl_5.0.2                    cachem_1.0.8                 
[101] GlobalOptions_0.1.2           BiocVersion_3.17.1           
[103] parallel_4.3.1                desc_1.4.2                   
[105] restfulr_0.0.15               pillar_1.9.0                 
[107] vctrs_0.6.3                   promises_1.2.1               
[109] dbplyr_2.3.3                  cluster_2.1.4                
[111] xtable_1.8-4                  tximport_1.28.0              
[113] evaluate_0.21                 VennDiagram_1.7.3            
[115] GenomicFeatures_1.52.2        cli_3.6.1                    
[117] locfit_1.5-9.8                compiler_4.3.1               
[119] futile.options_1.0.1          Rsamtools_2.16.0             
[121] rlang_1.1.1                   crayon_1.5.2                 
[123] ps_1.7.5                      plyr_1.8.8                   
[125] stringi_1.7.12                munsell_0.5.0                
[127] Biostrings_2.68.1             lazyeval_0.2.2               
[129] Matrix_1.6-1                  BSgenome_1.68.0              
[131] hms_1.1.3                     bit64_4.0.5                  
[133] KEGGREST_1.40.0               statmod_1.5.0                
[135] shiny_1.7.5                   interactiveDisplayBase_1.38.0
[137] AnnotationHub_3.8.0           memoise_2.0.1                
[139] bit_4.0.5                     xmlparsedata_1.0.5  