---
title: 'DE Analysis: Pairwise Brain Regions'
author: "Emma Jones"
date: "2023-05-22"
output: html_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Script for DE Analysis: Pairwise brain regions

For code review, I am tracking the amount of time that this analysis will take.

```{r proc time}
ptm <- proc.time()
```


#### Load in Packages

```{r load in packages}
suppressPackageStartupMessages({
  library(tidyverse)
  library(DESeq2)
  library(styler)
  library(lintr)
  library(here)
})

set.seed(123)
```

Read in data and functions

```{r read in data}
load(here("data", "cpm_out", "cpm_counts_metadata.RData"))
source(here("src", "de_analysis", "de_functions.R"))

# read in gene counts data
gene_counts <- read.table(
  here("data", "nextflow", "bambu", "counts_gene.txt"),
  header = TRUE
)

# rename columns
colnames(gene_counts) <- str_extract(colnames(gene_counts), "sample..")
```

### Run DESeq2

First I need to make the DESeq DataSet object

```{r prep data for deseq2}
dds <- DESeqDataSetFromMatrix(
  countData = round(gene_counts),
  colData = sample_collection_metadata,
  design = ~ sex + tissue
)
nrow(dds)

keep <- rowSums(counts(dds)) > 1
dds <- dds[keep, ]
nrow(dds)

# vst normalization
vsd <- vst(dds, blind = FALSE)
head(assay(vsd), 3)

colData(vsd)

plotPCA(vsd, intgroup = c("sex", "tissue"))

plotPCA(vsd, intgroup = c("tissue"))

plotPCA(vsd, intgroup = c("med_len"))

plotPCA(vsd, intgroup = c("RIN"))

vst_tissue_pca <- plotPCA(vsd, intgroup = c("tissue"))

ggsave(here("results", "plots", "pca", "vst_tissue_pca.png"),
       plot = vst_tissue_pca)
```

Next, I can run DESeq2. I don't think I need to include anything other than tissue and sex based on these plots. I'm leaving sex because I'm interested in it, thought I know it may not have a lot of sig differences.

```{r run DESeq2}
dds <- estimateSizeFactors(dds)
dds <- estimateDispersions(dds)
dds <- nbinomWaldTest(dds, maxit = 5000) # the more I increase this
# the less "did not converge" columns you will get
```

Run function for saving data

```{r get de results}
get_de_results("cerebellum", "cortex")
get_de_results("cerebellum", "hippocampus")
get_de_results("cerebellum", "striatum")
get_de_results("cortex", "hippocampus")
get_de_results("cortex", "striatum")
get_de_results("striatum", "hippocampus")
```
Expected output: 
[1] "cerebellum versus cortex had 5702 significant DE genes"
[1] "cerebellum versus hippocampus had 4687 significant DE genes"
[1] "cerebellum versus striatum had 5268 significant DE genes"
[1] "cortex versus hippocampus had 1145 significant DE genes"
[1] "cortex versus striatum had 2256 significant DE genes"
[1] "striatum versus hippocampus had 1855 significant DE genes"

## Do DESeq2 at the isoform level. Note: THIS WILL TREAT TRANSCRIPTS AS SEPARATE GENES

```{r redo at isoform level}
# read in transcript counts data
transcript_counts <- read.table(
  here("data", "nextflow", "bambu", "counts_transcript.txt"),
  header = TRUE
)

# rename columns
colnames(transcript_counts) <- str_extract(colnames(transcript_counts), "sample..")

# remove gene names
transcript_counts <- transcript_counts[,-2]

# add rownames
rownames(transcript_counts) <- transcript_counts[,1]

# remove isoform_id column
transcript_counts <- transcript_counts[,-1]
```


```{r prep data for deseq2}
dds_transcripts <- DESeqDataSetFromMatrix(
  countData = round(transcript_counts),
  colData = sample_collection_metadata,
  design = ~ sex + tissue
)
nrow(dds_transcripts)

keep <- rowSums(counts(dds_transcripts)) > 1
dds_transcripts <- dds_transcripts[keep, ]
nrow(dds_transcripts)

# vst normalization
vsd_transcripts <- vst(dds_transcripts, blind = FALSE)
head(assay(vsd_transcripts), 3)

colData(vsd_transcripts)

plotPCA(vsd_transcripts, intgroup = c("tissue"))
```

After filtering, I get 60,526 transcripts that it decides to keep.

Next, I can run DESeq2.

```{r run DESeq2}
dds_transcripts <- estimateSizeFactors(dds_transcripts)
dds_transcripts <- estimateDispersions(dds_transcripts)
dds_transcripts <- nbinomWaldTest(dds_transcripts, maxit = 5000) # the more I increase this
# the less "did not converge" columns you will get
```

Run function for saving data

```{r get de results}
get_de_results_transcripts("cerebellum", "cortex")
get_de_results_transcripts("cerebellum", "hippocampus")
get_de_results_transcripts("cerebellum", "striatum")
get_de_results_transcripts("cortex", "hippocampus")
get_de_results_transcripts("cortex", "striatum")
get_de_results_transcripts("hippocampus", "striatum")
```

Expected results:
[1] "cerebellum versus cortex had 6033 significant DE transcripts"
[1] "cerebellum versus hippocampus had 4871 significant DE transcripts"
[1] "cerebellum versus striatum had 5591 significant DE transcripts"
[1] "cortex versus hippocampus had 897 significant DE transcripts"
[1] "cortex versus striatum had 1866 significant DE transcripts"
[1] "hippocampus versus striatum had 1657 significant DE transcripts"

#### Clean up script

```{r tidy script}
style_file("DESeq2_region_region.Rmd")
lint("DESeq2_region_region.Rmd",
  linters = linters_with_defaults(
    object_length_linter = NULL,
    object_name_linter = NULL,
    object_usage_linter = NULL
  )
)
```

Also, needed to get processing time.

```{r proc time finish}
fptm <- proc.time() - ptm
fptm[3] / 60
```

This number is negative so something is wrong here lol. I observed it does not take more than 5 minutes.

#### Software versions

My software versions will be commented below.

```{r versions}
sessionInfo()
```

R version 4.3.0 beta (2023-04-12 r84240)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.2 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
 [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] here_1.0.1                  lintr_3.0.2                 styler_1.9.1               
 [4] DESeq2_1.40.0               SummarizedExperiment_1.30.0 Biobase_2.60.0             
 [7] MatrixGenerics_1.12.0       matrixStats_0.63.0          GenomicRanges_1.52.0       
[10] GenomeInfoDb_1.36.0         IRanges_2.34.0              S4Vectors_0.38.0           
[13] BiocGenerics_0.46.0         lubridate_1.9.2             forcats_1.0.0              
[16] stringr_1.5.0               dplyr_1.1.2                 purrr_1.0.1                
[19] readr_2.1.4                 tidyr_1.3.0                 tibble_3.2.1               
[22] ggplot2_3.4.2               tidyverse_2.0.0            

loaded via a namespace (and not attached):
 [1] tidyselect_1.2.0        farver_2.1.1            R.utils_2.12.2         
 [4] bitops_1.0-7            fastmap_1.1.1           RCurl_1.98-1.12        
 [7] lazyeval_0.2.2          rex_1.2.1               digest_0.6.31          
[10] timechange_0.2.0        lifecycle_1.0.3         processx_3.8.0         
[13] magrittr_2.0.3          compiler_4.3.0          rlang_1.1.1            
[16] tools_4.3.0             utf8_1.2.3              yaml_2.3.7             
[19] knitr_1.42              labeling_0.4.2          DelayedArray_0.25.0    
[22] xml2_1.3.4              BiocParallel_1.34.0     R.cache_0.16.0         
[25] withr_2.5.0             desc_1.4.2              R.oo_1.25.0            
[28] grid_4.3.0              fansi_1.0.4             colorspace_2.1-0       
[31] scales_1.2.1            cli_3.6.1               rmarkdown_2.21         
[34] crayon_1.5.2            generics_0.1.3          remotes_2.4.2          
[37] rstudioapi_0.14         tzdb_0.3.0              zlibbioc_1.46.0        
[40] parallel_4.3.0          XVector_0.40.0          vctrs_0.6.2            
[43] Matrix_1.5-4            callr_3.7.3             hms_1.1.3              
[46] cyclocomp_1.1.0         locfit_1.5-9.7          glue_1.6.2             
[49] codetools_0.2-19        ps_1.7.4                stringi_1.7.12         
[52] gtable_0.3.3            munsell_0.5.0           pillar_1.9.0           
[55] htmltools_0.5.5         GenomeInfoDbData_1.2.10 R6_2.5.1               
[58] rprojroot_2.0.3         evaluate_0.20           lattice_0.21-8         
[61] R.methodsS3_1.8.2       Rcpp_1.0.10             xfun_0.39              
[64] pkgconfig_2.0.3 
