---
title: "DESeq2 region others"
author: "Emma Jones"
date: "2023-05-25"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Script for DE analysis on one brain region vs others grouped together

For code review, I am tracking the amount of time that this analysis will take.

```{r proc time}
ptm <- proc.time()
```

#### Load in Packages

```{r load in packages}
suppressPackageStartupMessages({
  library(tidyverse)
  library(DESeq2)
  library(styler)
  library(lintr)
  library(here)
})

set.seed(123)
```

Read in data and functions

```{r read in data}
load(here("data", "cpm_out", "cpm_counts_metadata.RData"))
source(here("src", "de_analysis", "de_functions.R"))

# read in gene counts data
gene_counts <- read.table(
  here("data", "nextflow", "bambu", "counts_gene.txt"),
  header = TRUE
)

# rename columns
colnames(gene_counts) <- str_extract(colnames(gene_counts), "sample..")
```

### Run DESeq2 at gene level

Next, I can run DESeq2 at the gene level, meaning on the gene count matrices.

```{r run DESeq2 at the gene level}
deseq2_single_region(
  counts_table = gene_counts,
  metadata = sample_collection_metadata,
  region = "cerebellum",
  object_save_path = here("data", "deseq2_data"),
  results_save_path = here("results", "de_genes"),
  level = "gene"
)

deseq2_single_region(
  counts_table = gene_counts,
  metadata = sample_collection_metadata,
  region = "cortex",
  object_save_path = here("data", "deseq2_data"),
  results_save_path = here("results", "de_genes"),
  level = "gene"
)

deseq2_single_region(
  counts_table = gene_counts,
  metadata = sample_collection_metadata,
  region = "hippocampus",
  object_save_path = here("data", "deseq2_data"),
  results_save_path = here("results", "de_genes"),
  level = "gene"
)

deseq2_single_region(
  counts_table = gene_counts,
  metadata = sample_collection_metadata,
  region = "striatum",
  object_save_path = here("data", "deseq2_data"),
  results_save_path = here("results", "de_genes"),
  level = "gene"
)
```

Expected results:
[1] "cerebellum had 5916 significant DE genes"
[1] "cortex had 1969 significant DE genes"
[1] "hippocampus had 861 significant DE genes"
[1] "striatum had 2010 significant DE genes"

## Do DESeq2 at the isoform level. Note: THIS WILL TREAT TRANSCRIPTS AS SEPARATE GENES

```{r redo at isoform level}
# read in transcript counts data
transcript_counts <- read.table(
  here("data", "nextflow", "bambu", "counts_transcript.txt"),
  header = TRUE
)

# rename columns
colnames(transcript_counts) <- str_extract(
  colnames(transcript_counts),
  "sample.."
)

# remove gene names
transcript_counts <- transcript_counts[, -2]

# add rownames
rownames(transcript_counts) <- transcript_counts[, 1]

# remove isoform_id column
transcript_counts <- transcript_counts[, -1]
```

Now, do the same thing at the transcript level

```{r run DESeq2 at the transcript level}
deseq2_single_region(
  counts_table = transcript_counts,
  metadata = sample_collection_metadata,
  region = "cerebellum",
  object_save_path = here("data", "deseq2_data"),
  results_save_path = here("results", "de_transcripts"),
  level = "transcript"
)

deseq2_single_region(
  counts_table = transcript_counts,
  metadata = sample_collection_metadata,
  region = "cortex",
  object_save_path = here("data", "deseq2_data"),
  results_save_path = here("results", "de_transcripts"),
  level = "transcript"
)

deseq2_single_region(
  counts_table = transcript_counts,
  metadata = sample_collection_metadata,
  region = "hippocampus",
  object_save_path = here("data", "deseq2_data"),
  results_save_path = here("results", "de_transcripts"),
  level = "transcript"
)

deseq2_single_region(
  counts_table = transcript_counts,
  metadata = sample_collection_metadata,
  region = "striatum",
  object_save_path = here("data", "deseq2_data"),
  results_save_path = here("results", "de_transcripts"),
  level = "transcript"
)
```
Expected results: 
[1] "cerebellum had 6682 significant DE transcripts"

[1] "cortex had 1591 significant DE transcripts"

[1] "hippocampus had 747 significant DE transcripts"

[1] "striatum had 1770 significant DE transcripts"

#### Clean up script

```{r tidy script}
style_file("11_DESeq2_region_others.Rmd")
lint("11_DESeq2_region_others.Rmd",
  linters = linters_with_defaults(
    object_length_linter = NULL,
    object_name_linter = NULL,
    object_usage_linter = NULL
  )
)
```

Also, needed to get processing time.

```{r proc time finish}
fptm <- proc.time() - ptm
fptm[3] / 60
```

It says it took 219 minutes which is not true. I think the parallel processing messes this up because it only took me 10 minutes to run.

#### Software versions

My software versions will be commented below.

```{r versions}
sessionInfo()
```

R version 4.3.0 beta (2023-04-12 r84240)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.2 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
 [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] here_1.0.1                  lintr_3.0.2                 styler_1.9.1               
 [4] DESeq2_1.40.0               SummarizedExperiment_1.30.0 Biobase_2.60.0             
 [7] MatrixGenerics_1.12.0       matrixStats_0.63.0          GenomicRanges_1.52.0       
[10] GenomeInfoDb_1.36.0         IRanges_2.34.0              S4Vectors_0.38.0           
[13] BiocGenerics_0.46.0         lubridate_1.9.2             forcats_1.0.0              
[16] stringr_1.5.0               dplyr_1.1.2                 purrr_1.0.1                
[19] readr_2.1.4                 tidyr_1.3.0                 tibble_3.2.1               
[22] ggplot2_3.4.2               tidyverse_2.0.0            

loaded via a namespace (and not attached):
 [1] tidyselect_1.2.0        R.utils_2.12.2          bitops_1.0-7           
 [4] fastmap_1.1.1           RCurl_1.98-1.12         lazyeval_0.2.2         
 [7] rex_1.2.1               digest_0.6.31           timechange_0.2.0       
[10] lifecycle_1.0.3         processx_3.8.0          magrittr_2.0.3         
[13] compiler_4.3.0          rlang_1.1.1             tools_4.3.0            
[16] utf8_1.2.3              yaml_2.3.7              knitr_1.42             
[19] xmlparsedata_1.0.5      DelayedArray_0.25.0     xml2_1.3.4             
[22] BiocParallel_1.34.0     R.cache_0.16.0          withr_2.5.0            
[25] desc_1.4.2              R.oo_1.25.0             grid_4.3.0             
[28] fansi_1.0.4             colorspace_2.1-0        scales_1.2.1           
[31] cli_3.6.1               rmarkdown_2.21          crayon_1.5.2           
[34] generics_0.1.3          remotes_2.4.2           rstudioapi_0.14        
[37] tzdb_0.3.0              zlibbioc_1.46.0         parallel_4.3.0         
[40] XVector_0.40.0          vctrs_0.6.2             Matrix_1.5-4           
[43] callr_3.7.3             hms_1.1.3               cyclocomp_1.1.0        
[46] locfit_1.5-9.7          glue_1.6.2              codetools_0.2-19       
[49] ps_1.7.4                stringi_1.7.12          gtable_0.3.3           
[52] munsell_0.5.0           pillar_1.9.0            htmltools_0.5.5        
[55] GenomeInfoDbData_1.2.10 R6_2.5.1                rprojroot_2.0.3        
[58] evaluate_0.20           lattice_0.21-8          R.methodsS3_1.8.2      
[61] Rcpp_1.0.10             xfun_0.39               pkgconfig_2.0.3 
