---
title: "Incorporating DE Results"
author: "Emma Jones"
date: "2023-05-23"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Incorporating differential expression results

The purpose of this script is to wrap in all DESeq2 significance values into my isoformswitchanalyzeR objects and subsequently, plots. This enables us to plot them all with significance values for DGE, DTE, and DTU. This script is fully dependent on scripts 01-09 and the de_analysis DESeq2 scripts 10-12.

For code review, I am tracking the amount of time that this analysis will take.

```{r proc time}
ptm <- proc.time()
```


#### Load in Packages

```{r load packages and set seed}
# load packages
suppressPackageStartupMessages({
  library(tidyverse)
  library(IsoformSwitchAnalyzeR)
  library(ComplexHeatmap)
  library(viridis)
  library(org.Mm.eg.db)
  library(gprofiler2)
  library(styler)
  library(lintr)
  library(here)
})
# set seed
set.seed(123)
```

#### Read in functions

```{r read in functions}
source("de_functions.R", local = knitr::knit_global())
```

#### Read in Data

```{r load in switchlist orf objects}
files <- list.files(here("data", "switchlist_objects", "orf_added"),
  pattern = ".Rds"
)
# loop in files
for (i in files) {
  temp <- readRDS(
    here("data", "switchlist_objects", "orf_added", i)
  )
  assign(paste0(str_sub(i, end = -14), "_orf"),
    temp,
    envir = .GlobalEnv
  )
}
```

I am adding DESeq results to the region_region switchlist first.

```{r add gene q value}
features <- region_region_switchlist_orf$isoformFeatures

order_needed <- features[, c("gene_id", "condition_1", "condition_2")]

order_needed_transcripts <- features[, c(
  "isoform_id", "condition_1",
  "condition_2"
)]
```


```{r read in deseq2 results}
files <- list.files(here("data", "deseq2_data"),
  pattern = "results.Rds"
)
# loop in files
for (i in files) {
  temp <- readRDS(
    here("data", "deseq2_data", i)
  )
  assign(paste0(str_sub(i, end = -12), "res"),
    temp,
    envir = .GlobalEnv
  )
}
```

Next, I can use a function I made for formatting deseq2 results and matching to the format needed for my switchlist.

```{r make into function}
final_gene_column <- c()

format_deseq_results("cerebellum", "cortex")
format_deseq_results("cerebellum", "hippocampus")
format_deseq_results("cerebellum", "striatum")
format_deseq_results("cortex", "hippocampus")
format_deseq_results("cortex", "striatum")

hippocampus_striatum_res <- striatum_hippocampus_res # needed to rename

format_deseq_results("hippocampus", "striatum")

region_region_switchlist_orf$isoformFeatures$gene_q_value <- final_gene_column

switchPlot(region_region_switchlist_orf,
  gene = "Gng13",
  condition1 = "cerebellum", condition2 = "cortex"
)
```

Great, so now the switchplot has DGE values. Let's add DTE values.

### Also do for transcript level

First, I need to get the DTE expression results

```{r read in deseq2 results}
files <- list.files(here("data", "deseq2_data"),
  pattern = "_transcripts_results.Rds"
)
# loop in files
for (i in files) {
  temp <- readRDS(
    here("data", "deseq2_data", i)
  )
  assign(paste0(str_sub(i, end = -24), "dte_res"),
    temp,
    envir = .GlobalEnv
  )
}
```

Next, I can use a function I made for formatting deseq2 results and matching to the format needed for my switchlist.

```{r make into function}
final_transcript_column <- c()

format_deseq_results_dte("cerebellum", "cortex")
format_deseq_results_dte("cerebellum", "hippocampus")
format_deseq_results_dte("cerebellum", "striatum")
format_deseq_results_dte("cortex", "hippocampus")
format_deseq_results_dte("cortex", "striatum")
format_deseq_results_dte("hippocampus", "striatum")

region_region_switchlist_orf$isoformFeatures$iso_q_value <-
  final_transcript_column

saveRDS(
  region_region_switchlist_orf,
  here(
    "data", "switchlist_objects",
    "de_added", "region_region_orf_de.Rds"
  )
)

switchPlot(region_region_switchlist_orf,
  gene = "Gng13",
  condition1 = "cerebellum", condition2 = "cortex"
)

switchPlotTopSwitches(region_region_switchlist_orf,
  pathToOutput = here(
    "results", "plots", "switch_plots", "de_added",
    "region_region"
  ),
  splitFunctionalConsequences = FALSE
)
```

### Incorporate DE results for region vs others comparisons

```{r region others gene level}
format_deseq_results_region(region = "cerebellum",
                            save_path = 
                              here("data", "switchlist_objects", "de_added"))

format_deseq_results_region(region = "cortex",
                            save_path = 
                              here("data", "switchlist_objects", "de_added"))

format_deseq_results_region(region = "hippocampus",
                            save_path = 
                              here("data", "switchlist_objects", "de_added"))

format_deseq_results_region(region = "striatum",
                            save_path = 
                              here("data", "switchlist_objects", "de_added"))
```

```{r region others transcript level}
format_deseq_results_region_dte(region = "cerebellum",
                            save_path = 
                              here("data", "switchlist_objects", "de_added"))

format_deseq_results_region_dte(region = "cortex",
                            save_path = 
                              here("data", "switchlist_objects", "de_added"))

format_deseq_results_region_dte(region = "hippocampus",
                            save_path = 
                              here("data", "switchlist_objects", "de_added"))

format_deseq_results_region_dte(region = "striatum",
                            save_path = 
                              here("data", "switchlist_objects", "de_added"))
```

#### Save new region other plots with signficiance values

```{r save new plots}

switchPlotTopSwitches(cerebellum_switchlist_orf_dte,
  pathToOutput = here(
    "results", "plots", "switch_plots", "de_added",
    "cerebellum"
  ),
  splitFunctionalConsequences = FALSE
)

switchPlotTopSwitches(cortex_switchlist_orf_dte,
  pathToOutput = here(
    "results", "plots", "switch_plots", "de_added",
    "cortex"
  ),
  splitFunctionalConsequences = FALSE
)

switchPlotTopSwitches(cerebellum_switchlist_orf_dte,
  pathToOutput = here(
    "results", "plots", "switch_plots", "de_added",
    "cerebellum"
  ),
  splitFunctionalConsequences = FALSE
)

switchPlotTopSwitches(hippocampus_switchlist_orf_dte,
  pathToOutput = here(
    "results", "plots", "switch_plots", "de_added",
    "hippocampus"
  ),
  splitFunctionalConsequences = FALSE
)

switchPlotTopSwitches(striatum_switchlist_orf_dte,
  pathToOutput = here(
    "results", "plots", "switch_plots", "de_added",
    "striatum"
  ),
  splitFunctionalConsequences = FALSE
)

```

### DESeq2 by sex

Now I need to pull in the results from my individual tissue comparisons by sex. I had edited those original function again to make this a bit easier.

Start with gene-level, then move to transcript-level.

```{r incorporate deseq2 results by sex for within brain regions}
format_deseq_results_region_sex(region = "cerebellum",
                            save_path = 
                              here("data", "switchlist_objects", "de_added"))

format_deseq_results_region_sex(region = "cortex",
                            save_path = 
                              here("data", "switchlist_objects", "de_added"))

format_deseq_results_region_sex(region = "hippocampus",
                            save_path = 
                              here("data", "switchlist_objects", "de_added"))

format_deseq_results_region_sex(region = "striatum",
                            save_path = 
                              here("data", "switchlist_objects", "de_added"))


```

Now can do transcript level.

```{r incorporate deseq2 results by sex for within brain regions transcript level}
format_deseq_results_region_dte_sex(region = "cerebellum",
                            save_path = 
                              here("data", "switchlist_objects", "de_added"))

format_deseq_results_region_dte_sex(region = "cortex",
                            save_path = 
                              here("data", "switchlist_objects", "de_added"))

format_deseq_results_region_dte_sex(region = "hippocampus",
                            save_path = 
                              here("data", "switchlist_objects", "de_added"))

format_deseq_results_region_dte_sex(region = "striatum",
                            save_path = 
                              here("data", "switchlist_objects", "de_added"))


```

Finally, make new plots for these sex. The only signficnat plots are for striatum anyway, so I'm only running the plotting function once.

At some point, we lost the gene names for these sex objects. I will have to re-add the gene names.

```{r plot striatum sex}

get_gene_symbols_mod(tissue = "striatum",
                     save_path = here("data", "switchlist_objects", "de_added"))

switchPlotTopSwitches(striatum_sex_switchlist_orf_dte,
  pathToOutput = here(
    "results", "plots", "switch_plots", "de_added",
    "striatum"
  ),
  splitFunctionalConsequences = FALSE
)

```

#### Clean up script

```{r tidy script}
style_file("13_incorporate_de_results.Rmd")
lint("13_incorporate_de_results.Rmd",
  linters = linters_with_defaults(
    object_length_linter = NULL,
    object_name_linter = NULL,
    object_usage_linter = NULL
  )
)
```

Also, needed to get processing time.

```{r proc time finish}
fptm <- proc.time() - ptm
fptm[3] / 60
```

This took 12 minutes to run.

#### Software versions

My software versions will be commented below.

```{r versions}
sessionInfo()
```

R version 4.3.0 beta (2023-04-12 r84240)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.2 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
 [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] grid      stats4    stats     graphics  grDevices utils     datasets  methods  
[9] base     

other attached packages:
 [1] here_1.0.1                    lintr_3.0.2                  
 [3] styler_1.9.1                  gprofiler2_0.2.1             
 [5] org.Mm.eg.db_3.17.0           viridis_0.6.2                
 [7] viridisLite_0.4.1             ComplexHeatmap_2.16.0        
 [9] IsoformSwitchAnalyzeR_1.99.17 pfamAnalyzeR_1.0.0           
[11] sva_3.48.0                    genefilter_1.82.0            
[13] mgcv_1.8-42                   nlme_3.1-162                 
[15] satuRn_1.8.0                  DEXSeq_1.46.0                
[17] RColorBrewer_1.1-3            AnnotationDbi_1.62.0         
[19] DESeq2_1.40.0                 SummarizedExperiment_1.30.0  
[21] GenomicRanges_1.52.0          GenomeInfoDb_1.36.0          
[23] IRanges_2.34.0                S4Vectors_0.38.0             
[25] MatrixGenerics_1.12.0         matrixStats_0.63.0           
[27] Biobase_2.60.0                BiocGenerics_0.46.0          
[29] BiocParallel_1.34.0           limma_3.56.0                 
[31] lubridate_1.9.2               forcats_1.0.0                
[33] stringr_1.5.0                 dplyr_1.1.2                  
[35] purrr_1.0.1                   readr_2.1.4                  
[37] tidyr_1.3.0                   tibble_3.2.1                 
[39] ggplot2_3.4.2                 tidyverse_2.0.0              

loaded via a namespace (and not attached):
  [1] splines_4.3.0                 later_1.3.0                  
  [3] BiocIO_1.10.0                 bitops_1.0-7                 
  [5] filelock_1.0.2                R.oo_1.25.0                  
  [7] rex_1.2.1                     XML_3.99-0.14                
  [9] lifecycle_1.0.3               rprojroot_2.0.3              
 [11] edgeR_3.42.0                  doParallel_1.0.17            
 [13] processx_3.8.0                lattice_0.21-8               
 [15] ensembldb_2.24.0              magrittr_2.0.3               
 [17] plotly_4.10.1                 rmarkdown_2.21               
 [19] remotes_2.4.2                 yaml_2.3.7                   
 [21] httpuv_1.6.9                  pbapply_1.7-0                
 [23] DBI_1.1.3                     zlibbioc_1.46.0              
 [25] R.cache_0.16.0                R.utils_2.12.2               
 [27] AnnotationFilter_1.24.0       RCurl_1.98-1.12              
 [29] rappdirs_0.3.3                circlize_0.4.15              
 [31] GenomeInfoDbData_1.2.10       annotate_1.78.0              
 [33] codetools_0.2-19              DelayedArray_0.25.0          
 [35] xml2_1.3.4                    tidyselect_1.2.0             
 [37] shape_1.4.6                   futile.logger_1.4.3          
 [39] locfdr_1.1-8                  BiocFileCache_2.8.0          
 [41] GenomicAlignments_1.36.0      jsonlite_1.8.4               
 [43] GetoptLong_1.0.5              ellipsis_0.3.2               
 [45] survival_3.5-5                iterators_1.0.14             
 [47] foreach_1.5.2                 tools_4.3.0                  
 [49] progress_1.2.2                Rcpp_1.0.10                  
 [51] glue_1.6.2                    gridExtra_2.3                
 [53] xfun_0.39                     withr_2.5.0                  
 [55] formatR_1.14                  BiocManager_1.30.20          
 [57] fastmap_1.1.1                 boot_1.3-28.1                
 [59] fansi_1.0.4                   callr_3.7.3                  
 [61] digest_0.6.31                 timechange_0.2.0             
 [63] R6_2.5.1                      mime_0.12                    
 [65] colorspace_2.1-0              biomaRt_2.56.0               
 [67] RSQLite_2.3.1                 R.methodsS3_1.8.2            
 [69] utf8_1.2.3                    generics_0.1.3               
 [71] data.table_1.14.8             tximeta_1.18.0               
 [73] rtracklayer_1.60.0            htmlwidgets_1.6.2            
 [75] prettyunits_1.1.1             httr_1.4.5                   
 [77] pkgconfig_2.0.3               gtable_0.3.3                 
 [79] blob_1.2.4                    hwriter_1.3.2.1              
 [81] XVector_0.40.0                htmltools_0.5.5              
 [83] geneplotter_1.78.0            ProtGenerics_1.32.0          
 [85] clue_0.3-64                   scales_1.2.1                 
 [87] cyclocomp_1.1.0               png_0.1-8                    
 [89] knitr_1.42                    lambda.r_1.2.4               
 [91] rstudioapi_0.14               tzdb_0.3.0                   
 [93] reshape2_1.4.4                rjson_0.2.21                 
 [95] curl_5.0.0                    cachem_1.0.8                 
 [97] GlobalOptions_0.1.2           BiocVersion_3.17.1           
 [99] parallel_4.3.0                desc_1.4.2                   
[101] restfulr_0.0.15               pillar_1.9.0                 
[103] vctrs_0.6.2                   promises_1.2.0.1             
[105] dbplyr_2.3.2                  xtable_1.8-4                 
[107] cluster_2.1.4                 tximport_1.28.0              
[109] evaluate_0.20                 VennDiagram_1.7.3            
[111] GenomicFeatures_1.52.0        cli_3.6.1                    
[113] locfit_1.5-9.7                compiler_4.3.0               
[115] futile.options_1.0.1          Rsamtools_2.16.0             
[117] rlang_1.1.1                   crayon_1.5.2                 
[119] ps_1.7.4                      plyr_1.8.8                   
[121] stringi_1.7.12                munsell_0.5.0                
[123] Biostrings_2.68.0             lazyeval_0.2.2               
[125] Matrix_1.5-4                  BSgenome_1.68.0              
[127] hms_1.1.3                     bit64_4.0.5                  
[129] KEGGREST_1.40.0               statmod_1.5.0                
[131] shiny_1.7.4                   interactiveDisplayBase_1.38.0
[133] AnnotationHub_3.8.0           memoise_2.0.1                
[135] bit_4.0.5                     xmlparsedata_1.0.5 
