---
title: "comparing DTU and DGE"
author: "Emma Jones"
date: "2023-06-01"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Comparing DTU, DTE, and DGE

The purpose of this script is to compare genes with differential gene expression, differential transcript usage, and differential transcript expression. It requires the eulerr package, which is in docker 1.6 and above. To create figure 2 using cowplot, you must be using docker 1.7.

For code review, I am tracking the amount of time that this analysis will take.

```{r proc time}
ptm <- proc.time()
```

#### Load in Packages

```{r load packages}
# load packages
suppressPackageStartupMessages({
  library(tidyverse)
  library(IsoformSwitchAnalyzeR)
  library(ComplexHeatmap)
  library(viridis)
  library(styler)
  library(lintr)
  library(here)
  library(eulerr)
  library(cowplot)
})

# source functions
source("de_functions.R")
```

Set color palettes

```{r get stacked bar plot colors}
# get colors

# From Paul Tol: https://personal.sron.nl/~pault/

Tol_muted <- c(
  "#88CCEE", "#44AA99", "#117733", "#332288", "#DDCC77", "#999933",
  "#CC6677", "#882255", "#AA4499", "#DDDDDD"
)

Tol_light <- c(
  "#44BB99", "#BBCC33", "#FFAABB", "#EE8866", "#EEDD88", "#77AADD",
  "#99DDFF", "#DDDDDD", "#AAAA00"
)

# my palette borrows other colors from tol_light not similar to the three I
# chose for my venn diagrams

EJ_custom_palette <- c(
  "#EEDD88", "#77AADD", "#FFAABB", "#5D69B1", "#EE8866",
  "#52BCA3", "#99C945", "#99DDFF", "#DDDDDD"
)
```

### Load in Data

First, we want to load in the DGE data

```{r load in DGE data}
# DGE
dge_files <- list.files(here("results", "de_genes"))
# loop in files
for (i in dge_files) {
  temp <- read.csv(
    here("results", "de_genes", i),
    header = FALSE
  )
  genes <- temp$V1
  assign(paste0(str_sub(i, end = -5), "_dge_genes"),
    genes,
    envir = .GlobalEnv
  )
}
```

Next, we need to pull in the transcript counts files to get gene name information for all transcripts and load in the DTE lists.

```{r pull in DTE data}
transcript_counts <- read.table(
  here("data", "nextflow", "bambu", "counts_transcript.txt"),
  header = TRUE
)

tx_gene_names <- transcript_counts[, 1:2]

# DTE
dte_files <- list.files(here("results", "de_transcripts"))
# loop in files
for (i in dte_files) {
  temp <- read.csv(
    here("results", "de_transcripts", i),
    header = FALSE
  )
  transcripts <- temp$V1
  genes <- tx_gene_names$GENEID[tx_gene_names$TXNAME %in% transcripts]
  assign(paste0(str_sub(i, end = -5), "_dte_genes"),
    genes,
    envir = .GlobalEnv
  )
}
```

Finally, load in the DTU data for comparisons. It is important to note here that because some of it is stored in lists, I need two different ways of importing the data.

```{r load in DTU data}
# start with region region
region_region_switchlist <- readRDS(
  here("data", "switchlist_objects", "de_added", "region_region_orf_de.Rds")
)
sig_region_region <- dplyr::filter(
  region_region_switchlist$isoformFeatures,
  abs(dIF) > 0.1 & isoform_switch_q_value < 0.05
)
region_region_dtu_genes <- unique(sig_region_region$gene_id)

# region others
region_all_switchlist_list <- readRDS(
  here(
    "data", "switchlist_objects", "de_added",
    "region_all_switchlist_list_orf_de.Rds"
  )
)
for (i in names(region_all_switchlist_list)) {
  sig_temp <- dplyr::filter(
    region_all_switchlist_list[[i]]$isoformFeatures,
    abs(dIF) > 0.1 & isoform_switch_q_value < 0.05
  )
  genes <- sig_temp$gene_id
  genes <- unique(genes)
  assign(paste0(i, "_dtu_genes"),
    genes,
    envir = .GlobalEnv
  )
}

# region sex
region_sex_switchlist_list <- readRDS(
  here(
    "data", "switchlist_objects", "de_added",
    "region_sex_switchlist_list_orf_de.Rds"
  )
)
for (i in names(region_sex_switchlist_list)) {
  sig_temp <- dplyr::filter(
    region_sex_switchlist_list[[i]]$isoformFeatures,
    abs(dIF) > 0.1 & isoform_switch_q_value < 0.05
  )
  genes <- sig_temp$gene_id
  genes <- unique(genes)
  assign(paste0(i, "_sex_dtu_genes"),
    genes,
    envir = .GlobalEnv
  )
}
```

Next, we can pull what we need from each file to make single comparison lists.

```{r pull out comparisons}
# cerebellum cortex
cerebellum_cortex_subset <- filter(
  sig_region_region,
  condition_1 == "cerebellum" & condition_2 == "cortex"
)

cerebellum_cortex_dtu_genes <- cerebellum_cortex_subset$gene_id

# cerebellum hippocampus
cerebellum_hippocampus_subset <- filter(
  sig_region_region,
  condition_1 == "cerebellum" & condition_2 == "hippocampus"
)

cerebellum_hippocampus_dtu_genes <- cerebellum_hippocampus_subset$gene_id

# cerebellum striatum
cerebellum_striatum_subset <- filter(
  sig_region_region,
  condition_1 == "cerebellum" & condition_2 == "striatum"
)

cerebellum_striatum_dtu_genes <- cerebellum_striatum_subset$gene_id

# cortex hippocampus
cortex_hippocampus_subset <- filter(
  sig_region_region,
  condition_1 == "cortex" & condition_2 == "hippocampus"
)

cortex_hippocampus_dtu_genes <- cortex_hippocampus_subset$gene_id

# cortex striatum
cortex_striatum_subset <- filter(
  sig_region_region,
  condition_1 == "cortex" & condition_2 == "striatum"
)

cortex_striatum_dtu_genes <- cortex_striatum_subset$gene_id

# hippocampus striatum
hippocampus_striatum_subset <- filter(
  sig_region_region,
  condition_1 == "hippocampus" & condition_2 == "striatum"
)

hippocampus_striatum_dtu_genes <- hippocampus_striatum_subset$gene_id
```

Now that everything is extracted, we need to make it into lists. I'll start with region-region direct comparisons.

```{r prep region region lists}
cerebellum_cortex_list <- list(
  `cerebellum cortex DGE` = cerebellum_cortex_dge_genes,
  `cerebellum cortex DTE` = cerebellum_cortex_dte_genes,
  `cerebellum cortex DTU` = cerebellum_cortex_dtu_genes
)

cerebellum_hippocampus_list <- list(
  `cerebellum hippocampus DGE` = cerebellum_hippocampus_dge_genes,
  `cerebellum hippocampus DTE` = cerebellum_hippocampus_dte_genes,
  `cerebellum hippocampus DTU` = cerebellum_hippocampus_dtu_genes
)

cerebellum_striatum_list <- list(
  `cerebellum striatum DGE` = cerebellum_striatum_dge_genes,
  `cerebellum striatum DTE` = cerebellum_striatum_dte_genes,
  `cerebellum striatum DTU` = cerebellum_striatum_dtu_genes
)

cortex_hippocampus_list <- list(
  `cortex hippocampus DGE` = cortex_hippocampus_dge_genes,
  `cortex hippocampus DTE` = cortex_hippocampus_dte_genes,
  `cortex hippocampus DTU` = cortex_hippocampus_dtu_genes
)

cortex_striatum_list <- list(
  `cortex striatum DGE` = cortex_striatum_dge_genes,
  `cortex striatum DTE` = cortex_striatum_dte_genes,
  `cortex striatum DTU` = cortex_striatum_dtu_genes
)

hippocampus_striatum_list <- list(
  `hippocampus striatum DGE` = striatum_hippocampus_dge_genes,
  `hippocampus striatum DTE` = striatum_hippocampus_dte_genes,
  `hippocampus striatum DTU` = hippocampus_striatum_dtu_genes
)
```

Next, we can do for the individual regions.

```{r prep single region data}
# make results into a list
cerebellum_list <- list(
  `cerebellum DGE` = cerebellum_dge_genes,
  `cerebellum DTE` = cerebellum_dte_genes,
  `cerebellum DTU` = cerebellum_dtu_genes
)

cortex_list <- list(
  `cortex DGE` = cortex_dge_genes,
  `cortex DTE` = cortex_dte_genes,
  `cortex DTU` = cortex_dtu_genes
)

hippocampus_list <- list(
  `hippocampus DGE` = hippocampus_dge_genes,
  `hippocampus DTE` = hippocampus_dte_genes,
  `hippocampus DTU` = hippocampus_dtu_genes
)

striatum_list <- list(
  `striatum DGE` = striatum_dge_genes,
  `striatum DTE` = striatum_dte_genes,
  `striatum DTU` = striatum_dtu_genes
)
```

Finally, we want to compare to within brain regions across sex.

```{r make list for comparing within brain region across sex}
# make results into a list
cerebellum_sex_list <- list(
  `DGE` = cerebellum_sex_dge_genes,
  `DTE` = cerebellum_sex_dte_genes,
  `DTU` = cerebellum_sex_dtu_genes
)

cortex_sex_list <- list(
  `DGE` = cortex_sex_dge_genes,
  `DTE` = cortex_sex_dte_genes,
  `DTU` = cortex_sex_dtu_genes
)

hippocampus_sex_list <- list(
  `DGE` = hippocampus_sex_dge_genes,
  `DTE` = hippocampus_sex_dte_genes,
  `DTU` = hippocampus_sex_dtu_genes
)

striatum_sex_list <- list(
  `DGE` = striatum_sex_dge_genes,
  `DTE` = striatum_sex_dte_genes,
  `DTU` = striatum_sex_dtu_genes
)
```

### Plotting DGE, DTE, and DTU overlap

I created a make comparision plots function to create venn diagrams and upset plots for overlapping genes.

```{r plot comparisions for region region}
make_comparison_plots(cerebellum_cortex_list,
  save_path = here("results", "plots")
)

make_comparison_plots(cerebellum_hippocampus_list,
  save_path = here("results", "plots")
)

make_comparison_plots(cerebellum_striatum_list,
  save_path = here("results", "plots")
)

make_comparison_plots(cortex_hippocampus_list,
  save_path = here("results", "plots")
)

make_comparison_plots(cortex_striatum_list,
  save_path = here("results", "plots")
)

make_comparison_plots(hippocampus_striatum_list,
  save_path = here("results", "plots")
)
```

I also want to run this function on the single region comparisons.

```{r run plotting function on single brain regions}
make_comparison_plots(cerebellum_list, save_path = here("results", "plots"))

make_comparison_plots(cortex_list, save_path = here("results", "plots"))

make_comparison_plots(hippocampus_list, save_path = here("results", "plots"))

make_comparison_plots(striatum_list, save_path = here("results", "plots"))
```

And finally, plot comparisons within brain regions across sex.

```{r plot comps within region across sex}
make_comparison_plots(cerebellum_sex_list,
  save_path = here("results", "plots")
)

make_comparison_plots(cortex_sex_list,
  save_path = here("results", "plots")
)

make_comparison_plots(hippocampus_sex_list,
  save_path = here("results", "plots")
)

make_comparison_plots(striatum_sex_list,
  save_path = here("results", "plots")
)
```

I also want to look at the overlap of DGE and DTE with themselves, as compared to the DTU in script 5. This code allows us to make lists, the corresponding matrices and combination matrices required for the upset plots. I'll start with the pairwise genes identified, the largest gene lists.

First create the lists.

```{r create lists of DGE and DTE overlap}
dge_main_list <- list(
  `Cerebellum vs. Cortex` = cerebellum_cortex_dge_genes,
  `Cerebellum vs. Hippocampus` = cerebellum_hippocampus_dge_genes,
  `Cerebellum vs. Striatum` = cerebellum_striatum_dge_genes,
  `Cortex vs. Hippocampus` = cortex_hippocampus_dge_genes,
  `Cortex vs. Striatum` = cortex_striatum_dge_genes,
  `Hippocampus vs. Striatum` = striatum_hippocampus_dge_genes
)

dte_main_list <- list(
  `Cerebellum vs. Cortex` = cerebellum_cortex_dte_genes,
  `Cerebellum vs. Hippocampus` = cerebellum_hippocampus_dte_genes,
  `Cerebellum vs. Striatum` = cerebellum_striatum_dte_genes,
  `Cortex vs. Hippocampus` = cortex_hippocampus_dte_genes,
  `Cortex vs. Striatum` = cortex_striatum_dte_genes,
  `Hippocampus vs. Striatum` = striatum_hippocampus_dte_genes
)

dtu_main_list <- list(
  `Cerebellum vs. Cortex` = cerebellum_cortex_dtu_genes,
  `Cerebellum vs. Hippocampus` = cerebellum_hippocampus_dtu_genes,
  `Cerebellum vs. Striatum` = cerebellum_striatum_dtu_genes,
  `Cortex vs. Hippocampus` = cortex_hippocampus_dtu_genes,
  `Cortex vs. Striatum` = cortex_striatum_dtu_genes,
  `Hippocampus vs. Striatum` = hippocampus_striatum_dtu_genes
)
```

Also, I want to make lists of all dge and dtu single region comparisons.

```{r make lists of all dge and dtu single region comparisons}
dge_single_list <- list(
  `Cerebellum` = cerebellum_dge_genes,
  `Hippocampus` = hippocampus_dge_genes,
  `Cortex` = cortex_dge_genes,
  `Striatum` = striatum_dge_genes
)

dte_single_list <- list(
  `Cerebellum` = cerebellum_dte_genes,
  `Hippocampus` = hippocampus_dte_genes,
  `Cortex` = cortex_dte_genes,
  `Striatum` = striatum_dte_genes
)
```

And, I want to make make lists of all dge and dtu sex comparisons.

```{r make lists of all dge and dtu sex comparisons}
dge_sex_list <- list(
  `Cerebellum` = cerebellum_sex_dge_genes,
  `Hippocampus` = hippocampus_sex_dge_genes,
  `Cortex` = cortex_sex_dge_genes,
  `Striatum` = striatum_sex_dge_genes
)

dte_sex_list <- list(
  `Cerebellum` = cerebellum_sex_dte_genes,
  `Hippocampus` = hippocampus_sex_dte_genes,
  `Cortex` = cortex_sex_dte_genes,
  `Striatum` = striatum_sex_dte_genes
)
```

Once we have made these lists, we can use them to get some basic stats on how many genes overlap and how many are unique. This is mainly for the manuscript.

```{r get pairwise comparisions stats}
# aggregate all dge genes
all_pairwise_dge_genes <- unique(c(
  cerebellum_cortex_dge_genes, cerebellum_hippocampus_dge_genes,
  cerebellum_striatum_dge_genes, cortex_hippocampus_dge_genes,
  cortex_striatum_dge_genes,
  striatum_hippocampus_dge_genes
))

# aggregate all dte genes
all_pairwise_dte_genes <- unique(c(
  cerebellum_cortex_dte_genes, cerebellum_hippocampus_dte_genes,
  cerebellum_striatum_dte_genes, cortex_hippocampus_dte_genes,
  cortex_striatum_dte_genes,
  striatum_hippocampus_dte_genes
))

# how many intersect?
length(intersect(all_pairwise_dge_genes, all_pairwise_dte_genes))

# all dge or dte genes
all_dge_dte <- union(all_pairwise_dge_genes, all_pairwise_dte_genes)

# aggregate all pairwise dtu genes
all_pairwise_dtu_genes <- unique(c(
  cerebellum_cortex_dtu_genes, cerebellum_hippocampus_dtu_genes,
  cerebellum_striatum_dtu_genes, cortex_hippocampus_dtu_genes,
  cortex_striatum_dtu_genes,
  hippocampus_striatum_dtu_genes
))

# how many pairwise dtu genes overlap with dge
length(intersect(all_pairwise_dge_genes, all_pairwise_dtu_genes))

# and what percent?
length(intersect(all_pairwise_dge_genes, all_pairwise_dtu_genes)) /
  length(all_pairwise_dtu_genes)

# how many pairwise dtu genes overlap with dte
length(intersect(all_pairwise_dte_genes, all_pairwise_dtu_genes))

# and what percent?
length(intersect(all_pairwise_dte_genes, all_pairwise_dtu_genes)) /
  length(all_pairwise_dtu_genes)

# how many pairwise dtu genes overlap with dge or dte
length(intersect(all_dge_dte, all_pairwise_dtu_genes))

# and what percent?
length(intersect(all_dge_dte, all_pairwise_dtu_genes)) /
  length(all_pairwise_dtu_genes)

# leaving how many unidentified?
length(all_pairwise_dtu_genes) - length(intersect(
  all_dge_dte,
  all_pairwise_dtu_genes
))

# and what percent?
(length(all_pairwise_dtu_genes) - length(intersect(
  all_dge_dte,
  all_pairwise_dtu_genes
))) /
  length(all_pairwise_dtu_genes)
```

### Creating More UpSet Plots!

Then create the upset plots from combination matrices. 

First for DGE.

```{r create upset plots for examining DGE overlap}
# create matrix
dge_overlap_list_mat <- ComplexHeatmap::list_to_matrix(dge_main_list)

# generate combination matrix
dge_overlap_list_comb_mat <- make_comb_mat(dge_overlap_list_mat)

# make UpSet plot (ComplexHeatmap)
png(here("results", "plots", "upset_plots", "region_region_dge_upset.png"),
  width = 10, height = 4, units = "in", res = 300
)
dge_pairwise_upset_plot <- UpSet(dge_overlap_list_comb_mat,
  comb_order = order(-comb_size(dge_overlap_list_comb_mat)),
  comb_col = "#5D69B1",
  top_annotation = upset_top_annotation(dge_overlap_list_comb_mat,
    add_numbers = TRUE
  ),
  right_annotation = upset_right_annotation(dge_overlap_list_comb_mat,
    add_numbers = TRUE
  )
)
draw(dge_pairwise_upset_plot)
dev.off()

dge_pairwise_upset_plot
```

Region others

```{r do this for region others}
# create matrix
dge_overlap_list_mat <- ComplexHeatmap::list_to_matrix(dge_single_list)

# generate combination matrix
dge_overlap_list_comb_mat <- make_comb_mat(dge_overlap_list_mat)


# make UpSet plot (ComplexHeatmap)
pdf(here("results", "plots", "upset_plots", "region_others_dge_upset.pdf"),
  width = 12, height = 6
)
dge_single_upset_plot <- UpSet(dge_overlap_list_comb_mat,
  comb_order = order(-comb_size(dge_overlap_list_comb_mat)),
  comb_col = viridis(6)[comb_degree(dge_overlap_list_comb_mat)],
  top_annotation = upset_top_annotation(dge_overlap_list_comb_mat,
    add_numbers = TRUE
  ),
  right_annotation = upset_right_annotation(dge_overlap_list_comb_mat,
    add_numbers = TRUE
  )
)
draw(dge_single_upset_plot)
dev.off()

dge_single_upset_plot
```
And region sex

```{r do this for region others}
# create matrix
dge_overlap_list_mat <- ComplexHeatmap::list_to_matrix(dge_sex_list)

# generate combination matrix
dge_overlap_list_comb_mat <- make_comb_mat(dge_overlap_list_mat)


# make UpSet plot (ComplexHeatmap)
pdf(here("results", "plots", "upset_plots", "region_sex_dge_upset.pdf"),
  width = 12, height = 6
)
dge_sex_upset_plot <- UpSet(dge_overlap_list_comb_mat,
  comb_order = order(-comb_size(dge_overlap_list_comb_mat)),
  comb_col = viridis(6)[comb_degree(dge_overlap_list_comb_mat)],
  top_annotation = upset_top_annotation(dge_overlap_list_comb_mat,
    add_numbers = TRUE
  ),
  right_annotation = upset_right_annotation(dge_overlap_list_comb_mat,
    add_numbers = TRUE
  )
)
draw(dge_sex_upset_plot)
dev.off()

dge_sex_upset_plot
```
Then, for DTE.

```{r create upsetplots for DTE overlap}
# create matrix
dte_overlap_list_mat <- ComplexHeatmap::list_to_matrix(dte_main_list)

# generate combination matrix
dte_overlap_list_comb_mat <- make_comb_mat(dte_overlap_list_mat)

# make UpSet plot (ComplexHeatmap)
png(here("results", "plots", "upset_plots", "region_region_dte_upset.png"),
  width = 10, height = 4, units = "in", res = 300
)
dte_pairwise_upset_plot <- UpSet(dte_overlap_list_comb_mat,
  comb_order = order(-comb_size(dte_overlap_list_comb_mat)),
  comb_col = "#52BCA3",
  top_annotation = upset_top_annotation(dte_overlap_list_comb_mat,
    add_numbers = TRUE
  ),
  right_annotation = upset_right_annotation(dte_overlap_list_comb_mat,
    add_numbers = TRUE
  )
)
draw(dte_pairwise_upset_plot)
dev.off()

dte_pairwise_upset_plot
```

I can also do this for region other and region sex, just as above.

Region others

```{r do this for region others}
# create matrix
dte_overlap_list_mat <- ComplexHeatmap::list_to_matrix(dte_single_list)

# generate combination matrix
dte_overlap_list_comb_mat <- make_comb_mat(dte_overlap_list_mat)


# make UpSet plot (ComplexHeatmap)
pdf(here("results", "plots", "upset_plots", "region_others_dte_upset.pdf"),
  width = 12, height = 6
)
dte_single_upset_plot <- UpSet(dte_overlap_list_comb_mat,
  comb_order = order(-comb_size(dte_overlap_list_comb_mat)),
  comb_col = viridis(6)[comb_degree(dte_overlap_list_comb_mat)],
  top_annotation = upset_top_annotation(dte_overlap_list_comb_mat,
    add_numbers = TRUE
  ),
  right_annotation = upset_right_annotation(dte_overlap_list_comb_mat,
    add_numbers = TRUE
  )
)
draw(dte_single_upset_plot)
dev.off()

dte_single_upset_plot
```
And region sex

```{r do this for region others}
# create matrix
dte_overlap_list_mat <- ComplexHeatmap::list_to_matrix(dte_sex_list)

# generate combination matrix
dte_overlap_list_comb_mat <- make_comb_mat(dte_overlap_list_mat)


# make UpSet plot (ComplexHeatmap)
pdf(here("results", "plots", "upset_plots", "region_sex_dte_upset.pdf"),
  width = 12, height = 6
)
dte_sex_upset_plot <- UpSet(dte_overlap_list_comb_mat,
  comb_order = order(-comb_size(dte_overlap_list_comb_mat)),
  comb_col = viridis(6)[comb_degree(dte_overlap_list_comb_mat)],
  top_annotation = upset_top_annotation(dte_overlap_list_comb_mat,
    add_numbers = TRUE
  ),
  right_annotation = upset_right_annotation(dte_overlap_list_comb_mat,
    add_numbers = TRUE
  )
)
draw(dte_sex_upset_plot)
dev.off()

dte_sex_upset_plot
```

Finally, do this for the DTU. We did this already, but this version is actually for the manuscript figure.

```{r create upset plots for examining DGE overlap}
# create matrix
dtu_overlap_list_mat <- ComplexHeatmap::list_to_matrix(dtu_main_list)

# generate combination matrix
dtu_overlap_list_comb_mat <- make_comb_mat(dtu_overlap_list_mat)

# make UpSet plot (ComplexHeatmap)
png(here("results", "plots", "upset_plots", "region_region_dtu_upset.png"),
  width = 10, height = 4, units = "in", res = 300
)
dtu_pairwise_upset_plot <- UpSet(dtu_overlap_list_comb_mat,
  comb_order = order(-comb_size(dtu_overlap_list_comb_mat)),
  comb_col = "#99C945",
  top_annotation = upset_top_annotation(dtu_overlap_list_comb_mat,
    add_numbers = TRUE
  ),
  right_annotation = upset_right_annotation(dtu_overlap_list_comb_mat,
    add_numbers = TRUE
  )
)
draw(dtu_pairwise_upset_plot)
dev.off()

dtu_pairwise_upset_plot
```

### Stacked bar plots

One way to show numbers of genes is stacked bar plots, though this will not show the overlap. Maybe if I make a category for each in the stack it will be ok? So for each brain region or pairwise comparision, there would be 7 possible combinations.

For simplicity, I'll do this for only 4 brain regions first.

```{r stacked bar plots}
stacked_barplot <- data.frame(
  "brain_region" = c(
    rep("cerebellum", 3), rep("cortex", 3),
    rep("hippocampus", 3), rep("striatum", 3)
  ),
  "analysis" = rep(c("DGE", "DTE", "DTU"), 4),
  "n_genes" = c(
    length(cerebellum_dge_genes),
    length(cerebellum_dte_genes),
    length(cerebellum_dtu_genes),
    length(cortex_dge_genes),
    length(cortex_dte_genes),
    length(cortex_dtu_genes),
    length(hippocampus_dge_genes),
    length(hippocampus_dte_genes),
    length(hippocampus_dtu_genes),
    length(striatum_dge_genes),
    length(striatum_dte_genes),
    length(striatum_dtu_genes)
  )
)

ggplot(stacked_barplot, aes(fill = analysis, y = n_genes, x = brain_region)) +
  geom_bar(position = "stack", stat = "identity")

ggplot(stacked_barplot, aes(fill = analysis, y = n_genes, x = brain_region)) +
  geom_bar(position = "fill", stat = "identity")

ggsave(here("results/plots/bar_plots/stacked_region_others.png"),
  width = 6, height = 4
)
```

Next, we can create the dataframe for stacked bar plot with overlap.

```{r create stacked barplot}
stacked_barplot <- data.frame(
  "brain_region" = c(
    rep("cerebellum", 7), rep("cortex", 7),
    rep("hippocampus", 7), rep("striatum", 7)
  ),
  "analysis" = rep(c(
    "DGE only", "DTE only", "DTU only", "DGE and DTE",
    "DGE and DTU", "DTE and DTU", "all"
  ), 4),
  "n_genes" = c(
    # dge only
    length(setdiff(
      cerebellum_dge_genes,
      union(
        intersect(cerebellum_dge_genes, cerebellum_dte_genes),
        intersect(cerebellum_dge_genes, cerebellum_dtu_genes)
      )
    )),
    # dte only
    length(setdiff(
      cerebellum_dte_genes,
      union(
        intersect(cerebellum_dge_genes, cerebellum_dte_genes),
        intersect(cerebellum_dte_genes, cerebellum_dtu_genes)
      )
    )),
    # dtu only
    length(setdiff(
      cerebellum_dtu_genes,
      union(
        intersect(cerebellum_dge_genes, cerebellum_dtu_genes),
        intersect(cerebellum_dte_genes, cerebellum_dtu_genes)
      )
    )),
    # dge and dte
    length(intersect(
      cerebellum_dge_genes, cerebellum_dte_genes
    )) -
      length(intersect(
        intersect(cerebellum_dge_genes, cerebellum_dte_genes),
        cerebellum_dtu_genes
      )),
    # dge and dtu
    length(intersect(cerebellum_dge_genes, cerebellum_dtu_genes)) -
      length(intersect(
        intersect(cerebellum_dge_genes, cerebellum_dte_genes),
        cerebellum_dtu_genes
      )),
    # dte and dtu
    length(intersect(cerebellum_dtu_genes, cerebellum_dte_genes)) -
      length(intersect(
        intersect(cerebellum_dge_genes, cerebellum_dte_genes),
        cerebellum_dtu_genes
      )),
    # all
    length(intersect(
      intersect(cerebellum_dge_genes, cerebellum_dte_genes),
      cerebellum_dtu_genes
    )),

    # dge only
    length(setdiff(cortex_dge_genes, union(intersect(
      cortex_dge_genes,
      cortex_dte_genes
    ), intersect(cortex_dge_genes, cortex_dtu_genes)))),
    # dte only
    length(setdiff(cortex_dte_genes, union(intersect(
      cortex_dge_genes,
      cortex_dte_genes
    ), intersect(cortex_dte_genes, cortex_dtu_genes)))),
    # dtu only
    length(setdiff(cortex_dtu_genes, union(intersect(
      cortex_dge_genes,
      cortex_dtu_genes
    ), intersect(cortex_dte_genes, cortex_dtu_genes)))),
    # dge and dte
    length(intersect(cortex_dge_genes, cortex_dte_genes)) - length(intersect(
      intersect(cortex_dge_genes, cortex_dte_genes),
      cortex_dtu_genes
    )),
    # dge and dtu
    length(intersect(cortex_dge_genes, cortex_dtu_genes)) - length(intersect(
      intersect(cortex_dge_genes, cortex_dte_genes),
      cortex_dtu_genes
    )),
    # dte and dtu
    length(intersect(cortex_dtu_genes, cortex_dte_genes)) - length(intersect(
      intersect(cortex_dge_genes, cortex_dte_genes),
      cortex_dtu_genes
    )),
    # all
    length(intersect(
      intersect(cortex_dge_genes, cortex_dte_genes),
      cortex_dtu_genes
    )),

    # dge only
    length(setdiff(hippocampus_dge_genes, union(
      intersect(hippocampus_dge_genes, hippocampus_dte_genes),
      intersect(hippocampus_dge_genes, hippocampus_dtu_genes)
    ))),
    # dte only
    length(setdiff(hippocampus_dte_genes, union(
      intersect(hippocampus_dge_genes, hippocampus_dte_genes),
      intersect(hippocampus_dte_genes, hippocampus_dtu_genes)
    ))),
    # dtu only
    length(setdiff(hippocampus_dtu_genes, union(
      intersect(hippocampus_dge_genes, hippocampus_dtu_genes),
      intersect(hippocampus_dte_genes, hippocampus_dtu_genes)
    ))),
    # dge and dte
    length(intersect(hippocampus_dge_genes, hippocampus_dte_genes)) -
      length(intersect(
        intersect(hippocampus_dge_genes, hippocampus_dte_genes),
        hippocampus_dtu_genes
      )),
    # dge and dtu
    length(intersect(hippocampus_dge_genes, hippocampus_dtu_genes)) -
      length(intersect(
        intersect(hippocampus_dge_genes, hippocampus_dte_genes),
        hippocampus_dtu_genes
      )),
    # dte and dtu
    length(intersect(hippocampus_dtu_genes, hippocampus_dte_genes)) -
      length(intersect(
        intersect(hippocampus_dge_genes, hippocampus_dte_genes),
        hippocampus_dtu_genes
      )),
    # all
    length(intersect(
      intersect(hippocampus_dge_genes, hippocampus_dte_genes),
      hippocampus_dtu_genes
    )),

    # dge only
    length(setdiff(striatum_dge_genes, union(
      intersect(
        striatum_dge_genes, striatum_dte_genes
      ),
      intersect(
        striatum_dge_genes, striatum_dtu_genes
      )
    ))),
    # dte only
    length(setdiff(striatum_dte_genes, union(
      intersect(
        striatum_dge_genes, striatum_dte_genes
      ),
      intersect(
        striatum_dte_genes, striatum_dtu_genes
      )
    ))),
    # dtu only
    length(setdiff(striatum_dtu_genes, union(
      intersect(
        striatum_dge_genes, striatum_dtu_genes
      ),
      intersect(
        striatum_dte_genes, striatum_dtu_genes
      )
    ))),
    # dge and dte
    length(intersect(striatum_dge_genes, striatum_dte_genes)) -
      length(intersect(
        intersect(striatum_dge_genes, striatum_dte_genes),
        striatum_dtu_genes
      )),
    # dge and dtu
    length(intersect(striatum_dge_genes, striatum_dtu_genes)) -
      length(intersect(
        intersect(striatum_dge_genes, striatum_dte_genes),
        striatum_dtu_genes
      )),
    # dte and dtu
    length(intersect(striatum_dtu_genes, striatum_dte_genes)) -
      length(intersect(
        intersect(striatum_dge_genes, striatum_dte_genes),
        striatum_dtu_genes
      )),
    # all
    length(intersect(
      intersect(striatum_dge_genes, striatum_dte_genes),
      striatum_dtu_genes
    ))
  )
)
```

Now that the dataframe has been created, we can plot.

```{r plot stacked barplot for unpaired}
ggplot(stacked_barplot, aes(fill = analysis, y = n_genes, x = brain_region)) +
  geom_bar(position = "stack", stat = "identity")

ggplot(stacked_barplot, aes(fill = analysis, y = n_genes, x = brain_region)) +
  geom_bar(position = "fill", stat = "identity") +
  scale_fill_manual(values = EJ_custom_palette) +
  xlab("Brain Region") +
  ylab("Proportion of Total Genes") +
  theme_minimal()

ggsave(here("results/plots/bar_plots/stacked_region_others_overlap.png"),
  width = 6, height = 4
)
```

This is a neat idea in theory but in practice not so sure about it because of all the arguments and lengths needed to manually calculate. there should be a way to simplify this the same way an upset plot works. and at that point, it just is an upset plot.

```{r create paired stacked barplot dataframe}
stacked_barplot_paired <- data.frame(
  "brain_regions" = c(
    rep("Cerebellum vs. Cortex", 7),
    rep("Cerebellum vs. Hippocampus", 7),
    rep("Cerebellum vs. Striatum", 7),
    rep("Cortex vs. Hippocampus", 7),
    rep("Cortex vs. Striatum", 7),
    rep("Hippocampus vs. Striatum", 7)
  ),
  "Analysis" = rep(c(
    "DGE only", "DTE only", "DTU only", "DGE and DTE",
    "DGE and DTU", "DTE and DTU", "all"
  ), 6),
  "n_genes" = c(
    # dge only
    length(setdiff(
      cerebellum_cortex_dge_genes,
      union(
        intersect(
          cerebellum_cortex_dge_genes,
          cerebellum_cortex_dte_genes
        ),
        intersect(
          cerebellum_cortex_dge_genes,
          cerebellum_cortex_dtu_genes
        )
      )
    )),
    # dte only
    length(setdiff(
      cerebellum_cortex_dte_genes,
      union(
        intersect(
          cerebellum_cortex_dge_genes,
          cerebellum_cortex_dte_genes
        ),
        intersect(
          cerebellum_cortex_dte_genes,
          cerebellum_cortex_dtu_genes
        )
      )
    )),
    # dtu only
    length(setdiff(
      cerebellum_cortex_dtu_genes,
      union(
        intersect(
          cerebellum_cortex_dge_genes,
          cerebellum_cortex_dtu_genes
        ),
        intersect(
          cerebellum_cortex_dte_genes,
          cerebellum_cortex_dtu_genes
        )
      )
    )),
    # dge and dte
    length(intersect(
      cerebellum_cortex_dge_genes,
      cerebellum_cortex_dte_genes
    )) -
      length(intersect(
        intersect(
          cerebellum_cortex_dge_genes,
          cerebellum_cortex_dte_genes
        ),
        cerebellum_cortex_dtu_genes
      )),
    # dge and dtu
    length(intersect(
      cerebellum_cortex_dge_genes,
      cerebellum_cortex_dtu_genes
    )) -
      length(intersect(
        intersect(
          cerebellum_cortex_dge_genes,
          cerebellum_cortex_dte_genes
        ),
        cerebellum_cortex_dtu_genes
      )),
    # dte and dtu
    length(intersect(
      cerebellum_cortex_dtu_genes,
      cerebellum_cortex_dte_genes
    )) -
      length(intersect(
        intersect(
          cerebellum_cortex_dge_genes,
          cerebellum_cortex_dte_genes
        ),
        cerebellum_cortex_dtu_genes
      )),
    # all
    length(intersect(
      intersect(
        cerebellum_cortex_dge_genes,
        cerebellum_cortex_dte_genes
      ),
      cerebellum_cortex_dtu_genes
    )),

    # dge only
    length(setdiff(
      cerebellum_hippocampus_dge_genes,
      union(
        intersect(
          cerebellum_hippocampus_dge_genes,
          cerebellum_hippocampus_dte_genes
        ),
        intersect(
          cerebellum_hippocampus_dge_genes,
          cerebellum_hippocampus_dtu_genes
        )
      )
    )),
    # dte only
    length(setdiff(
      cerebellum_hippocampus_dte_genes,
      union(
        intersect(
          cerebellum_hippocampus_dge_genes,
          cerebellum_hippocampus_dte_genes
        ),
        intersect(
          cerebellum_hippocampus_dte_genes,
          cerebellum_hippocampus_dtu_genes
        )
      )
    )),
    # dtu only
    length(setdiff(
      cerebellum_hippocampus_dtu_genes,
      union(
        intersect(
          cerebellum_hippocampus_dge_genes,
          cerebellum_hippocampus_dtu_genes
        ),
        intersect(
          cerebellum_hippocampus_dte_genes,
          cerebellum_hippocampus_dtu_genes
        )
      )
    )),
    # dge and dte
    length(intersect(
      cerebellum_hippocampus_dge_genes,
      cerebellum_hippocampus_dte_genes
    )) - length(intersect(
      intersect(
        cerebellum_hippocampus_dge_genes,
        cerebellum_hippocampus_dte_genes
      ),
      cerebellum_hippocampus_dtu_genes
    )),
    # dge and dtu
    length(intersect(
      cerebellum_hippocampus_dge_genes,
      cerebellum_hippocampus_dtu_genes
    )) - length(intersect(
      intersect(
        cerebellum_hippocampus_dge_genes,
        cerebellum_hippocampus_dte_genes
      ),
      cerebellum_hippocampus_dtu_genes
    )),
    # dte and dtu
    length(intersect(
      cerebellum_hippocampus_dtu_genes,
      cerebellum_hippocampus_dte_genes
    )) - length(intersect(
      intersect(
        cerebellum_hippocampus_dge_genes,
        cerebellum_hippocampus_dte_genes
      ),
      cerebellum_hippocampus_dtu_genes
    )),
    # all
    length(intersect(
      intersect(
        cerebellum_hippocampus_dge_genes,
        cerebellum_hippocampus_dte_genes
      ),
      cerebellum_hippocampus_dtu_genes
    )),

    # dge only
    length(setdiff(
      cerebellum_striatum_dge_genes,
      union(
        intersect(
          cerebellum_striatum_dge_genes,
          cerebellum_striatum_dte_genes
        ),
        intersect(
          cerebellum_striatum_dge_genes,
          cerebellum_striatum_dtu_genes
        )
      )
    )),
    # dte only
    length(setdiff(
      cerebellum_striatum_dte_genes,
      union(
        intersect(
          cerebellum_striatum_dge_genes,
          cerebellum_striatum_dte_genes
        ),
        intersect(
          cerebellum_striatum_dte_genes,
          cerebellum_striatum_dtu_genes
        )
      )
    )),
    # dtu only
    length(setdiff(
      cerebellum_striatum_dtu_genes,
      union(
        intersect(
          cerebellum_striatum_dge_genes,
          cerebellum_striatum_dtu_genes
        ),
        intersect(
          cerebellum_striatum_dte_genes,
          cerebellum_striatum_dtu_genes
        )
      )
    )),
    # dge and dte
    length(intersect(
      cerebellum_striatum_dge_genes,
      cerebellum_striatum_dte_genes
    )) -
      length(intersect(
        intersect(
          cerebellum_striatum_dge_genes,
          cerebellum_striatum_dte_genes
        ),
        cerebellum_striatum_dtu_genes
      )),
    # dge and dtu
    length(intersect(
      cerebellum_striatum_dge_genes,
      cerebellum_striatum_dtu_genes
    )) -
      length(intersect(
        intersect(
          cerebellum_striatum_dge_genes,
          cerebellum_striatum_dte_genes
        ),
        cerebellum_striatum_dtu_genes
      )),
    # dte and dtu
    length(intersect(
      cerebellum_striatum_dtu_genes,
      cerebellum_striatum_dte_genes
    )) -
      length(intersect(
        intersect(
          cerebellum_striatum_dge_genes,
          cerebellum_striatum_dte_genes
        ),
        cerebellum_striatum_dtu_genes
      )),
    # all
    length(intersect(
      intersect(
        cerebellum_striatum_dge_genes,
        cerebellum_striatum_dte_genes
      ),
      cerebellum_striatum_dtu_genes
    )),

    # dge only
    length(setdiff(
      cortex_hippocampus_dge_genes,
      union(
        intersect(
          cortex_hippocampus_dge_genes,
          cortex_hippocampus_dte_genes
        ),
        intersect(
          cortex_hippocampus_dge_genes,
          cortex_hippocampus_dtu_genes
        )
      )
    )),
    # dte only
    length(setdiff(
      cortex_hippocampus_dte_genes,
      union(
        intersect(
          cortex_hippocampus_dge_genes,
          cortex_hippocampus_dte_genes
        ),
        intersect(
          cortex_hippocampus_dte_genes,
          cortex_hippocampus_dtu_genes
        )
      )
    )),
    # dtu only
    length(setdiff(
      cortex_hippocampus_dtu_genes,
      union(
        intersect(
          cortex_hippocampus_dge_genes,
          cortex_hippocampus_dtu_genes
        ),
        intersect(
          cortex_hippocampus_dte_genes,
          cortex_hippocampus_dtu_genes
        )
      )
    )),
    # dge and dte
    length(intersect(
      cortex_hippocampus_dge_genes,
      cortex_hippocampus_dte_genes
    )) -
      length(intersect(intersect(
        cortex_hippocampus_dge_genes,
        cortex_hippocampus_dte_genes
      ), cortex_hippocampus_dtu_genes)),
    # dge and dtu
    length(intersect(
      cortex_hippocampus_dge_genes,
      cortex_hippocampus_dtu_genes
    )) - length(intersect(intersect(
      cortex_hippocampus_dge_genes,
      cortex_hippocampus_dte_genes
    ), cortex_hippocampus_dtu_genes)),
    # dte and dtu
    length(intersect(
      cortex_hippocampus_dtu_genes,
      cortex_hippocampus_dte_genes
    )) - length(intersect(intersect(
      cortex_hippocampus_dge_genes,
      cortex_hippocampus_dte_genes
    ), cortex_hippocampus_dtu_genes)),
    # all
    length(intersect(
      intersect(
        cortex_hippocampus_dge_genes,
        cortex_hippocampus_dte_genes
      ),
      cortex_hippocampus_dtu_genes
    )),

    # dge only
    length(setdiff(
      cortex_striatum_dge_genes,
      union(
        intersect(
          cortex_striatum_dge_genes,
          cortex_striatum_dte_genes
        ),
        intersect(cortex_striatum_dge_genes, cortex_striatum_dtu_genes)
      )
    )),
    # dte only
    length(setdiff(
      cortex_striatum_dte_genes,
      union(
        intersect(
          cortex_striatum_dge_genes,
          cortex_striatum_dte_genes
        ),
        intersect(
          cortex_striatum_dte_genes,
          cortex_striatum_dtu_genes
        )
      )
    )),
    # dtu only
    length(setdiff(
      cortex_striatum_dtu_genes,
      union(
        intersect(
          cortex_striatum_dge_genes,
          cortex_striatum_dtu_genes
        ),
        intersect(
          cortex_striatum_dte_genes,
          cortex_striatum_dtu_genes
        )
      )
    )),
    # dge and dte
    length(intersect(cortex_striatum_dge_genes, cortex_striatum_dte_genes)) -
      length(intersect(
        intersect(
          cortex_striatum_dge_genes,
          cortex_striatum_dte_genes
        ),
        cortex_striatum_dtu_genes
      )),
    # dge and dtu
    length(intersect(cortex_striatum_dge_genes, cortex_striatum_dtu_genes)) -
      length(intersect(
        intersect(
          cortex_striatum_dge_genes,
          cortex_striatum_dte_genes
        ),
        cortex_striatum_dtu_genes
      )),
    # dte and dtu
    length(intersect(cortex_striatum_dtu_genes, cortex_striatum_dte_genes)) -
      length(intersect(
        intersect(
          cortex_striatum_dge_genes,
          cortex_striatum_dte_genes
        ),
        cortex_striatum_dtu_genes
      )),
    # all
    length(intersect(
      intersect(
        cortex_striatum_dge_genes,
        cortex_striatum_dte_genes
      ),
      cortex_striatum_dtu_genes
    )),

    # dge only
    length(setdiff(
      striatum_hippocampus_dge_genes,
      union(
        intersect(
          striatum_hippocampus_dge_genes,
          striatum_hippocampus_dte_genes
        ),
        intersect(
          striatum_hippocampus_dge_genes,
          hippocampus_striatum_dtu_genes
        )
      )
    )),
    # dte only
    length(setdiff(
      hippocampus_striatum_dte_genes,
      union(
        intersect(
          striatum_hippocampus_dge_genes,
          striatum_hippocampus_dte_genes
        ),
        intersect(
          hippocampus_striatum_dte_genes,
          hippocampus_striatum_dtu_genes
        )
      )
    )),
    # dtu only
    length(setdiff(
      hippocampus_striatum_dtu_genes,
      union(
        intersect(
          striatum_hippocampus_dge_genes,
          hippocampus_striatum_dtu_genes
        ),
        intersect(
          striatum_hippocampus_dte_genes,
          hippocampus_striatum_dtu_genes
        )
      )
    )),
    # dge and dte
    length(intersect(
      striatum_hippocampus_dge_genes,
      striatum_hippocampus_dte_genes
    )) -
      length(intersect(
        intersect(
          striatum_hippocampus_dge_genes,
          striatum_hippocampus_dte_genes
        ),
        hippocampus_striatum_dtu_genes
      )),
    # dge and dtu
    length(intersect(
      striatum_hippocampus_dge_genes,
      hippocampus_striatum_dtu_genes
    )) -
      length(intersect(
        intersect(
          striatum_hippocampus_dge_genes,
          striatum_hippocampus_dte_genes
        ),
        hippocampus_striatum_dtu_genes
      )),
    # dte and dtu
    length(intersect(
      hippocampus_striatum_dtu_genes,
      striatum_hippocampus_dte_genes
    )) -
      length(intersect(
        intersect(
          striatum_hippocampus_dge_genes,
          striatum_hippocampus_dte_genes
        ),
        hippocampus_striatum_dtu_genes
      )),
    # all
    length(intersect(
      intersect(
        striatum_hippocampus_dge_genes,
        striatum_hippocampus_dte_genes
      ),
      hippocampus_striatum_dtu_genes
    ))
  )
)
```

Now that we have made the dataframe, we can make the ggplot.

```{r create stacked paired barplot}
stacked_barplot_paired_plot <-
  ggplot(
    stacked_barplot_paired,
    aes(fill = analysis, y = n_genes, x = brain_regions)
  ) +
  geom_bar(position = "fill", stat = "identity") +
  scale_fill_manual(values = EJ_custom_palette) +
  xlab("Brain Regions") +
  ylab("Proportion of Total Genes") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1))

stacked_barplot_paired_plot

# make horizontal
stacked_barplot_paired_plot_horiz_rot <-
  ggplot(
    stacked_barplot_paired,
    aes(fill = analysis, y = n_genes, x = brain_regions)
  ) +
  geom_bar(position = "fill", stat = "identity") +
  scale_fill_manual(values = EJ_custom_palette) +
  xlab("Brain Regions") +
  ylab("Proportion of Total Genes") +
  theme_minimal() +
  theme(axis.text.y = element_text(angle = 45, hjust = .8)) +
  coord_flip() +
  theme(text = element_text(face = "bold"))

stacked_barplot_paired_plot_horiz_rot

# try without rotated labels
stacked_barplot_paired_plot_horiz <-
  ggplot(
    stacked_barplot_paired,
    aes(fill = Analysis, y = n_genes, x = brain_regions)
  ) +
  geom_bar(position = "fill", stat = "identity") +
  scale_fill_manual(values = EJ_custom_palette) +
  xlab("Brain Regions") +
  ylab("Proportion of Total Genes") +
  theme_minimal() +
  coord_flip() +
  theme(text = element_text(face = "bold"), legend.position = "bottom",
        axis.text.y = element_text(color = "black")) +
  guides(fill = guide_legend(nrow = 4))

stacked_barplot_paired_plot_horiz

ggsave(
  here(
    "results/plots/bar_plots/stacked_region_region_overlap.png",
    stacked_barplot_paired_plot
  ),
  width = 6, height = 4
)
```

Now that I have all of the individual plots we are interested in, we can assemble figure 2.

```{r assemble figure 2}
# export complexheatmap objects
dge_pairwise_upset_plot_grab <- grid.grabExpr(draw(dge_pairwise_upset_plot))
dte_pairwise_upset_plot_grab <- grid.grabExpr(draw(dte_pairwise_upset_plot))
dtu_pairwise_upset_plot_grab <- grid.grabExpr(draw(dtu_pairwise_upset_plot))

# make left side
left_side <-
  plot_grid(dge_pairwise_upset_plot_grab, dte_pairwise_upset_plot_grab,
    dtu_pairwise_upset_plot_grab,
    ncol = 1
  )

left_side

right_side <- plot_grid(NULL, stacked_barplot_paired_plot_horiz, NULL,
                        ncol = 1, rel_heights = c(0.2, 1, 0.2))

right_side

full_figure2 <-
  plot_grid(left_side, right_side, rel_widths = c(3, 1))

full_figure2

full_figure2 <-
  plot_grid(left_side, stacked_barplot_paired_plot_horiz, rel_widths = c(3, 1))

full_figure2

```

```{r try making upset plots smaller}
# dtu
dtu_comb_mat_small <- dtu_overlap_list_comb_mat[comb_size(dtu_overlap_list_comb_mat) >= 5]

dtu_comb_mat_small

dtu_pairwise_upset_plot_smaller <- UpSet(dtu_comb_mat_small,
  comb_order = order(-comb_size(dtu_comb_mat_small)),
  comb_col = "#99C945",
  top_annotation = upset_top_annotation(dtu_comb_mat_small,
    add_numbers = TRUE, annotation_name_gp = gpar(fontface = "bold"),
    axis_param = list(gp = gpar(fontface = "bold")),
    numbers_gp = gpar(fontface = "bold")
  ),
  right_annotation = upset_right_annotation(dtu_comb_mat_small,
    add_numbers = TRUE, annotation_name_gp = gpar(fontface = "bold"),
    axis_param = list(gp = gpar(fontface = "bold")),
    numbers_gp = gpar(fontface = "bold")
  ),
  row_names_gp = gpar(fontface = "bold")
)

dtu_pairwise_upset_plot_smaller

# dge
dge_comb_mat_small <- dge_overlap_list_comb_mat[comb_size(dge_overlap_list_comb_mat) >= 40]

dge_comb_mat_small

dge_pairwise_upset_plot_smaller <- UpSet(dge_comb_mat_small,
  comb_order = order(-comb_size(dge_comb_mat_small)),
  comb_col = "#5D69B1",
  top_annotation = upset_top_annotation(dge_comb_mat_small,
    add_numbers = TRUE, annotation_name_gp = gpar(fontface = "bold"),
    axis_param = list(gp = gpar(fontface = "bold")),
    numbers_gp = gpar(fontface = "bold")
  ),
  right_annotation = upset_right_annotation(dge_comb_mat_small,
    add_numbers = TRUE, annotation_name_gp = gpar(fontface = "bold"),
    axis_param = list(gp = gpar(fontface = "bold")),
    numbers_gp = gpar(fontface = "bold")
  ),
  row_names_gp = gpar(fontface = "bold")
)

dge_pairwise_upset_plot_smaller

# dte
dte_comb_mat_small <- dte_overlap_list_comb_mat[comb_size(dte_overlap_list_comb_mat) >= 40]

dte_comb_mat_small

dte_pairwise_upset_plot_smaller <- UpSet(dte_comb_mat_small,
  comb_order = order(-comb_size(dte_comb_mat_small)),
  comb_col = "#52BCA3",
  top_annotation = upset_top_annotation(dte_comb_mat_small,
    add_numbers = TRUE, annotation_name_gp = gpar(fontface = "bold"),
    axis_param = list(gp = gpar(fontface = "bold")),
    numbers_gp = gpar(fontface = "bold")
  ),
  right_annotation = upset_right_annotation(dte_comb_mat_small,
    add_numbers = TRUE, annotation_name_gp = gpar(fontface = "bold"),
    axis_param = list(gp = gpar(fontface = "bold")),
    numbers_gp = gpar(fontface = "bold")
  ),
  row_names_gp = gpar(fontface = "bold")
)

dte_pairwise_upset_plot_smaller

```

```{r assemble figure 2 with smaller upsets}
# export complexheatmap objects
dge_pairwise_upset_plot_grab <- grid.grabExpr(draw(dge_pairwise_upset_plot_smaller))
dte_pairwise_upset_plot_grab <- grid.grabExpr(draw(dte_pairwise_upset_plot_smaller))
dtu_pairwise_upset_plot_grab <- grid.grabExpr(draw(dtu_pairwise_upset_plot_smaller))

# make left side
left_side <-
  plot_grid(dge_pairwise_upset_plot_grab, dte_pairwise_upset_plot_grab,
    dtu_pairwise_upset_plot_grab,
    ncol = 1
  )

left_side

right_side <- plot_grid(NULL, stacked_barplot_paired_plot_horiz, NULL,
                        ncol = 1, rel_heights = c(0.4, 1, 0.2))

final_figure2 <-
  plot_grid(left_side, right_side, rel_widths = c(1, 0.7))

final_figure2

png(here("results", "finished_figures", "figure2.png"), width = 12, height = 9, units = "in", res = 300)
final_figure2
dev.off()

```

#### Clean up script

```{r tidy script}
style_file("15_compare_DTU_DGE.Rmd")
lint("15_compare_DTU_DGE.Rmd",
  linters = linters_with_defaults(
    object_length_linter = NULL,
    object_name_linter = NULL,
    object_usage_linter = NULL
  )
)
```

Also, needed to get processing time.

```{r proc time finish}
fptm <- proc.time() - ptm
fptm[3] / 60
```

#### Software versions

My software versions will be commented below.

```{r versions}
sessionInfo()
```

R version 4.3.0 beta (2023-04-12 r84240)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.2 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
 [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] grid      stats4    stats     graphics  grDevices utils     datasets  methods  
[9] base     

other attached packages:
 [1] here_1.0.1                    lintr_3.0.2                  
 [3] styler_1.9.1                  ggvenn_0.1.10                
 [5] viridis_0.6.2                 viridisLite_0.4.2            
 [7] ComplexHeatmap_2.16.0         IsoformSwitchAnalyzeR_1.99.17
 [9] pfamAnalyzeR_1.0.0            sva_3.48.0                   
[11] genefilter_1.82.0             mgcv_1.8-42                  
[13] nlme_3.1-162                  satuRn_1.8.0                 
[15] DEXSeq_1.46.0                 RColorBrewer_1.1-3           
[17] AnnotationDbi_1.62.0          DESeq2_1.40.0                
[19] SummarizedExperiment_1.30.0   GenomicRanges_1.52.0         
[21] GenomeInfoDb_1.36.0           IRanges_2.34.0               
[23] S4Vectors_0.38.0              MatrixGenerics_1.12.0        
[25] matrixStats_0.63.0            Biobase_2.60.0               
[27] BiocGenerics_0.46.0           BiocParallel_1.34.0          
[29] limma_3.56.0                  lubridate_1.9.2              
[31] forcats_1.0.0                 stringr_1.5.0                
[33] dplyr_1.1.2                   purrr_1.0.1                  
[35] readr_2.1.4                   tidyr_1.3.0                  
[37] tibble_3.2.1                  ggplot2_3.4.2                
[39] tidyverse_2.0.0              

loaded via a namespace (and not attached):
  [1] splines_4.3.0                 later_1.3.0                  
  [3] BiocIO_1.10.0                 bitops_1.0-7                 
  [5] filelock_1.0.2                R.oo_1.25.0                  
  [7] rex_1.2.1                     XML_3.99-0.14                
  [9] lifecycle_1.0.3               rprojroot_2.0.3              
 [11] edgeR_3.42.0                  doParallel_1.0.17            
 [13] processx_3.8.0                lattice_0.21-8               
 [15] ensembldb_2.24.0              magrittr_2.0.3               
 [17] rmarkdown_2.21                remotes_2.4.2                
 [19] yaml_2.3.7                    httpuv_1.6.9                 
 [21] pbapply_1.7-0                 DBI_1.1.3                    
 [23] zlibbioc_1.46.0               R.cache_0.16.0               
 [25] R.utils_2.12.2                AnnotationFilter_1.24.0      
 [27] RCurl_1.98-1.12               rappdirs_0.3.3               
 [29] circlize_0.4.15               GenomeInfoDbData_1.2.10      
 [31] annotate_1.78.0               codetools_0.2-19             
 [33] DelayedArray_0.25.0           xml2_1.3.4                   
 [35] tidyselect_1.2.0              shape_1.4.6                  
 [37] futile.logger_1.4.3           locfdr_1.1-8                 
 [39] BiocFileCache_2.8.0           GenomicAlignments_1.36.0     
 [41] jsonlite_1.8.4                GetoptLong_1.0.5             
 [43] ellipsis_0.3.2                survival_3.5-5               
 [45] iterators_1.0.14              foreach_1.5.2                
 [47] tools_4.3.0                   progress_1.2.2               
 [49] Rcpp_1.0.10                   glue_1.6.2                   
 [51] gridExtra_2.3                 xfun_0.39                    
 [53] withr_2.5.0                   formatR_1.14                 
 [55] BiocManager_1.30.20           fastmap_1.1.1                
 [57] boot_1.3-28.1                 fansi_1.0.4                  
 [59] callr_3.7.3                   digest_0.6.31                
 [61] timechange_0.2.0              R6_2.5.1                     
 [63] mime_0.12                     colorspace_2.1-0             
 [65] biomaRt_2.56.0                RSQLite_2.3.1                
 [67] R.methodsS3_1.8.2             utf8_1.2.3                   
 [69] generics_0.1.3                tximeta_1.18.0               
 [71] rtracklayer_1.60.0            prettyunits_1.1.1            
 [73] httr_1.4.5                    pkgconfig_2.0.3              
 [75] gtable_0.3.3                  blob_1.2.4                   
 [77] hwriter_1.3.2.1               XVector_0.40.0               
 [79] htmltools_0.5.5               geneplotter_1.78.0           
 [81] ProtGenerics_1.32.0           clue_0.3-64                  
 [83] scales_1.2.1                  cyclocomp_1.1.0              
 [85] png_0.1-8                     knitr_1.42                   
 [87] lambda.r_1.2.4                rstudioapi_0.14              
 [89] tzdb_0.3.0                    reshape2_1.4.4               
 [91] rjson_0.2.21                  curl_5.0.0                   
 [93] cachem_1.0.8                  GlobalOptions_0.1.2          
 [95] BiocVersion_3.17.1            parallel_4.3.0               
 [97] desc_1.4.2                    restfulr_0.0.15              
 [99] pillar_1.9.0                  vctrs_0.6.3                  
[101] promises_1.2.0.1              dbplyr_2.3.2                 
[103] xtable_1.8-4                  cluster_2.1.4                
[105] tximport_1.28.0               evaluate_0.20                
[107] VennDiagram_1.7.3             GenomicFeatures_1.52.0       
[109] cli_3.6.1                     locfit_1.5-9.7               
[111] fptm futile.options_1.0.1         
[113] Rsamtools_2.16.0              rlang_1.1.1                  
[115] crayon_1.5.2                  ps_1.7.4                     
[117] plyr_1.8.8                    stringi_1.7.12               
[119] munsell_0.5.0                 Biostrings_2.68.0            
[121] lazyeval_0.2.2                Matrix_1.5-4                 
[123] BSgenome_1.68.0               hms_1.1.3                    
[125] bit64_4.0.5                   KEGGREST_1.40.0              
[127] statmod_1.5.0                 shiny_1.7.4                  
[129] interactiveDisplayBase_1.38.0 AnnotationHub_3.8.0          
[131] memoise_2.0.1                 bit_4.0.5                    
[133] xmlparsedata_1.0.5     
