---
title: "Functional enrichment analysis"
author: "Emma Jones"
date: "2023-10-11"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Functional enrichment analysis script

The purpose of this script is to perform functional enrichment analysis with the package gprofiler2 on the DGE and DTE genes. The DTU genes already had it performed back in scripts 5-7. I'm most interested in the pairwise comparisons since they had the most significant results, so I'm running gprofiler on those first. I have edited the gprofiler function to accommodate any background gene list. This script is fully dependent on scripts 00-12.

### Load in packages

```{r load packages and set seed}
suppressPackageStartupMessages({
  library(tidyverse)
  library(IsoformSwitchAnalyzeR)
  library(ComplexHeatmap)
  library(viridis)
  library(org.Mm.eg.db)
  library(gprofiler2)
  library(styler)
  library(lintr)
  library(here)
})

set.seed(123)
```


```{r read in functions}
source("../dtu_analyses/functions.R", local = knitr::knit_global())
```


```{r load in DGE results}
de_results <- list.files(here("data", "deseq2_data"))

dge_results <- de_results[!grepl("transcripts", de_results)]

# loop in files
for (i in dge_results) {
  temp <- readRDS(
    here("data", "deseq2_data", i)
  )
  results <- as.data.frame(temp)

  sig_results <- subset(results, padj < 0.05)
  sig_results <- subset(sig_results, abs(log2FoldChange) > 1.5)

  sig_results <-
    na.omit(str_extract(rownames(sig_results), "ENSMUSG..........."))

  assign(paste0(str_sub(i, end = -5)),
    results,
    envir = .GlobalEnv
  )

  assign(paste0("sig_", str_sub(i, end = -5)),
    sig_results,
    envir = .GlobalEnv
  )
}
```

The current issue with these results is that the transcript results are not mapped to genes. I will have to map these to genes first.

```{r load in only DTE results}
# map txs to genes based on counts table
transcript_counts <- read.table(
  here("data", "nextflow", "bambu", "counts_transcript.txt"),
  header = TRUE
)

tx_gene_names <- transcript_counts[, 1:2]

# list files
dte_results <- de_results[grepl("transcripts", de_results)]

# loop in files
for (i in dte_results) {
  temp <- readRDS(
    here("data", "deseq2_data", i)
  )

  results <- as.data.frame(temp)

  sig_results <- subset(results, padj < 0.05)
  sig_results <- subset(sig_results, abs(log2FoldChange) > 1.5)

  sig_results <- unique(
    tx_gene_names$GENEID[
      tx_gene_names$TXNAME %in% rownames(sig_results)
    ]
  )
  sig_results <- na.omit(str_extract(sig_results, "ENSMUSG..........."))

  assign(paste0(str_sub(i, end = -5)),
    results,
    envir = .GlobalEnv
  )

  assign(paste0("sig_", str_sub(i, end = -5)),
    sig_results,
    envir = .GlobalEnv
  )
}
```

Can also get background genes in this way

```{r get bg genes from txs}
pairwise_dte_genes <- unique(
  tx_gene_names$GENEID[
    tx_gene_names$TXNAME %in% rownames(cerebellum_cortex_transcripts_results)
  ]
)
```

The pairwise results DO share the same background genes, but the single regions and single region sex DO NOT share the same background genes. For bg genes, you can just get the rownames for gene level. For transcript level, I'll need to pull the gene names of all the transcripts that were measured the way we got above. I'm just going to start with pairwise comparisons, since thatis what we use in the manuscript anyway.

```{r create background genes list}
pairwise_dge_genes <- rownames(cerebellum_cortex_results)

# data wrangle the ensembl IDs for gprofiler and drop novel genes
pairwise_dge_genes <-
  na.omit(str_extract(pairwise_dge_genes, "ENSMUSG..........."))

pairwise_dte_genes <-
  na.omit(str_extract(pairwise_dte_genes, "ENSMUSG..........."))
```

```{r run gprofiler on dge genes}
# run gprofiler2
run_plot_gprofiler(
  gene_list = sig_cerebellum_cortex_results,
  name = "cerebellum_cortex",
  custom_bg_genes = as.vector(pairwise_dge_genes),
  plot_save_path = here("results", "plots", "gprofiler2", "dge"),
  csv_save_path = here("results", "gprofiler2", "dge")
)

run_plot_gprofiler(
  gene_list = sig_cerebellum_hippocampus_results,
  name = "cerebellum_hippocampus",
  custom_bg_genes = as.vector(pairwise_dge_genes),
  plot_save_path = here("results", "plots", "gprofiler2", "dge"),
  csv_save_path = here("results", "gprofiler2", "dge")
)

run_plot_gprofiler(
  gene_list = sig_cerebellum_striatum_results,
  name = "cerebellum_striatum",
  custom_bg_genes = as.vector(pairwise_dge_genes),
  plot_save_path = here("results", "plots", "gprofiler2", "dge"),
  csv_save_path = here("results", "gprofiler2", "dge")
)

run_plot_gprofiler(
  gene_list = sig_cortex_hippocampus_results,
  name = "cortex_hippocampus",
  custom_bg_genes = as.vector(pairwise_dge_genes),
  plot_save_path = here("results", "plots", "gprofiler2", "dge"),
  csv_save_path = here("results", "gprofiler2", "dge")
)

run_plot_gprofiler(
  gene_list = sig_cortex_striatum_results,
  name = "cortex_striatum",
  custom_bg_genes = as.vector(pairwise_dge_genes),
  plot_save_path = here("results", "plots", "gprofiler2", "dge"),
  csv_save_path = here("results", "gprofiler2", "dge")
)

run_plot_gprofiler(
  gene_list = sig_striatum_hippocampus_results,
  name = "striatum_hippocampus",
  custom_bg_genes = as.vector(pairwise_dge_genes),
  plot_save_path = here("results", "plots", "gprofiler2", "dge"),
  csv_save_path = here("results", "gprofiler2", "dge")
)
```

Now, do the same thing for DTE.

```{r run gprofiler on dte genes}
run_plot_gprofiler(
  gene_list = sig_cerebellum_cortex_transcripts_results,
  name = "cerebellum_cortex",
  custom_bg_genes = as.vector(pairwise_dte_genes),
  plot_save_path = here("results", "plots", "gprofiler2", "dte"),
  csv_save_path = here("results", "gprofiler2", "dte")
)

run_plot_gprofiler(
  gene_list = sig_cerebellum_hippocampus_transcripts_results,
  name = "cerebellum_hippocampus",
  custom_bg_genes = as.vector(pairwise_dte_genes),
  plot_save_path = here("results", "plots", "gprofiler2", "dte"),
  csv_save_path = here("results", "gprofiler2", "dte")
)

run_plot_gprofiler(
  gene_list = sig_cerebellum_striatum_transcripts_results,
  name = "cerebellum_striatum",
  custom_bg_genes = as.vector(pairwise_dte_genes),
  plot_save_path = here("results", "plots", "gprofiler2", "dte"),
  csv_save_path = here("results", "gprofiler2", "dte")
)

run_plot_gprofiler(
  gene_list = sig_cortex_hippocampus_transcripts_results,
  name = "cortex_hippocampus",
  custom_bg_genes = as.vector(pairwise_dte_genes),
  plot_save_path = here("results", "plots", "gprofiler2", "dte"),
  csv_save_path = here("results", "gprofiler2", "dte")
)

run_plot_gprofiler(
  gene_list = sig_cortex_striatum_transcripts_results,
  name = "cortex_striatum",
  custom_bg_genes = as.vector(pairwise_dte_genes),
  plot_save_path = here("results", "plots", "gprofiler2", "dte"),
  csv_save_path = here("results", "gprofiler2", "dte")
)

run_plot_gprofiler(
  gene_list = sig_striatum_hippocampus_transcripts_results,
  name = "striatum_hippocampus",
  custom_bg_genes = as.vector(pairwise_dte_genes),
  plot_save_path = here("results", "plots", "gprofiler2", "dte"),
  csv_save_path = here("results", "gprofiler2", "dte")
)
```

#### Tidy script

```{r tidy script}
style_file("20_enrichment_analysis.Rmd")
lint("20_enrichment_analysis.Rmd",
  linters = linters_with_defaults(
    object_length_linter = NULL,
    object_name_linter = NULL,
    object_usage_linter = NULL
  )
)
```


#### Software versions

```{r versions}
sessionInfo()
```
R version 4.3.1 (2023-06-16)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.3 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
 [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] grid      stats4    stats     graphics  grDevices utils     datasets  methods  
[9] base     

other attached packages:
 [1] here_1.0.1                  lintr_3.1.0                 styler_1.10.2              
 [4] gprofiler2_0.2.2            org.Mm.eg.db_3.17.0         viridis_0.6.4              
 [7] viridisLite_0.4.2           ComplexHeatmap_2.16.0       IsoformSwitchAnalyzeR_2.1.3
[10] pfamAnalyzeR_1.0.1          sva_3.48.0                  genefilter_1.82.1          
[13] mgcv_1.9-0                  nlme_3.1-163                satuRn_1.8.0               
[16] DEXSeq_1.46.0               RColorBrewer_1.1-3          AnnotationDbi_1.62.2       
[19] DESeq2_1.40.2               SummarizedExperiment_1.30.2 GenomicRanges_1.52.0       
[22] GenomeInfoDb_1.36.3         IRanges_2.34.1              S4Vectors_0.38.1           
[25] MatrixGenerics_1.12.3       matrixStats_1.0.0           Biobase_2.60.0             
[28] BiocGenerics_0.46.0         BiocParallel_1.34.2         limma_3.56.2               
[31] lubridate_1.9.2             forcats_1.0.0               stringr_1.5.0              
[34] dplyr_1.1.3                 purrr_1.0.2                 readr_2.1.4                
[37] tidyr_1.3.0                 tibble_3.2.1                ggplot2_3.4.3              
[40] tidyverse_2.0.0            

loaded via a namespace (and not attached):
  [1] splines_4.3.1                 later_1.3.1                  
  [3] BiocIO_1.10.0                 bitops_1.0-7                 
  [5] filelock_1.0.2                R.oo_1.25.0                  
  [7] rex_1.2.1                     XML_3.99-0.14                
  [9] lifecycle_1.0.3               rprojroot_2.0.3              
 [11] edgeR_3.42.4                  doParallel_1.0.17            
 [13] processx_3.8.2                lattice_0.21-8               
 [15] ensembldb_2.24.0              backports_1.4.1              
 [17] magrittr_2.0.3                plotly_4.10.2                
 [19] rmarkdown_2.24                remotes_2.4.2.1              
 [21] yaml_2.3.7                    httpuv_1.6.11                
 [23] pbapply_1.7-2                 DBI_1.1.3                    
 [25] abind_1.4-5                   zlibbioc_1.46.0              
 [27] R.cache_0.16.0                R.utils_2.12.2               
 [29] AnnotationFilter_1.24.0       RCurl_1.98-1.12              
 [31] rappdirs_0.3.3                circlize_0.4.15              
 [33] GenomeInfoDbData_1.2.10       annotate_1.78.0              
 [35] codetools_0.2-19              DelayedArray_0.26.7          
 [37] xml2_1.3.5                    tidyselect_1.2.0             
 [39] shape_1.4.6                   futile.logger_1.4.3          
 [41] locfdr_1.1-8                  farver_2.1.1                 
 [43] BiocFileCache_2.8.0           GenomicAlignments_1.36.0     
 [45] jsonlite_1.8.7                GetoptLong_1.0.5             
 [47] ellipsis_0.3.2                survival_3.5-7               
 [49] iterators_1.0.14              systemfonts_1.0.4            
 [51] foreach_1.5.2                 tools_4.3.1                  
 [53] progress_1.2.2                ragg_1.2.5                   
 [55] Rcpp_1.0.11                   glue_1.6.2                   
 [57] gridExtra_2.3                 xfun_0.40                    
 [59] withr_2.5.0                   formatR_1.14                 
 [61] BiocManager_1.30.22           fastmap_1.1.1                
 [63] boot_1.3-28.1                 fansi_1.0.4                  
 [65] callr_3.7.3                   digest_0.6.33                
 [67] timechange_0.2.0              R6_2.5.1                     
 [69] mime_0.12                     textshaping_0.3.6            
 [71] colorspace_2.1-0              biomaRt_2.56.1               
 [73] RSQLite_2.3.1                 R.methodsS3_1.8.2            
 [75] utf8_1.2.3                    generics_0.1.3               
 [77] data.table_1.14.8             tximeta_1.18.3               
 [79] rtracklayer_1.60.1            htmlwidgets_1.6.2            
 [81] prettyunits_1.1.1             httr_1.4.7                   
 [83] S4Arrays_1.0.6                pkgconfig_2.0.3              
 [85] gtable_0.3.4                  blob_1.2.4                   
 [87] hwriter_1.3.2.1               XVector_0.40.0               
 [89] htmltools_0.5.6               geneplotter_1.78.0           
 [91] ProtGenerics_1.32.0           clue_0.3-64                  
 [93] scales_1.2.1                  cyclocomp_1.1.1              
 [95] png_0.1-8                     knitr_1.44                   
 [97] lambda.r_1.2.4                rstudioapi_0.15.0            
 [99] tzdb_0.4.0                    reshape2_1.4.4               
[101] rjson_0.2.21                  curl_5.0.2                   
[103] cachem_1.0.8                  GlobalOptions_0.1.2          
[105] BiocVersion_3.17.1            parallel_4.3.1               
[107] desc_1.4.2                    restfulr_0.0.15              
[109] pillar_1.9.0                  vctrs_0.6.3                  
[111] promises_1.2.1                dbplyr_2.3.3                 
[113] cluster_2.1.4                 xtable_1.8-4                 
[115] tximport_1.28.0               evaluate_0.21                
[117] VennDiagram_1.7.3             GenomicFeatures_1.52.2       
[119] cli_3.6.1                     locfit_1.5-9.8               
[121] compiler_4.3.1                futile.options_1.0.1         
[123] Rsamtools_2.16.0              rlang_1.1.1                  
[125] crayon_1.5.2                  ps_1.7.5                     
[127] plyr_1.8.8                    stringi_1.7.12               
[129] munsell_0.5.0                 Biostrings_2.68.1            
[131] lazyeval_0.2.2                Matrix_1.6-1                 
[133] BSgenome_1.68.0               hms_1.1.3                    
[135] bit64_4.0.5                   KEGGREST_1.40.0              
[137] statmod_1.5.0                 shiny_1.7.5                  
[139] interactiveDisplayBase_1.38.0 AnnotationHub_3.8.0          
[141] memoise_2.0.1                 bit_4.0.5  
